// francium-sdk/dist/constants/smartFarming/orca/pools.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.activeSmartFarmingOrcaPools = exports.orcaSmartFarmingPools = exports.GLOBAL_FARM_DATA_LAYOUT = void 0;
const tokens_1 = require("../../tokens");
const web3_js_1 = require("@solana/web3.js");
const buffer_layout_1 = require("buffer-layout");
const borsh_1 = require("@project-serum/borsh");
const info_1 = require("../../../constants/farm/orca/info");
const base_1 = require("../base");
const lodash_1 = require("lodash");
exports.GLOBAL_FARM_DATA_LAYOUT = (0, buffer_layout_1.struct)([
    (0, buffer_layout_1.u8)("isInitialized"),
    (0, buffer_layout_1.u8)("accountType"),
    (0, buffer_layout_1.u8)("nonce"),
    (0, borsh_1.publicKey)("tokenProgramId"),
    (0, borsh_1.publicKey)("emissionsAuthority"),
    (0, borsh_1.publicKey)("removeRewardsAuthority"),
    (0, borsh_1.publicKey)("baseTokenMint"),
    (0, borsh_1.publicKey)("baseTokenVault"),
    (0, borsh_1.publicKey)("rewardTokenVault"),
    (0, borsh_1.publicKey)("farmTokenMint"),
    (0, borsh_1.u64)("emissionsPerSecondNumerator"),
    (0, borsh_1.u64)("emissionsPerSecondDenominator"),
    (0, borsh_1.u64)("lastUpdatedTimestamp"),
    (0, buffer_layout_1.blob)(32, "cumulativeEmissionsPerFarmToken"),
]);
exports.orcaSmartFarmingPools = {
    "ORCA-USDC": {
        smartPoolInfo: {
            poolType: base_1.POOL_VAULT_TYPE_ORCA,
            tokenMint0: tokens_1.TOKENS.USDC.mintAddress,
            tokenMint1: tokens_1.TOKENS.ORCA.mintAddress,
            liquidityTokenMint: new web3_js_1.PublicKey("n8Mpu28RjeYD7oUX3LG1tPxzhRZh3YYLRSHcHRdS3Zx"),
            rewardsTokenMint: tokens_1.TOKENS.FRC.mintAddress,
            basedRewardsTokenMint: tokens_1.TOKENS.ORCA.mintAddress,
            basedRewardsTokenMintB: undefined,
            poolInfoAccount: new web3_js_1.PublicKey('GWooQLRmaAnwYjQx4qAeSrctdGkKf2RF8oXRMdAnyNYm'),
            poolAuthority: new web3_js_1.PublicKey('31sjDreudj5Yhd4Hvx1BSKmGEQtasQoe3NLPLNnMBYtQ'),
            vaultManagerAccount: new web3_js_1.PublicKey('jLVVrKn132jvjiD27A4UadSWc9YcNmK6ZKbPBFKc95v'),
            tokenAccount0: new web3_js_1.PublicKey('7NkNkmHokq3Kw96NtNMU1WEdgUAp4PFkRx5KvGypJPz9'),
            tokenAccount1: new web3_js_1.PublicKey('8c12EGPi2fCVkWLQbi8AxJnGi4hBbwrEPpkK7BKCJpY3'),
            liquidityTokenAccount: new web3_js_1.PublicKey('3WdSJbGtBYkDg1VeMPbYkSqjSMt9eKvR7gu1o4dV3VsP'),
            farmInfoAccount: new web3_js_1.PublicKey('4ZGbh6woeEY9Ccm4TgNmFs3Zvynfggf8YuyWoaQi7PaK'),
            basedRewardsTokenAccount: new web3_js_1.PublicKey('8c12EGPi2fCVkWLQbi8AxJnGi4hBbwrEPpkK7BKCJpY3'),
            farmTokenAccount: new web3_js_1.PublicKey('9w1LcWUEKrQwpzFH3iTY7jMFvsr6z4Lq4xNqGDomg6o2'),
            borrowCreditTokenAccount0: new web3_js_1.PublicKey('FUayGPex5FX2rLLfY8pD4zTdYP7UCx8JxMF5TnKNHRiH'),
            borrowCreditTokenAccount1: new web3_js_1.PublicKey('5nrx5ZR79PjWFDgdUyafEqWbe6Vk1ifq6p5KEMKDb7YR'),
            farmInfoAccountB: undefined,
            basedRewardsTokenAccountB: undefined,
            farmTokenAccountB: undefined,
            rewardsTokenAccount: new web3_js_1.PublicKey('6aQxooPrWnpRuYbUTqgDXj5bA8cJihxzFGcpJKBjVFE7'),
            rewardsTargetTokenAccount: new web3_js_1.PublicKey('7NkNkmHokq3Kw96NtNMU1WEdgUAp4PFkRx5KvGypJPz9'),
            rewardsTargetTokenAccountB: undefined,
        },
        basedLendingInfo: {
            "0": {
                tknMint: tokens_1.TOKENS.USDC.mintAddress,
                lendingPoolInfoAccount: new web3_js_1.PublicKey("Hx6LbkMHe69DYawhPyVNs8Apa6tyfogfzQV6a7XkwBUU"),
                lendingPoolTknAccount: new web3_js_1.PublicKey("CFp9kt8z3Epb1QSiEp3xA44KbSwuJxhFR3wQoerFqYS9"),
                lendingPoolFeeAccount: new web3_js_1.PublicKey("6h6vVc42moDmRs7KahQpFs8YWL5GfWBrxT9UrsUvpApn"),
                lendingPoolShareMint: new web3_js_1.PublicKey("62fDf5daUJ9jBz8Xtj6Bmw1bh1DvHn8AG4L9hMmxCzpu"),
                lendingPoolShareAccount: new web3_js_1.PublicKey("Gyc1V1xbA9NjzuURE662ATw6W4AdhwvsL26yUnSGhbkz"),
                lendingPoolCreditMint: new web3_js_1.PublicKey("CZ2s85dnuAVyGbRWBNauHZwj9oTV2i9xTvFwc1Cedr7E"),
                lendingPoolCreditAccount: new web3_js_1.PublicKey("Bd71jEvypKFs8N5gByyhD2149tSqLy4PVAzFFShCTJTB"),
            },
            "1": {
                tknMint: tokens_1.TOKENS.ORCA.mintAddress,
                lendingPoolInfoAccount: new web3_js_1.PublicKey("9bo5JunUhp4XD7TDeeS9ARvWTYGBYtuQKUHGUb7RRvkf"),
                lendingPoolTknAccount: new web3_js_1.PublicKey("F3EhkuRuZaYNY2bLTosjv6V7QXiZ8db9nXTZeLnmHC7V"),
                lendingPoolFeeAccount: new web3_js_1.PublicKey("HBc25ZyXiGMERxUX2KupZjc3DAivA48wBSgZGxCQgtT6"),
                lendingPoolShareMint: new web3_js_1.PublicKey("DKoKBD7YheCGZjWp5CaHDPtheAcUveRSMvWkaivbNzWh"),
                lendingPoolShareAccount: new web3_js_1.PublicKey("AMUE9EgiEiTP1YVBicwTVGtumBypR5thjTgbsXZM2PPG"),
                lendingPoolCreditMint: new web3_js_1.PublicKey("CJNd1LZZxZr243dpNN9DtEavXn65kstrg251MPb8vmwW"),
                lendingPoolCreditAccount: new web3_js_1.PublicKey("74m1zyzV8x9d8kZLEaVaNWRsnx8pH47tfX8cquot11ef"),
            }
        },
        basedFarmInfo: {
            ammInfo: {
                swapProgramId: info_1.Orca_Swap_ProgramId,
                swapPoolId: new web3_js_1.PublicKey("2p7nYbtPBgtmY69NsE8DAW6szpRJn7tQvDnqvoEWQvjY"),
                swapPoolAuthority: new web3_js_1.PublicKey("3fr1AhdiAmWLeNrS24CMoAu9pPgbzVhwLtJ6QUPmw2ob"),
                lpMint: new web3_js_1.PublicKey("n8Mpu28RjeYD7oUX3LG1tPxzhRZh3YYLRSHcHRdS3Zx"),
                lpDecimals: 6,
                swapFeeAccount: new web3_js_1.PublicKey("7CXZED4jfRp3qdHB9Py3up6v1C4UhHofFvfT6RXbJLRN"),
                swapTknVault0: new web3_js_1.PublicKey("6UczejMUv1tzdvUzKpULKHxrK9sqLm8edR1v9jinVWm9"),
                swapTknVault1: new web3_js_1.PublicKey("9vYWHBPz817wJdQpE8u3h8UoY3sZ16ZXdCcvLB7jY4Dj"),
            },
            rewardsAmmInfo: {
                stakeProgramId: info_1.Orca_Stake_ProgramId,
                stakeUserFarmInfoLayout: info_1.ORCA_USER_FARM_INFO_ACCOUNT_LAYOUT,
                stakePoolRewardsMint: tokens_1.TOKENS.ORCA.mintAddress,
                stakePoolFarmInfo: new web3_js_1.PublicKey("9S1BsxbDNQXQccjFamVEGgxiYQHTeudvhEYwFr4oWeaf"),
                stakePoolAuthority: new web3_js_1.PublicKey("66xaEjFoYfRcspc18oDj61mXDyznr9zam6tFNeqvs2jK"),
                stakePoolFarmTknMint: new web3_js_1.PublicKey("Gc7W5U66iuHQcC1cQyeX9hxkPF2QUVJPTf1NWbW8fNrt"),
                stakePoolRewardsTknVault: new web3_js_1.PublicKey("DEiqe2Ta9TRMRtWdBqiFV13dhVrqCeG8MMmVwywvXvJo"),
                stakePoolBaseTokenVault: new web3_js_1.PublicKey("45BAAQCZYd2kP3Z3WvRwdtfUhvuW4FvpqVK4m8qrR5x1"),
                rewardsSwapPoolId: new web3_js_1.PublicKey("2p7nYbtPBgtmY69NsE8DAW6szpRJn7tQvDnqvoEWQvjY"),
                rewardsSwapPoolAuthority: new web3_js_1.PublicKey("3fr1AhdiAmWLeNrS24CMoAu9pPgbzVhwLtJ6QUPmw2ob"),
                rewardsTargetLpMint: new web3_js_1.PublicKey("n8Mpu28RjeYD7oUX3LG1tPxzhRZh3YYLRSHcHRdS3Zx"),
                rewardsSwapFeeAccount: new web3_js_1.PublicKey("7CXZED4jfRp3qdHB9Py3up6v1C4UhHofFvfT6RXbJLRN"),
                rewardsSwapPoolRewardsVault: new web3_js_1.PublicKey("9vYWHBPz817wJdQpE8u3h8UoY3sZ16ZXdCcvLB7jY4Dj"),
                rewardsSwapPoolTargetVault: new web3_js_1.PublicKey("6UczejMUv1tzdvUzKpULKHxrK9sqLm8edR1v9jinVWm9"),
            },
            rewardsAmmInfoB: undefined,
        },
    },
};
const activeSmartFarmingOrcaPools = [];
exports.activeSmartFarmingOrcaPools = activeSmartFarmingOrcaPools;
(0, lodash_1.forEach)(exports.orcaSmartFarmingPools, (value, key) => {
    activeSmartFarmingOrcaPools.push({
        pool: key,
        platform: 'orca',
        alias: value === null || value === void 0 ? void 0 : value.alias,
        poolInfoAccount: value.smartPoolInfo.poolInfoAccount,
        lpDecimals: value.basedFarmInfo.ammInfo.lpDecimals
    });
});
