// @apricot-lend/sdk-ts/dist/constants/commands.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ASSIST_MODE_STABLE_ONLY = exports.MAX_ASSIST_ACTIONS = exports.AMOUNT_MULTIPLIER = exports.INVALID_PAGE = exports.SWAP_ORCA = exports.SWAP_MERCURIAL = exports.SWAP_SABER = exports.SWAP_RAYDIUM = exports.SWAP_SERUM = exports.SWAP_FAKE = exports.CMD_LP_UNSTAKE_SECOND = exports.CMD_LP_STAKE_SECOND = exports.CMD_MAKE_LM_REWARD_AVAILABLE = exports.CMD_CLAIM_APT_LM_REWARD = exports.CMD_LP_UNSTAKE = exports.CMD_LP_STAKE = exports.CMD_LP_OP_ENDCHECK = exports.CMD_LP_OP_CHECK = exports.CMD_LP_REDEEM = exports.CMD_LP_CREATE = exports.CMD_WITHDRAW_AND_REMOVE_USER = exports.CMD_UPDATE_USER_ASSET_CONFIG = exports.CMD_MARGIN_SWAP = exports.CMD_UPDATE_USER_CONFIG = exports.CMD_EXTERN_LIQUIDATE = exports.CMD_REPAY = exports.CMD_BORROW = exports.CMD_WITHDRAW = exports.CMD_DEPOSIT = exports.CMD_ADD_USER_AND_DEPOSIT = exports.CMD_REFRESH_USER = void 0;
exports.CMD_REFRESH_USER = 0x0a;
exports.CMD_ADD_USER_AND_DEPOSIT = 0x10;
exports.CMD_DEPOSIT = 0x11;
exports.CMD_WITHDRAW = 0x12;
exports.CMD_BORROW = 0x13;
exports.CMD_REPAY = 0x14;
exports.CMD_EXTERN_LIQUIDATE = 0x15;
// DEPRECATED self-liquidate
exports.CMD_UPDATE_USER_CONFIG = 0x17;
exports.CMD_MARGIN_SWAP = 0x18;
exports.CMD_UPDATE_USER_ASSET_CONFIG = 0x19;
exports.CMD_WITHDRAW_AND_REMOVE_USER = 0x1a;
// export const CMD_TOKEN_DEPOSIT = 0x1b;
// export const CMD_TOKEN_WITHDRAW = 0x1c;
exports.CMD_LP_CREATE = 0x1d;
exports.CMD_LP_REDEEM = 0x1e;
exports.CMD_LP_OP_CHECK = 0x21;
exports.CMD_LP_OP_ENDCHECK = 0x22;
exports.CMD_LP_STAKE = 0x23;
exports.CMD_LP_UNSTAKE = 0x24;
exports.CMD_CLAIM_APT_LM_REWARD = 0x29;
exports.CMD_MAKE_LM_REWARD_AVAILABLE = 0x2a;
exports.CMD_LP_STAKE_SECOND = 0x81;
exports.CMD_LP_UNSTAKE_SECOND = 0x82;
// other trivia
// swap identifiers
exports.SWAP_FAKE = 0x00;
exports.SWAP_SERUM = 0x01;
exports.SWAP_RAYDIUM = 0x02;
exports.SWAP_SABER = 0x03;
exports.SWAP_MERCURIAL = 0x04;
exports.SWAP_ORCA = 0x05;
exports.INVALID_PAGE = 65535;
exports.AMOUNT_MULTIPLIER = 16777216;
exports.MAX_ASSIST_ACTIONS = 6;
exports.ASSIST_MODE_STABLE_ONLY = 2;
