// @port.finance/port-sdk/dist/cjs/Environment.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Environment = void 0;
const web3_js_1 = require("@solana/web3.js");
const models_1 = require("./models");
const constants_1 = require("./constants");
const AssetConfigs_1 = require("./utils/AssetConfigs");
const env_1 = require("./utils/env");
class Environment {
    constructor(env, lendingProgramPk, stakingProgramPk, tokenProgramPk, assetConfigs) {
        this.env = env;
        this.lendingProgramPk = lendingProgramPk;
        this.stakingProgramPk = stakingProgramPk;
        this.tokenProgramPk = tokenProgramPk;
        this.assetContext = models_1.AssetContext.index(assetConfigs);
    }
    static forMainNet() {
        return new Environment(env_1.ENV.MainnetBeta, constants_1.PORT_LENDING, constants_1.PORT_STAKING, new web3_js_1.PublicKey("TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA"), AssetConfigs_1.MAINNET_ASSETS);
    }
    getEnv() {
        return this.env;
    }
    getLendingProgramPk() {
        return this.lendingProgramPk;
    }
    getStakingProgramPk() {
        return this.stakingProgramPk;
    }
    getTokenProgramPk() {
        return this.tokenProgramPk;
    }
    getAssetContext() {
        return this.assetContext;
    }
}
exports.Environment = Environment;
