// @port.finance/port-sdk/dist/cjs/models/basic/Decimal.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DecimalField = exports.Decimal = void 0;
const big_js_1 = __importDefault(require("big.js"));
const BigField_1 = require("../../serialization/BigField");
const Field_1 = require("../../serialization/Field");
const Numerical_1 = require("./Numerical");
class Decimal extends Numerical_1.Numerical {
    constructor(raw) {
        super();
        this.raw = new big_js_1.default(raw);
    }
    getRaw() {
        return this.raw;
    }
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    isCompatibleWith(that) {
        return true;
    }
}
exports.Decimal = Decimal;
// eslint-disable-next-line @typescript-eslint/no-explicit-any
class DecimalField extends Field_1.Field {
    constructor(type, property) {
        super(type.getBytes(), property);
        this.delegate = BigField_1.BigField.forType(type, property);
    }
    decode(b, offset) {
        return this.fromBig(this.delegate.decode(b, offset));
    }
}
exports.DecimalField = DecimalField;
