// @port.finance/port-sdk/dist/cjs/models/basic/Lamport.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Lamport = void 0;
const spl_token_1 = require("@solana/spl-token");
const Decimal_1 = require("./Decimal");
class Lamport extends Decimal_1.Decimal {
    constructor(raw) {
        super(raw);
    }
    static zero() {
        return Lamport.ZERO;
    }
    static max() {
        return Lamport.MAX;
    }
    static of(raw) {
        return new Lamport(raw);
    }
    static field(type, property) {
        return new LamportField(type, property);
    }
    isTrivial() {
        return this.lt(Lamport.ONE);
    }
    isMax() {
        return this.eq(Lamport.MAX);
    }
    toU64() {
        // eslint-disable-next-line new-cap
        return new spl_token_1.u64(this.raw.toFixed(0, 0)); // RoundDown
    }
    replaceWithValue(value) {
        return Lamport.of(value);
    }
}
exports.Lamport = Lamport;
Lamport.ZERO = new Lamport(0);
Lamport.ONE = new Lamport(1);
Lamport.MAX = new Lamport("18446744073709551615");
class LamportField extends Decimal_1.DecimalField {
    constructor(type, property) {
        super(type, property);
    }
    fromBig(big) {
        return Lamport.of(big);
    }
}
