// @port.finance/port-sdk/dist/cjs/models/basic/Percentage.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Percentage = void 0;
const big_js_1 = __importDefault(require("big.js"));
const Decimal_1 = require("./Decimal");
const BigType_1 = require("../../serialization/BigType");
class Percentage extends Decimal_1.Decimal {
    constructor(value) {
        super(value);
    }
    static zero() {
        return Percentage.PCT_ZERO;
    }
    static hundred() {
        return Percentage.PCT_HUNDRED;
    }
    static fromOneBased(oneBased) {
        return Percentage.fromRaw(oneBased, false);
    }
    static fromHundredBased(hundredBased) {
        return Percentage.fromRaw(hundredBased, true);
    }
    static fromRaw(raw, isHundredBased) {
        const big = new big_js_1.default(raw);
        if (!isHundredBased) {
            return new Percentage(big);
        }
        return new Percentage(big.div(100));
    }
    static field(property) {
        return new PercentageField(true, property);
    }
    isTrivial() {
        return this.lt(Percentage.PCT_BIP);
    }
    isHundredPct() {
        return this.eq(Percentage.PCT_HUNDRED);
    }
    toOneBasedNumber(dp) {
        return this.raw.round(dp).toNumber();
    }
    toHundredBasedNumber(dp) {
        return this.raw.mul(100).round(dp, 0).toNumber();
    }
    print() {
        if (this.gt(Percentage.PCT_THOUSAND)) {
            return this.raw.round(1, 1).toString() + "x"; // RoundHalfUp
        }
        return this.raw.mul(100).round(2, 1).toString() + "%"; // RoundHalfUp
    }
    replaceWithValue(value) {
        return new Percentage(value);
    }
}
exports.Percentage = Percentage;
Percentage.PCT_BIP = new Percentage(0.0001);
Percentage.PCT_ZERO = new Percentage(0);
Percentage.PCT_HUNDRED = new Percentage(1);
Percentage.PCT_THOUSAND = new Percentage(10);
class PercentageField extends Decimal_1.DecimalField {
    constructor(isHundredBased, property) {
        super(BigType_1.BigType.U8, property);
        this.isHundredBased = isHundredBased;
    }
    fromBig(big) {
        return Percentage.fromRaw(big, this.isHundredBased);
    }
}
