// francium-sdk/dist/constants/smartLending/pools.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.activeSmartLendingPools = exports.smartLendingPools = exports.lendingPoolMarketInfo = exports.smartPoolMarketInfo = void 0;
const web3_js_1 = require("@solana/web3.js");
const lodash_1 = require("lodash");
const tokens_1 = require("../tokens");
exports.smartPoolMarketInfo = {
    marketInfoAccount: new web3_js_1.PublicKey("CNSCJKxXwu4vLbuLgadKFjjPxYEtxTvU5gXywvSZsKzm"),
    owner: new web3_js_1.PublicKey("7MBLg6oV5phip11YBbJPuq7u38kdzSi9PM3BifKSpLaR"),
    authority: new web3_js_1.PublicKey("Hac3juGDAL2wbuUaWgzGqxiF4kWDsjkm6T192r1siMNV"),
    authorityBump: 254,
    feeReceiver: new web3_js_1.PublicKey("7MBLg6oV5phip11YBbJPuq7u38kdzSi9PM3BifKSpLaR"),
    botAddress: new web3_js_1.PublicKey("7MBLg6oV5phip11YBbJPuq7u38kdzSi9PM3BifKSpLaR"),
};
exports.lendingPoolMarketInfo = {
    marketInfoAccount: new web3_js_1.PublicKey("4XNif294wbrxj6tJ8K5Rg7SuaEACnu9s2L27i28MQB6E"),
    marketAuthority: new web3_js_1.PublicKey("sCDiYj7X7JmXg5fVq2nqED2q1Wqjo7PnqMgH3casMem"),
    marketOwner: new web3_js_1.PublicKey("7MBLg6oV5phip11YBbJPuq7u38kdzSi9PM3BifKSpLaR"),
};
exports.smartLendingPools = {
    FRCCP: {
        smartPoolInfo: {
            liquidityTokenMint: tokens_1.TOKENS.FRCCP.mintAddress,
            rewardsTokenMint: tokens_1.TOKENS.FRCCP.mintAddress,
            poolInfoAccount: new web3_js_1.PublicKey('HPJMajbSRGihKjjWnPjoWbtjzQeamcJwWmnQYk34ET24'),
            poolAuthority: new web3_js_1.PublicKey('4fP3Xh7EtFeZpWJAhHPj55JTbdSX8fDy1eQL3GrSREVV'),
            vaultManagerAccount: new web3_js_1.PublicKey('CPcZ3xRw7DLaEcQjanK7kLt7MHsoMwHJ5u5XrZyMcTW5'),
            liquidityTokenAccount: new web3_js_1.PublicKey('AMNLVuTgepAVEUY8aKxmDJh48dkpX9D3jjTiF41AxrRn'),
            lendingShareAccount: new web3_js_1.PublicKey('tpod4ukA7P37v5TZtX5oEm4pSbkYh4Vwx7Q8JmbEZea'),
            rewardsTokenAccount: new web3_js_1.PublicKey('37SDME2qFFTgo19JqvXgL91TsKRgkW2oEGTLbhNYN5HE'),
        },
        basedLendingPool: {
            tknMint: tokens_1.TOKENS.FRCCP.mintAddress,
            lendingPoolInfoAccount: new web3_js_1.PublicKey('6yECBcn6eaXio8XVSAQBibFdKVTcy8XFVN3xrEyvg2Li'),
            lendingPoolTknAccount: new web3_js_1.PublicKey('2HGdKaLcnZywmaVpe6XWHmKMVtWzPqrQ5JRjk1WvVTid'),
            lendingPoolShareMint: new web3_js_1.PublicKey('BqtNZijgPyVBzhzS2PpHFH93KGyvWBiVPseEGexegJaB'),
            lendingPoolShareAccount: new web3_js_1.PublicKey('HqPpnAxXmMmNvqkh7sgu6HbZ32s7LKrtTtbXySNJKgyL'),
            lendingPoolCreditMint: new web3_js_1.PublicKey('5LKhwGY9xwHj2T5GNSjiRMfmH1qMHa2bJCDSt1aVGzu4'),
            lendingPoolCreditAccount: new web3_js_1.PublicKey('5JugdwBt4XTS4T61Woy8Cud2zmgpnKXxToKy86GroYSb'),
            lendingPoolFeeAccount: new web3_js_1.PublicKey('9fGuDMCyXqczyssLNe7BDSgmzFHGC3LHXoVPqeMnegrE'),
        }
    },
    "FRC": {
        smartPoolInfo: {
            liquidityTokenMint: tokens_1.TOKENS.FRC.mintAddress,
            rewardsTokenMint: tokens_1.TOKENS.FRC.mintAddress,
            poolInfoAccount: new web3_js_1.PublicKey('HoKqo4vXKwD1htBgX9CvBQgaPHvQn9p4dfzmirushswg'),
            poolAuthority: new web3_js_1.PublicKey('7VN1QgZzi2NfLiyUc6rpbsNGh1d8cXPyBBFHV6dKNnxz'),
            vaultManagerAccount: new web3_js_1.PublicKey('9WfmEHobYdfech4QuX4mqvvYQPuQXUYuRNeyPSt8nZLb'),
            liquidityTokenAccount: new web3_js_1.PublicKey('F3umFk1YtiU6hD9343KDU5uwsLMDfgMckRqeaYw6wbEY'),
            lendingShareAccount: new web3_js_1.PublicKey('8SSKc7wUqmB59p9Q9VdhDgiFCdbV9zLscHBKRUN1mAan'),
            rewardsTokenAccount: new web3_js_1.PublicKey('5pJEBbuCT2Y1XdWJc2teBkHbEhh5tS8C1dM1i8rKCpwk'),
        },
        basedLendingPool: {
            tknMint: tokens_1.TOKENS.FRC.mintAddress,
            lendingPoolInfoAccount: new web3_js_1.PublicKey('Cht1VvnoBDtBpU38WGoSu3VebvpmjWh377DRHgU2ABDw'),
            lendingPoolTknAccount: new web3_js_1.PublicKey('AVf6kFtrUGBAcLsPRpueYYEZsegui9ehXHxEUMwMedh1'),
            lendingPoolShareMint: new web3_js_1.PublicKey('vRQDjZjLQuaLnG1P25hbpzmrwUQjZauMcQLUqP76Lt6'),
            lendingPoolShareAccount: new web3_js_1.PublicKey('57CdJiuxs5xp35Qsj7UuA5bKFadsxPYndYiH4gHhyFVB'),
            lendingPoolCreditMint: new web3_js_1.PublicKey('9QbcYBsBn9jNM4vSzQJRi6GrTyQvCy3swspkmL6R2KpC'),
            lendingPoolCreditAccount: new web3_js_1.PublicKey('4YuEGUY6SMPZk11g6yztdULQCAwsLDsL4pAqfjvv2zpW'),
            lendingPoolFeeAccount: new web3_js_1.PublicKey('FHRMrLYLgE2zdseBBN5beXqsgRsB1ysDKZ3FrmZfMARk'),
        }
    },
    "USDC": {
        smartPoolInfo: {
            liquidityTokenMint: tokens_1.TOKENS.USDC.mintAddress,
            rewardsTokenMint: tokens_1.TOKENS.FRC.mintAddress,
            poolInfoAccount: new web3_js_1.PublicKey('JAHEucvkjhwqHDVjEcEfYjDncoBJuwr9jGmTibKhB9RH'),
            poolAuthority: new web3_js_1.PublicKey('2RrxxHCx7LDLSfFSP625pW3Bw45XRLRCpqNYSnyjib6s'),
            vaultManagerAccount: new web3_js_1.PublicKey('4bV7Su79Wk5VLX4gkYoZocX8GTwq1uYTcgcUfcZbnYAJ'),
            liquidityTokenAccount: new web3_js_1.PublicKey('5ehdF5mwy84GAficX1TWTDxhVMeWLwgnBbX7XhfnvWHb'),
            lendingShareAccount: new web3_js_1.PublicKey('FNfCXCUNzXB9FRQ4q93sruUUFnHrJnjGFWtRErYygBmR'),
            rewardsTokenAccount: new web3_js_1.PublicKey('6mWPAnr4pQGyVLA2XwAGA5VXkMoTaSyRe68s8iNHMCen'),
        },
        basedLendingPool: {
            tknMint: tokens_1.TOKENS.USDC.mintAddress,
            lendingPoolInfoAccount: new web3_js_1.PublicKey("Hx6LbkMHe69DYawhPyVNs8Apa6tyfogfzQV6a7XkwBUU"),
            lendingPoolTknAccount: new web3_js_1.PublicKey("CFp9kt8z3Epb1QSiEp3xA44KbSwuJxhFR3wQoerFqYS9"),
            lendingPoolFeeAccount: new web3_js_1.PublicKey("6h6vVc42moDmRs7KahQpFs8YWL5GfWBrxT9UrsUvpApn"),
            lendingPoolShareMint: new web3_js_1.PublicKey("62fDf5daUJ9jBz8Xtj6Bmw1bh1DvHn8AG4L9hMmxCzpu"),
            lendingPoolShareAccount: new web3_js_1.PublicKey("Gyc1V1xbA9NjzuURE662ATw6W4AdhwvsL26yUnSGhbkz"),
            lendingPoolCreditMint: new web3_js_1.PublicKey("CZ2s85dnuAVyGbRWBNauHZwj9oTV2i9xTvFwc1Cedr7E"),
            lendingPoolCreditAccount: new web3_js_1.PublicKey("Bd71jEvypKFs8N5gByyhD2149tSqLy4PVAzFFShCTJTB"),
        }
    },
    "SOL": {
        smartPoolInfo: {
            liquidityTokenMint: tokens_1.TOKENS.SOL.mintAddress,
            rewardsTokenMint: tokens_1.TOKENS.FRC.mintAddress,
            poolInfoAccount: new web3_js_1.PublicKey('9QEasBh3eMVmsCa3Z4istZHE62ctcjhBGC69Fbube7Mu'),
            poolAuthority: new web3_js_1.PublicKey('EzhhhvC37SbMCiAfiBrRwNc13duNjMzd3LyDNEH1riZm'),
            vaultManagerAccount: new web3_js_1.PublicKey('AbNddut5GLNADoLzQgPoiRMo7nTRZ5854t75P3z1pPhS'),
            liquidityTokenAccount: new web3_js_1.PublicKey('84Kd8CYgs8kYe6ogHpQoJrxthKNeoGrjS3spVfnbzFv'),
            lendingShareAccount: new web3_js_1.PublicKey('3cUuQwkpfMR8cD4xCmBjCfru5kJXd9S9cEH9rU3zRBwu'),
            rewardsTokenAccount: new web3_js_1.PublicKey('2U7zayUHhRQodM98vyxJAtGmjK5hCYyphA9s8AZVZMht'),
        },
        basedLendingPool: {
            tknMint: tokens_1.TOKENS.SOL.mintAddress,
            lendingPoolInfoAccount: new web3_js_1.PublicKey("EnZC8MB6QLTxwN1LNqCXYC7XMpvXqitnYQPf5y5AcQRn"),
            lendingMarketAuthority: new web3_js_1.PublicKey("sCDiYj7X7JmXg5fVq2nqED2q1Wqjo7PnqMgH3casMem"),
            lendingPoolTknAccount: new web3_js_1.PublicKey("GiJwrxQW5NQWWAva4rnDMXwHxfrhyUzwMX5to1pmCCgF"),
            lendingPoolFeeAccount: new web3_js_1.PublicKey("wLq5xWzkgfH1T2bbByBoUFxKk5g4K2urQv8q71gCsbM"),
            lendingPoolShareMint: new web3_js_1.PublicKey("92Zst3rEoJsyZUW2yNc2811GtkBwFrm3tqgxMTtFWuSB"),
            lendingPoolShareAccount: new web3_js_1.PublicKey("Hyi1jMgc9MMEZyaG9ziL2PvT7zbeRNKhb5K1kj9bxz4x"),
            lendingPoolCreditMint: new web3_js_1.PublicKey("7x61NrCzeekHwgW5Nk4afMg2sxjMK6RyMkx8tCzzX4qG"),
            lendingPoolCreditAccount: new web3_js_1.PublicKey("WuPCgCuyLSXatdjyWqkGjCPSQFgKLwcfsNR9fDq3hVL"),
        }
    },
    "RAY": {
        smartPoolInfo: {
            liquidityTokenMint: tokens_1.TOKENS.RAY.mintAddress,
            rewardsTokenMint: tokens_1.TOKENS.FRC.mintAddress,
            poolInfoAccount: new web3_js_1.PublicKey('9GthXJzMQeQkzAzeMwtWfP4bXy6nCnRJRHkMP9Jiq5E1'),
            poolAuthority: new web3_js_1.PublicKey('BBm3NNhr2md8H3aDqXhQ29gSu2p6cQLPuc8y8DHJ1Zcp'),
            vaultManagerAccount: new web3_js_1.PublicKey('48iPmFXCgN29NaKmkDWKWMBXTxoiAzRDM4hZK5mrQxYy'),
            liquidityTokenAccount: new web3_js_1.PublicKey('GqQjrUmAVeKDAQZdLpRYQdwo4PqidTXxWPn2j7Ku9n8S'),
            lendingShareAccount: new web3_js_1.PublicKey('CHMaC2EVy12UpWjcQLgqxLiG6ADWGkMfhqr7cZzYX1Bw'),
            rewardsTokenAccount: new web3_js_1.PublicKey('81rvRP61YCoqMXb4dTRGBoiohU7WGbSVJBNDqnxu3rud'),
        },
        basedLendingPool: {
            tknMint: tokens_1.TOKENS.RAY.mintAddress,
            lendingPoolInfoAccount: new web3_js_1.PublicKey("ENZ7vdrvNGdAfVa8DoGpT2GrfoSFUtWKRoesf2dvqpuq"),
            lendingPoolTknAccount: new web3_js_1.PublicKey("BUK8XbZeEqVULYRxGp4Va2R1BRqrYzXDrCVnWt6B7TZf"),
            lendingPoolFeeAccount: new web3_js_1.PublicKey("CbZydUBTvy2xqKshUByMdQR7FQ39EB5n2dyZpFXYDnXm"),
            lendingPoolShareMint: new web3_js_1.PublicKey("5uZERkJVyhCABEdocEssPZyMXYa8GJpLhngi95yXr9jc"),
            lendingPoolShareAccount: new web3_js_1.PublicKey("GrVpvHCRkTHESGwyaj4gtNLn95NAdkBSnZbKusoFumup"),
            lendingPoolCreditMint: new web3_js_1.PublicKey("GZqhoJhN1bz5Frq9GCaUSJ2nimnizXzHwsH12Qgh4YmR"),
            lendingPoolCreditAccount: new web3_js_1.PublicKey("6EZzqDqfTfdoFF4KD1gmki3btEHfULK4SjNcuJMTEDUt"),
        }
    },
    "ORCA": {
        smartPoolInfo: {
            liquidityTokenMint: tokens_1.TOKENS.ORCA.mintAddress,
            rewardsTokenMint: tokens_1.TOKENS.FRC.mintAddress,
            poolInfoAccount: new web3_js_1.PublicKey('2PLedtrsdm4kX557f2MEfqY7o7XDbeySbxiN68cpx3BZ'),
            poolAuthority: new web3_js_1.PublicKey('8tpUnYDVESYtAgj46xbHEg2MWXxa2gm6pbvhqPWvQsVU'),
            vaultManagerAccount: new web3_js_1.PublicKey('DmrxXcrn5zNPamd5iA5Wow3uchgTpByf6wMR2VQm29pi'),
            liquidityTokenAccount: new web3_js_1.PublicKey('3r4sccxqtdfE7z45KZFKwqu2Rg34usG3nRietCAo6PiM'),
            lendingShareAccount: new web3_js_1.PublicKey('2KACdqgpytLtEfJbtiQ7T43kZkM8ZKHR52Wrb5Ju5Ufe'),
            rewardsTokenAccount: new web3_js_1.PublicKey('3idxuCTZa8SS6Krxd7W8AZkzFQb5Y4AbdrkpSx3243sd'),
        },
        basedLendingPool: {
            tknMint: tokens_1.TOKENS.ORCA.mintAddress,
            lendingPoolInfoAccount: new web3_js_1.PublicKey("9bo5JunUhp4XD7TDeeS9ARvWTYGBYtuQKUHGUb7RRvkf"),
            lendingPoolTknAccount: new web3_js_1.PublicKey("F3EhkuRuZaYNY2bLTosjv6V7QXiZ8db9nXTZeLnmHC7V"),
            lendingPoolFeeAccount: new web3_js_1.PublicKey("HBc25ZyXiGMERxUX2KupZjc3DAivA48wBSgZGxCQgtT6"),
            lendingPoolShareMint: new web3_js_1.PublicKey("DKoKBD7YheCGZjWp5CaHDPtheAcUveRSMvWkaivbNzWh"),
            lendingPoolShareAccount: new web3_js_1.PublicKey("AMUE9EgiEiTP1YVBicwTVGtumBypR5thjTgbsXZM2PPG"),
            lendingPoolCreditMint: new web3_js_1.PublicKey("CJNd1LZZxZr243dpNN9DtEavXn65kstrg251MPb8vmwW"),
            lendingPoolCreditAccount: new web3_js_1.PublicKey("74m1zyzV8x9d8kZLEaVaNWRsnx8pH47tfX8cquot11ef"),
        }
    },
    "whETH": {
        smartPoolInfo: {
            liquidityTokenMint: tokens_1.TOKENS.whETH.mintAddress,
            rewardsTokenMint: tokens_1.TOKENS.FRC.mintAddress,
            poolInfoAccount: new web3_js_1.PublicKey('7toPbrv4amdXVR1whezJ367EQQV2fUd7i4yeY7maqGhr'),
            poolAuthority: new web3_js_1.PublicKey('7VWMeTcq3eDg4ZUzbvnq38rZzt63a2oTsLUUf4dRcCDp'),
            vaultManagerAccount: new web3_js_1.PublicKey('BwnQkxMesgk6V5HNYdhhyiD7ZPR2QSpptUbPcXfGqeUU'),
            liquidityTokenAccount: new web3_js_1.PublicKey('6jQSnxPHq25EGMRK9VyvHnKYzSSfjfXujvMKk1HNpBbL'),
            lendingShareAccount: new web3_js_1.PublicKey('GG7XsAq8RJw7hdRPyYauydWccaZKCvHR93kbieWkuDsP'),
            rewardsTokenAccount: new web3_js_1.PublicKey('7PKGDvgKDago7uwQ7H1thNzsNak6BQgF76isDWmWfvPe'),
        },
        basedLendingPool: {
            tknMint: tokens_1.TOKENS.whETH.mintAddress,
            lendingPoolInfoAccount: new web3_js_1.PublicKey('EEokU6r9bBdTz1umHjGhkjgfikARsJzJBXhYxnTmN4Xk'),
            lendingPoolTknAccount: new web3_js_1.PublicKey('9Mq6KEyW1nF24TcGrdLLfnK2pE5VdMxbY2SZtcP84R5X'),
            lendingPoolFeeAccount: new web3_js_1.PublicKey('EeNLz3tqBogKGs7pbtE556rfQEaV1AtYwKrjo4hHgd5V'),
            lendingPoolShareMint: new web3_js_1.PublicKey('Hk8b6i2C7PxFBPdd1TCrGssuZpCDEg4fh8yFuyHNJuJe'),
            lendingPoolShareAccount: new web3_js_1.PublicKey('FNG4A4GWF1EsQVEjNrQMRT9Dr3H28TycQabtnsQF1MiD'),
            lendingPoolCreditMint: new web3_js_1.PublicKey('7f7mU3tQvxD9t3f5ckhxcGSdwqoHHdWrJg43HXiUFWri'),
            lendingPoolCreditAccount: new web3_js_1.PublicKey('7LvrwstXyshwVn7BPDgjR8Cndersy9GzbbGhyHwQcX3W'),
        }
    },
};
const activeSmartLendingPools = [];
exports.activeSmartLendingPools = activeSmartLendingPools;
(0, lodash_1.forEach)(exports.smartLendingPools, (value, key) => {
    activeSmartLendingPools.push({
        pool: key,
        poolInfoAccount: value.smartPoolInfo.poolInfoAccount
    });
});
