// @port.finance/port-sdk/dist/esm/models/staking/StakeAccount.js
import { StakeAccountId } from "./StakeAccountId";
import { ExchangeRate } from "../ExchangeRate";
import { Lamport } from "../basic";
import { StakeAccountLayout } from "../../structs";
export class StakeAccount {
    constructor(stakeAccountId, stakingPoolId, startRate, depositedAmount, unclaimedReward) {
        this.stakeAccountId = stakeAccountId;
        this.stakingPoolId = stakingPoolId;
        this.startRate = startRate;
        this.depositedAmount = depositedAmount;
        this.unclaimedReward = unclaimedReward;
    }
    static newAccount(stakeAccountId, stakingPoolId) {
        return new StakeAccount(stakeAccountId, stakingPoolId, ExchangeRate.zero(), Lamport.zero(), Lamport.zero());
    }
    static fromRaw(raw) {
        const buffer = Buffer.from(raw.account.data);
        const proto = StakeAccountLayout.decode(buffer);
        const stakeAccountId = StakeAccountId.of(raw.pubkey);
        return new StakeAccount(stakeAccountId, proto.poolPubkey, proto.startRate, proto.depositedAmount, proto.unclaimedRewardWads);
    }
    getId() {
        return this.getStakeAccountId();
    }
    getStakeAccountId() {
        return this.stakeAccountId;
    }
    getStakingPoolId() {
        return this.stakingPoolId;
    }
    getStartRate() {
        return this.startRate;
    }
    getDepositAmount() {
        return this.depositedAmount;
    }
    getUnclaimedReward() {
        return this.unclaimedReward;
    }
    getTotalClaimableReward(rate) {
        const rateDiff = rate.subtract(this.getStartRate());
        if (rateDiff.isNegative()) {
            throw new Error("Rate lower than start rate");
        }
        const extraClaimable = this.getDepositAmount().multiply(rateDiff.getRaw());
        return extraClaimable.add(this.getUnclaimedReward());
    }
}
