// @apricot-lend/solana-swaps-js/dist/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Dex = exports.IsMultiSwapper = exports.MultiMarket = exports.Market = exports.SwapperType = exports.TokenID = void 0;
var TokenID;
(function (TokenID) {
    TokenID["APT"] = "APT";
    TokenID["BTC"] = "BTC";
    TokenID["ETH"] = "ETH";
    TokenID["SOL"] = "SOL";
    TokenID["mSOL"] = "mSOL";
    TokenID["RAY"] = "RAY";
    TokenID["SRM"] = "SRM";
    TokenID["USDT"] = "USDT";
    TokenID["USDC"] = "USDC";
    TokenID["UST"] = "UST";
    TokenID["PAI"] = "PAI";
    TokenID["SBR"] = "SBR";
    TokenID["ORCA"] = "ORCA";
    TokenID["USTv2"] = "USTv2";
    TokenID["MNDE"] = "MNDE";
    TokenID["FTT"] = "FTT";
    TokenID["stSOL"] = "stSOL";
    TokenID["whETH"] = "whETH";
    TokenID["scnSOL"] = "scnSOL";
})(TokenID = exports.TokenID || (exports.TokenID = {}));
var SwapperType;
(function (SwapperType) {
    SwapperType["Single"] = "Single";
    SwapperType["Multi"] = "Multi";
})(SwapperType = exports.SwapperType || (exports.SwapperType = {}));
// market contains meta data
var Market = /** @class */ (function () {
    function Market(name, tokenIds) {
        this.name = name;
        this.tokenIds = tokenIds;
    }
    return Market;
}());
exports.Market = Market;
var MultiMarket = /** @class */ (function () {
    function MultiMarket(markets) {
        this.markets = markets;
    }
    return MultiMarket;
}());
exports.MultiMarket = MultiMarket;
function IsMultiSwapper(object) {
    return 'createMultiSwapInstructions' in object;
}
exports.IsMultiSwapper = IsMultiSwapper;
var Dex = /** @class */ (function () {
    function Dex(markets) {
        this.markets = {};
        for (var _i = 0, markets_1 = markets; _i < markets_1.length; _i++) {
            var market = markets_1[_i];
            this.markets[market.name] = market;
        }
    }
    return Dex;
}());
exports.Dex = Dex;
