// @apricot-lend/solana-swaps-js/dist/saber/saberConstants.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SABER_USDT_USDC_MARKET = exports.SABER_USTv2_USDC_MARKET = exports.SaberMarket = exports.SABER_SWAP_PROGRAM = void 0;
var web3_js_1 = require("@solana/web3.js");
var types_1 = require("../types");
var Parser_1 = require("../utils/Parser");
var spl_token_1 = require("@solana/spl-token");
var tiny_invariant_1 = __importDefault(require("tiny-invariant"));
exports.SABER_SWAP_PROGRAM = new web3_js_1.PublicKey("SSwpkEEcbUqx4vtoEByFjSkhKdCT862DNVb52nZg1UZ");
var SaberMarket = /** @class */ (function (_super) {
    __extends(SaberMarket, _super);
    function SaberMarket(name, tokenIdA, tokenIdB, swap, swapAuthority, swapVaultA, swapVaultB, adminFeeDestinationA, adminFeeDestinationB) {
        var _this = _super.call(this, name, [tokenIdA, tokenIdB]) || this;
        _this.name = name;
        _this.tokenIdA = tokenIdA;
        _this.tokenIdB = tokenIdB;
        _this.swap = swap;
        _this.swapAuthority = swapAuthority;
        _this.swapVaultA = swapVaultA;
        _this.swapVaultB = swapVaultB;
        _this.adminFeeDestinationA = adminFeeDestinationA;
        _this.adminFeeDestinationB = adminFeeDestinationB;
        _this.INST_LAYOUT = new Parser_1.Parser()
            .u8("cmd")
            .u64("in_amount")
            .u64("min_out_amount");
        return _this;
    }
    SaberMarket.prototype.getSwapper = function () {
        return this;
    };
    SaberMarket.prototype.createSwapInstructions = function (fromToken, fromAmount, fromTokenAccount, toToken, minToAmount, toTokenAccount, tradeOwner) {
        return __awaiter(this, void 0, void 0, function () {
            var buffer, swapSource, swapDestination, adminFeeDestination, ix;
            return __generator(this, function (_a) {
                (0, tiny_invariant_1.default)([this.tokenIdA, this.tokenIdB].includes(fromToken));
                (0, tiny_invariant_1.default)([this.tokenIdA, this.tokenIdB].includes(toToken));
                buffer = this.INST_LAYOUT.encode({
                    cmd: 1,
                    in_amount: fromAmount,
                    min_out_amount: minToAmount
                });
                swapSource = fromToken === this.tokenIdA ? this.swapVaultA : this.swapVaultB;
                swapDestination = toToken === this.tokenIdB ? this.swapVaultB : this.swapVaultA;
                adminFeeDestination = toToken === this.tokenIdB ? this.adminFeeDestinationB : this.adminFeeDestinationA;
                ix = new web3_js_1.TransactionInstruction({
                    programId: exports.SABER_SWAP_PROGRAM,
                    keys: [
                        { pubkey: this.swap, isSigner: false, isWritable: false },
                        { pubkey: this.swapAuthority, isSigner: false, isWritable: false },
                        { pubkey: tradeOwner, isSigner: true, isWritable: false },
                        { pubkey: fromTokenAccount, isSigner: false, isWritable: true },
                        { pubkey: swapSource, isSigner: false, isWritable: true },
                        { pubkey: swapDestination, isSigner: false, isWritable: true },
                        { pubkey: toTokenAccount, isSigner: false, isWritable: true },
                        { pubkey: adminFeeDestination, isSigner: false, isWritable: true },
                        { pubkey: spl_token_1.TOKEN_PROGRAM_ID, isSigner: false, isWritable: false },
                        { pubkey: web3_js_1.SYSVAR_CLOCK_PUBKEY, isSigner: false, isWritable: false },
                    ],
                    data: buffer,
                });
                return [2 /*return*/, [ix]];
            });
        });
    };
    return SaberMarket;
}(types_1.Market));
exports.SaberMarket = SaberMarket;
exports.SABER_USTv2_USDC_MARKET = new SaberMarket("USTv2/USDC", types_1.TokenID.USTv2, types_1.TokenID.USDC, new web3_js_1.PublicKey("KwnjUuZhTMTSGAaavkLEmSyfobY16JNH4poL9oeeEvE"), // swap
new web3_js_1.PublicKey("9osV5a7FXEjuMujxZJGBRXVAyQ5fJfBFNkyAf6fSz9kw"), // swapAuthority
new web3_js_1.PublicKey("J63v6qEZmQpDqCD8bd4PXu2Pq5ZbyXrFcSa3Xt1HdAPQ"), // USTv2 valut
new web3_js_1.PublicKey("BnKQtTdLw9qPCDgZkWX3sURkBAoKCUYL1yahh6Mw7mRK"), // USDC vault
new web3_js_1.PublicKey("BYgyVxdrGa3XNj1cx1XHAVyRG8qYhBnv1DS59Bsvmg5h"), // USTv2 admin fee dest
new web3_js_1.PublicKey("G9nt2GazsDj3Ey3KdA49Sfaq9K95Dc72Ejps4NKTP2SR"));
exports.SABER_USDT_USDC_MARKET = new SaberMarket("USDT/USDC", types_1.TokenID.USDT, types_1.TokenID.USDC, new web3_js_1.PublicKey("YAkoNb6HKmSxQN9L8hiBE5tPJRsniSSMzND1boHmZxe"), // swap
new web3_js_1.PublicKey("5C1k9yV7y4CjMnKv8eGYDgWND8P89Pdfj79Trk2qmfGo"), // swapAuthority
new web3_js_1.PublicKey("EnTrdMMpdhugeH6Ban6gYZWXughWxKtVGfCwFn78ZmY3"), // USDT valut
new web3_js_1.PublicKey("CfWX7o2TswwbxusJ4hCaPobu2jLCb1hfXuXJQjVq3jQF"), // USDC vault
new web3_js_1.PublicKey("63aJYYuZddSnCGyE8FNrCVQWnXhjh6CQSRwcDeSMhdVC"), // USDT admin fee dest
new web3_js_1.PublicKey("XZuQG7CQrAA6y6tHM9CLrDjDUWwuUU2SBoV7pLaGDQT"));
