// @apricot-lend/solana-swaps-js/dist/mercurial/mercurialConstants.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MERCURIAL_USTv1_USDC_MARKET = exports.MercurialMarket = exports.MERCURIAL_SWAP_PROGRAM = void 0;
var web3_js_1 = require("@solana/web3.js");
var types_1 = require("../types");
var Parser_1 = require("../utils/Parser");
var spl_token_1 = require("@solana/spl-token");
exports.MERCURIAL_SWAP_PROGRAM = new web3_js_1.PublicKey("MERLuDFBMmsHnsBPZw2sDQZHvXFMwp8EdjudcU2HKky");
var MercurialMarket = /** @class */ (function (_super) {
    __extends(MercurialMarket, _super);
    function MercurialMarket(name, tokenIdA, tokenIdB, swap, swapAuthority, vaultA, vaultB, vaultC) {
        var _this = _super.call(this, name, [tokenIdA, tokenIdB]) || this;
        _this.name = name;
        _this.tokenIdA = tokenIdA;
        _this.tokenIdB = tokenIdB;
        _this.swap = swap;
        _this.swapAuthority = swapAuthority;
        _this.vaultA = vaultA;
        _this.vaultB = vaultB;
        _this.vaultC = vaultC;
        _this.INST_LAYOUT = new Parser_1.Parser()
            .u8("cmd")
            .u64("in_amount")
            .u64("min_out_amount");
        return _this;
    }
    MercurialMarket.prototype.getSwapper = function () {
        return this;
    };
    MercurialMarket.prototype.createSwapInstructions = function (_fromToken, fromAmount, fromTokenAccount, _toToken, minToAmount, toTokenAccount, tradeOwner) {
        return __awaiter(this, void 0, void 0, function () {
            var buffer, ix;
            return __generator(this, function (_a) {
                buffer = this.INST_LAYOUT.encode({
                    cmd: 4,
                    in_amount: fromAmount,
                    min_out_amount: minToAmount
                });
                ix = new web3_js_1.TransactionInstruction({
                    programId: exports.MERCURIAL_SWAP_PROGRAM,
                    keys: [
                        { pubkey: this.swap, isSigner: false, isWritable: false },
                        { pubkey: spl_token_1.TOKEN_PROGRAM_ID, isSigner: false, isWritable: false },
                        { pubkey: this.swapAuthority, isSigner: false, isWritable: false },
                        { pubkey: tradeOwner, isSigner: true, isWritable: false },
                        { pubkey: this.vaultA, isSigner: false, isWritable: true },
                        { pubkey: this.vaultB, isSigner: false, isWritable: true },
                        { pubkey: this.vaultC, isSigner: false, isWritable: true },
                        // below 2 accounts on solscan are displayed incorrectly but on solana exporer
                        { pubkey: fromTokenAccount, isSigner: false, isWritable: true },
                        { pubkey: toTokenAccount, isSigner: false, isWritable: true },
                    ],
                    data: buffer,
                });
                return [2 /*return*/, [ix]];
            });
        });
    };
    return MercurialMarket;
}(types_1.Market));
exports.MercurialMarket = MercurialMarket;
exports.MERCURIAL_USTv1_USDC_MARKET = new MercurialMarket("UST/USDC", types_1.TokenID.UST, types_1.TokenID.USDC, new web3_js_1.PublicKey("USD6kaowtDjwRkN5gAjw1PDMQvc9xRp8xW9GK8Z5HBA"), // swap Mercurial UST 3-Pool (USDC-USDT-UST)
new web3_js_1.PublicKey("FDonWCo5RJhx8rzSwtToUXiLEL7dAqLmUhnyH76F888D"), // swapAuthority
new web3_js_1.PublicKey("Go93HBNLMK6QnSHyZTViAoti9GCmSRhkeZMiVoJmcrud"), // vaultA
new web3_js_1.PublicKey("2uZVPMdS8VFwMEZjZNzP1f78m1zaqvT3R2JwMs1w9fU3"), // vaultB
new web3_js_1.PublicKey("5nazSj6MfaEeuAoQezMG3raWuZiyFjE7fmURUgDeP8cF"));
