// francium-sdk/dist/constants/lend/pools.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.lendingPoolList = exports.lendingPools = exports.LendingPoolLayout = exports.lendProgramId = void 0;
const spl_token_1 = require("@solana/spl-token");
const web3_js_1 = require("@solana/web3.js");
const BufferLayout = require("buffer-layout");
const borsh_1 = require("@project-serum/borsh");
const tokens_1 = require("../tokens");
exports.lendProgramId = new web3_js_1.PublicKey("FC81tbGt6JWRXidaWYFXxGnTk4VgobhJHATvTRVMqgWj");
exports.LendingPoolLayout = BufferLayout.struct([
    BufferLayout.u8("version"),
    BufferLayout.nu64("last_update_slot"),
    BufferLayout.u8('last_update_stale'),
    (0, borsh_1.publicKey)("lending_market"),
    (0, borsh_1.publicKey)("liquidity_mint_pubkey"),
    BufferLayout.u8('liquidity_mint_decimals'),
    (0, borsh_1.publicKey)('liquidity_supply_pubkey'),
    (0, borsh_1.publicKey)('liquidity_fee_receiver'),
    BufferLayout.blob(36, "oracle"),
    BufferLayout.nu64('liquidity_available_amount'),
    BufferLayout.blob(16, "liquidity_borrowed_amount_wads"),
    BufferLayout.blob(16, "liquidity_cumulative_borrow_rate_wads"),
    BufferLayout.blob(8, "liquidity_market_price"),
    BufferLayout.blob(32, "share_mint_pubkey"),
    BufferLayout.blob(8, "share_mint_total_supply"),
    BufferLayout.blob(32, "share_supply_pubkey"),
    BufferLayout.blob(32, "credit_mint_pubkey"),
    BufferLayout.blob(8, "credit_mint_total_supply"),
    BufferLayout.blob(32, "credit_supply_pubkey"),
    BufferLayout.u8("threshold_1"),
    BufferLayout.u8("threshold_2"),
    BufferLayout.u8("base_1"),
    BufferLayout.u16("factor_1"),
    BufferLayout.u8("base_2"),
    BufferLayout.u16("factor_2"),
    BufferLayout.u8("base_3"),
    BufferLayout.u16("factor_3"),
    BufferLayout.u8("interest_reverse_rate"),
    BufferLayout.nu64('accumulated_interest_reverse'),
    BufferLayout.blob(108, "padding"),
]);
exports.lendingPools = {
    USDC: {
        programId: exports.lendProgramId,
        tokenMint: new web3_js_1.PublicKey("EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v"),
        marketInfoAccount: new web3_js_1.PublicKey("4XNif294wbrxj6tJ8K5Rg7SuaEACnu9s2L27i28MQB6E"),
        marketOwner: new web3_js_1.PublicKey("7MBLg6oV5phip11YBbJPuq7u38kdzSi9PM3BifKSpLaR"),
        lendingPoolInfoAccount: new web3_js_1.PublicKey("Hx6LbkMHe69DYawhPyVNs8Apa6tyfogfzQV6a7XkwBUU"),
        lendingMarketAuthority: new web3_js_1.PublicKey("sCDiYj7X7JmXg5fVq2nqED2q1Wqjo7PnqMgH3casMem"),
        lendingPoolTknAccount: new web3_js_1.PublicKey("CFp9kt8z3Epb1QSiEp3xA44KbSwuJxhFR3wQoerFqYS9"),
        lendingPoolFeeAccount: new web3_js_1.PublicKey("CugsLyJk1Jghc7LkgvnFuUwshJhz1FE9mpkF7Z4acAMU"),
        lendingPoolShareMint: new web3_js_1.PublicKey("62fDf5daUJ9jBz8Xtj6Bmw1bh1DvHn8AG4L9hMmxCzpu"),
        lendingPoolShareAccount: new web3_js_1.PublicKey("Gyc1V1xbA9NjzuURE662ATw6W4AdhwvsL26yUnSGhbkz"),
        lendingPoolCreditMint: new web3_js_1.PublicKey("CZ2s85dnuAVyGbRWBNauHZwj9oTV2i9xTvFwc1Cedr7E"),
        lendingPoolCreditAccount: new web3_js_1.PublicKey("Bd71jEvypKFs8N5gByyhD2149tSqLy4PVAzFFShCTJTB")
    },
    PAI: {
        programId: exports.lendProgramId,
        tokenMint: new web3_js_1.PublicKey("Ea5SjE2Y6yvCeW5dYTn7PYMuW5ikXkvbGdcmSnXeaLjS"),
        marketInfoAccount: new web3_js_1.PublicKey("4XNif294wbrxj6tJ8K5Rg7SuaEACnu9s2L27i28MQB6E"),
        marketOwner: new web3_js_1.PublicKey("7MBLg6oV5phip11YBbJPuq7u38kdzSi9PM3BifKSpLaR"),
        lendingPoolInfoAccount: new web3_js_1.PublicKey("EzofB5BK23PHDfEAThk5oJANb9FWTXzezrafiTDzcBBA"),
        lendingMarketAuthority: new web3_js_1.PublicKey("sCDiYj7X7JmXg5fVq2nqED2q1Wqjo7PnqMgH3casMem"),
        lendingPoolTknAccount: new web3_js_1.PublicKey("4Kvd1ULAy45k2EQt1pdePa82UBz9tW6N1rLJWb3EAJgQ"),
        lendingPoolFeeAccount: new web3_js_1.PublicKey("sKmPL3zkAf7777bhXnMaSfTU1jg1keNHNPz38wBTNd5"),
        lendingPoolShareMint: new web3_js_1.PublicKey("HDvD8a4VWbkHNG7hb4CBumNhn41DyKL51qVYBNH73o23"),
        lendingPoolShareAccount: new web3_js_1.PublicKey("9tk9XwzJt1J2DoNcBtBmGMsnbnwvMUAoJF1pjdcfJe2E"),
        lendingPoolCreditMint: new web3_js_1.PublicKey("HkMze549cFMdLMRDsJyfH8n6EE5jjTuVZr9re4WRYuk9"),
        lendingPoolCreditAccount: new web3_js_1.PublicKey("7h8YziEXSZgyzqDb5c6PP3uUztQR7hQHwqyWWe9BTqY3")
    },
    RAY: {
        programId: exports.lendProgramId,
        tokenMint: new web3_js_1.PublicKey("4k3Dyjzvzp8eMZWUXbBCjEvwSkkk59S5iCNLY3QrkX6R"),
        marketInfoAccount: new web3_js_1.PublicKey("4XNif294wbrxj6tJ8K5Rg7SuaEACnu9s2L27i28MQB6E"),
        marketOwner: new web3_js_1.PublicKey("7MBLg6oV5phip11YBbJPuq7u38kdzSi9PM3BifKSpLaR"),
        lendingPoolInfoAccount: new web3_js_1.PublicKey("ENZ7vdrvNGdAfVa8DoGpT2GrfoSFUtWKRoesf2dvqpuq"),
        lendingMarketAuthority: new web3_js_1.PublicKey("sCDiYj7X7JmXg5fVq2nqED2q1Wqjo7PnqMgH3casMem"),
        lendingPoolTknAccount: new web3_js_1.PublicKey("BUK8XbZeEqVULYRxGp4Va2R1BRqrYzXDrCVnWt6B7TZf"),
        lendingPoolFeeAccount: new web3_js_1.PublicKey("C2HJD7r6NYCCVoXPPExxYEQ73LZJ3qmb5H13GnwMZfwa"),
        lendingPoolShareMint: new web3_js_1.PublicKey("5uZERkJVyhCABEdocEssPZyMXYa8GJpLhngi95yXr9jc"),
        lendingPoolShareAccount: new web3_js_1.PublicKey("GrVpvHCRkTHESGwyaj4gtNLn95NAdkBSnZbKusoFumup"),
        lendingPoolCreditMint: new web3_js_1.PublicKey("GZqhoJhN1bz5Frq9GCaUSJ2nimnizXzHwsH12Qgh4YmR"),
        lendingPoolCreditAccount: new web3_js_1.PublicKey("6EZzqDqfTfdoFF4KD1gmki3btEHfULK4SjNcuJMTEDUt")
    },
    USDT: {
        programId: exports.lendProgramId,
        tokenMint: new web3_js_1.PublicKey("Es9vMFrzaCERmJfrF4H2FYD4KCoNkY11McCe8BenwNYB"),
        marketInfoAccount: new web3_js_1.PublicKey("4XNif294wbrxj6tJ8K5Rg7SuaEACnu9s2L27i28MQB6E"),
        marketOwner: new web3_js_1.PublicKey("7MBLg6oV5phip11YBbJPuq7u38kdzSi9PM3BifKSpLaR"),
        lendingPoolInfoAccount: new web3_js_1.PublicKey("EyqWUeme2B1dhuwJvJVi2coYPcGWmWrakcpsSFeikrhb"),
        lendingMarketAuthority: new web3_js_1.PublicKey("sCDiYj7X7JmXg5fVq2nqED2q1Wqjo7PnqMgH3casMem"),
        lendingPoolTknAccount: new web3_js_1.PublicKey("Acx3Ky9pk8CyYftA9HYF31EbVhBDi21EqJsFtMW2XxXW"),
        lendingPoolFeeAccount: new web3_js_1.PublicKey("7gAN5Y4hGAG2A9vyr9dURGosjmo4Rwiaqz4iKZGNenE1"),
        lendingPoolShareMint: new web3_js_1.PublicKey("8TtsTZQ6mU4YSW5jw4sDjSDEG1CW8QLpK7C9g5TVgBvn"),
        lendingPoolShareAccount: new web3_js_1.PublicKey("EfLgyXRGFi9MUCMgrsFBuAJ7pMPXBceKhD6x4PgdvLLc"),
        lendingPoolCreditMint: new web3_js_1.PublicKey("CoheXVD8cAdVaKznfay22dBpfML5Fbz5g4j67syKgRg4"),
        lendingPoolCreditAccount: new web3_js_1.PublicKey("EAQxDXsA6RXxWtA1PwZNnTnj2MapHBtMzvUmfuGPJCPX")
    },
    SRM: {
        programId: exports.lendProgramId,
        tokenMint: new web3_js_1.PublicKey("SRMuApVNdxXokk5GT7XD5cUUgXMBCoAz2LHeuAoKWRt"),
        marketInfoAccount: new web3_js_1.PublicKey("4XNif294wbrxj6tJ8K5Rg7SuaEACnu9s2L27i28MQB6E"),
        marketOwner: new web3_js_1.PublicKey("7MBLg6oV5phip11YBbJPuq7u38kdzSi9PM3BifKSpLaR"),
        lendingPoolInfoAccount: new web3_js_1.PublicKey("B1zB1EuTjnFPLdwySeBYhzeAf3h9buWLbDoG7AHcUTMF"),
        lendingMarketAuthority: new web3_js_1.PublicKey("sCDiYj7X7JmXg5fVq2nqED2q1Wqjo7PnqMgH3casMem"),
        lendingPoolTknAccount: new web3_js_1.PublicKey("FLQtBThLEVvhXdKqq2CREL8sFt8jAFS8szm4HaMYqmJk"),
        lendingPoolFeeAccount: new web3_js_1.PublicKey("FdSBYwMEfy21H3k64cvYdir3mxzmnZotiYUsC5iPPoxM"),
        lendingPoolShareMint: new web3_js_1.PublicKey("6CZhCFwA7hbqDEoVUQdFpHaCGQU7fSSwhaJWdG4DmsQ6"),
        lendingPoolShareAccount: new web3_js_1.PublicKey("C3EjtH3hVuLrU3j1y5ArMFRbAhxFf5hXNfk3b9SU91qN"),
        lendingPoolCreditMint: new web3_js_1.PublicKey("DqKuxgMxLPDCXtoQzLG1p85S2i3tP3EaeUVDAw7P8Nt4"),
        lendingPoolCreditAccount: new web3_js_1.PublicKey("Ednqn9Zo5HpDX2DRmCWu6x46ZXnHNmTmrq9G8hrTuBnz")
    },
    ETH: {
        programId: exports.lendProgramId,
        tokenMint: new web3_js_1.PublicKey("2FPyTwcZLUg1MDrwsyoP4D6s1tM7hAkHYRjkNb5w6Pxk"),
        marketInfoAccount: new web3_js_1.PublicKey("4XNif294wbrxj6tJ8K5Rg7SuaEACnu9s2L27i28MQB6E"),
        marketOwner: new web3_js_1.PublicKey("7MBLg6oV5phip11YBbJPuq7u38kdzSi9PM3BifKSpLaR"),
        lendingPoolInfoAccount: new web3_js_1.PublicKey("CKMQxUz1nkn3NS5B9AUD1uyWNL8iN2piG9LVt1RvWXzj"),
        lendingMarketAuthority: new web3_js_1.PublicKey("sCDiYj7X7JmXg5fVq2nqED2q1Wqjo7PnqMgH3casMem"),
        lendingPoolTknAccount: new web3_js_1.PublicKey("9MH38iiDX4Pk37U6TXLqz2783RspNhwBTYwBNHp8WUzP"),
        lendingPoolFeeAccount: new web3_js_1.PublicKey("8UPe7Fcm2f1QEFQh2YNr1jg2vgQmj4CXhLYEWgStHd8B"),
        lendingPoolShareMint: new web3_js_1.PublicKey("B8QXcUv5FFwyHH5V88g6PhYBc8fQvwSpza4C9PsiRpQD"),
        lendingPoolShareAccount: new web3_js_1.PublicKey("C5X2Q2K2jQtwpuqHKnLVJ1ZsvL9BMRwddMgqaQ5UGNkC"),
        lendingPoolCreditMint: new web3_js_1.PublicKey("BKKbCDggM83SPLFd1jRPu1ZsCv9nXkwZVne11N9FEs9w"),
        lendingPoolCreditAccount: new web3_js_1.PublicKey("9A3KAmmv1VyqNqVGBM6T9b1dp9Ax9xxdeXEgedo8U7Gh")
    },
    SOL: {
        programId: exports.lendProgramId,
        tokenMint: spl_token_1.NATIVE_MINT,
        marketInfoAccount: new web3_js_1.PublicKey("4XNif294wbrxj6tJ8K5Rg7SuaEACnu9s2L27i28MQB6E"),
        marketOwner: new web3_js_1.PublicKey("7MBLg6oV5phip11YBbJPuq7u38kdzSi9PM3BifKSpLaR"),
        lendingPoolInfoAccount: new web3_js_1.PublicKey("EnZC8MB6QLTxwN1LNqCXYC7XMpvXqitnYQPf5y5AcQRn"),
        lendingMarketAuthority: new web3_js_1.PublicKey("sCDiYj7X7JmXg5fVq2nqED2q1Wqjo7PnqMgH3casMem"),
        lendingPoolTknAccount: new web3_js_1.PublicKey("GiJwrxQW5NQWWAva4rnDMXwHxfrhyUzwMX5to1pmCCgF"),
        lendingPoolFeeAccount: new web3_js_1.PublicKey("2BDZqoVKAX3PJKeZJytjk7oEpbUu3jvdgdi6tA8VusnY"),
        lendingPoolShareMint: new web3_js_1.PublicKey("92Zst3rEoJsyZUW2yNc2811GtkBwFrm3tqgxMTtFWuSB"),
        lendingPoolShareAccount: new web3_js_1.PublicKey("Hyi1jMgc9MMEZyaG9ziL2PvT7zbeRNKhb5K1kj9bxz4x"),
        lendingPoolCreditMint: new web3_js_1.PublicKey("7x61NrCzeekHwgW5Nk4afMg2sxjMK6RyMkx8tCzzX4qG"),
        lendingPoolCreditAccount: new web3_js_1.PublicKey("WuPCgCuyLSXatdjyWqkGjCPSQFgKLwcfsNR9fDq3hVL")
    },
    ORCA: {
        programId: exports.lendProgramId,
        tokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
        marketInfoAccount: new web3_js_1.PublicKey("4XNif294wbrxj6tJ8K5Rg7SuaEACnu9s2L27i28MQB6E"),
        marketOwner: new web3_js_1.PublicKey("7MBLg6oV5phip11YBbJPuq7u38kdzSi9PM3BifKSpLaR"),
        lendingPoolInfoAccount: new web3_js_1.PublicKey("9bo5JunUhp4XD7TDeeS9ARvWTYGBYtuQKUHGUb7RRvkf"),
        lendingMarketAuthority: new web3_js_1.PublicKey("sCDiYj7X7JmXg5fVq2nqED2q1Wqjo7PnqMgH3casMem"),
        lendingPoolTknAccount: new web3_js_1.PublicKey("F3EhkuRuZaYNY2bLTosjv6V7QXiZ8db9nXTZeLnmHC7V"),
        lendingPoolFeeAccount: new web3_js_1.PublicKey("F3EhkuRuZaYNY2bLTosjv6V7QXiZ8db9nXTZeLnmHC7V"),
        lendingPoolShareMint: new web3_js_1.PublicKey("DKoKBD7YheCGZjWp5CaHDPtheAcUveRSMvWkaivbNzWh"),
        lendingPoolShareAccount: new web3_js_1.PublicKey("AMUE9EgiEiTP1YVBicwTVGtumBypR5thjTgbsXZM2PPG"),
        lendingPoolCreditMint: new web3_js_1.PublicKey("CJNd1LZZxZr243dpNN9DtEavXn65kstrg251MPb8vmwW"),
        lendingPoolCreditAccount: new web3_js_1.PublicKey("74m1zyzV8x9d8kZLEaVaNWRsnx8pH47tfX8cquot11ef")
    },
    mSOL: {
        programId: exports.lendProgramId,
        tokenMint: new web3_js_1.PublicKey("mSoLzYCxHdYgdzU16g5QSh3i5K3z3KZK7ytfqcJm7So"),
        marketInfoAccount: new web3_js_1.PublicKey("4XNif294wbrxj6tJ8K5Rg7SuaEACnu9s2L27i28MQB6E"),
        marketOwner: new web3_js_1.PublicKey("7MBLg6oV5phip11YBbJPuq7u38kdzSi9PM3BifKSpLaR"),
        lendingMarketAuthority: new web3_js_1.PublicKey("sCDiYj7X7JmXg5fVq2nqED2q1Wqjo7PnqMgH3casMem"),
        lendingPoolInfoAccount: new web3_js_1.PublicKey('492Hr5QDQyrsLcyXqMZ7A5osjmqwtn4cdKNQLDBYBzvy'),
        lendingPoolTknAccount: new web3_js_1.PublicKey('ADW9ZJuRQ9xbzTtWCwX1Th24Vxq7GWHWwYtctrbyjSr1'),
        lendingPoolFeeAccount: new web3_js_1.PublicKey('ADW9ZJuRQ9xbzTtWCwX1Th24Vxq7GWHWwYtctrbyjSr1'),
        lendingPoolShareMint: new web3_js_1.PublicKey('4mntRxKJWib1YTWNV4aLfszF6FeziBHdwnNAwGSfKKMX'),
        lendingPoolShareAccount: new web3_js_1.PublicKey('EbWWhmDLKXDdq2guubVJCdSztuUaN6XhtwVMy9TMMwtV'),
        lendingPoolCreditMint: new web3_js_1.PublicKey('Fdu3QfaQ9VL4GF85NrkpzhKRAJECv9cx7gchHca5MsRt'),
        lendingPoolCreditAccount: new web3_js_1.PublicKey('HdTwUP1U3WQ1EnYPP8c8mx8sqHhXYQ9zwKVsLGxWi7W1'),
    },
    BTC: {
        programId: exports.lendProgramId,
        tokenMint: new web3_js_1.PublicKey("9n4nbM75f5Ui33ZbPYXn59EwSgE8CGsHtAeTH5YFeJ9E"),
        marketInfoAccount: new web3_js_1.PublicKey("4XNif294wbrxj6tJ8K5Rg7SuaEACnu9s2L27i28MQB6E"),
        marketOwner: new web3_js_1.PublicKey("7MBLg6oV5phip11YBbJPuq7u38kdzSi9PM3BifKSpLaR"),
        lendingMarketAuthority: new web3_js_1.PublicKey("sCDiYj7X7JmXg5fVq2nqED2q1Wqjo7PnqMgH3casMem"),
        lendingPoolInfoAccount: new web3_js_1.PublicKey('DxAuEAxKYXsKMYG5Ma4TM8AsrAo7LQBSnbEGicU7i1ek'),
        lendingPoolTknAccount: new web3_js_1.PublicKey('2MN34BxuLVrjZpKYKHmW1c6ZGeQs4aWQRrEvmrYfZdtG'),
        lendingPoolFeeAccount: new web3_js_1.PublicKey('2MN34BxuLVrjZpKYKHmW1c6ZGeQs4aWQRrEvmrYfZdtG'),
        lendingPoolShareMint: new web3_js_1.PublicKey('2G9iwy9zfLaXB2bFiqSA7YbKEvtAEXVmdvGTF28jQVgg'),
        lendingPoolShareAccount: new web3_js_1.PublicKey('HGejymLhai1TDvRA4vBEbS5VRbKP1u74ZdzZcv8Y2o1M'),
        lendingPoolCreditMint: new web3_js_1.PublicKey('6WzqPd81AqsCwbXp33A6ro87axfD9je32hn3JE4KsnGi'),
        lendingPoolCreditAccount: new web3_js_1.PublicKey('E7HkZYeY8mqrtC45kusWiArqLNpGnV5Foioj6yEPHB1y'),
    },
    whETH: {
        programId: exports.lendProgramId,
        tokenMint: new web3_js_1.PublicKey("7vfCXTUXx5WJV5JADk17DUJ4ksgau7utNKj4b963voxs"),
        marketInfoAccount: new web3_js_1.PublicKey("4XNif294wbrxj6tJ8K5Rg7SuaEACnu9s2L27i28MQB6E"),
        marketOwner: new web3_js_1.PublicKey("7MBLg6oV5phip11YBbJPuq7u38kdzSi9PM3BifKSpLaR"),
        lendingMarketAuthority: new web3_js_1.PublicKey("sCDiYj7X7JmXg5fVq2nqED2q1Wqjo7PnqMgH3casMem"),
        lendingPoolInfoAccount: new web3_js_1.PublicKey('EEokU6r9bBdTz1umHjGhkjgfikARsJzJBXhYxnTmN4Xk'),
        lendingPoolTknAccount: new web3_js_1.PublicKey('9Mq6KEyW1nF24TcGrdLLfnK2pE5VdMxbY2SZtcP84R5X'),
        lendingPoolFeeAccount: new web3_js_1.PublicKey('9Mq6KEyW1nF24TcGrdLLfnK2pE5VdMxbY2SZtcP84R5X'),
        lendingPoolShareMint: new web3_js_1.PublicKey('Hk8b6i2C7PxFBPdd1TCrGssuZpCDEg4fh8yFuyHNJuJe'),
        lendingPoolShareAccount: new web3_js_1.PublicKey('FNG4A4GWF1EsQVEjNrQMRT9Dr3H28TycQabtnsQF1MiD'),
        lendingPoolCreditMint: new web3_js_1.PublicKey('7f7mU3tQvxD9t3f5ckhxcGSdwqoHHdWrJg43HXiUFWri'),
        lendingPoolCreditAccount: new web3_js_1.PublicKey('7LvrwstXyshwVn7BPDgjR8Cndersy9GzbbGhyHwQcX3W'),
    },
    weSUSHI: {
        programId: exports.lendProgramId,
        tokenMint: tokens_1.TOKENS.weSUSHI.mintAddress,
        marketInfoAccount: new web3_js_1.PublicKey("4XNif294wbrxj6tJ8K5Rg7SuaEACnu9s2L27i28MQB6E"),
        marketOwner: new web3_js_1.PublicKey("7MBLg6oV5phip11YBbJPuq7u38kdzSi9PM3BifKSpLaR"),
        lendingMarketAuthority: new web3_js_1.PublicKey("sCDiYj7X7JmXg5fVq2nqED2q1Wqjo7PnqMgH3casMem"),
        lendingPoolInfoAccount: new web3_js_1.PublicKey('2oTEVdMpSRsamFeNYzqn3wifsPHRomA8TShgbTnDtTgN'),
        lendingPoolTknAccount: new web3_js_1.PublicKey('4VAyWCoRbW5YXiV5Rz7pZy7miYBJLjfZYsykDuTnWyqj'),
        lendingPoolFeeAccount: new web3_js_1.PublicKey('4VAyWCoRbW5YXiV5Rz7pZy7miYBJLjfZYsykDuTnWyqj'),
        lendingPoolShareMint: new web3_js_1.PublicKey('EEhiV55jAt5JDpeH3GF4VGrStiPn5gCeWmqffyTp9B4E'),
        lendingPoolShareAccount: new web3_js_1.PublicKey('5SDoskkCcV8NBJcP97g9CS7BEGLqXT5q2F6Ve2W3VN71'),
        lendingPoolCreditMint: new web3_js_1.PublicKey('E2ocWnUUuBpUDLRSxB5VdUqr635DMPiWuV4PYDtiyGbi'),
        lendingPoolCreditAccount: new web3_js_1.PublicKey('L9idBQvg4SKzdtHPgZijpJGkDT4WVbhnA7KmTVPT477'),
    },
    weUNI: {
        programId: exports.lendProgramId,
        tokenMint: tokens_1.TOKENS.weUNI.mintAddress,
        marketInfoAccount: new web3_js_1.PublicKey("4XNif294wbrxj6tJ8K5Rg7SuaEACnu9s2L27i28MQB6E"),
        marketOwner: new web3_js_1.PublicKey("7MBLg6oV5phip11YBbJPuq7u38kdzSi9PM3BifKSpLaR"),
        lendingMarketAuthority: new web3_js_1.PublicKey("sCDiYj7X7JmXg5fVq2nqED2q1Wqjo7PnqMgH3casMem"),
        lendingPoolInfoAccount: new web3_js_1.PublicKey('H4uTQ8CCFJYVQYZ8c2bH7hHGrrok6k4pWDhk5NQt9KR8'),
        lendingPoolTknAccount: new web3_js_1.PublicKey('E3fN9Wqesn2NbYGPjybFo5HTMpcyLPnWNHovQpQqqw8G'),
        lendingPoolFeeAccount: new web3_js_1.PublicKey('E3fN9Wqesn2NbYGPjybFo5HTMpcyLPnWNHovQpQqqw8G'),
        lendingPoolShareMint: new web3_js_1.PublicKey('CGj7e1g4ojW1RhSocF8AKWxHFaqCQjK8kBsbHqNB4BxP'),
        lendingPoolShareAccount: new web3_js_1.PublicKey('8BtR6Dzg4fLmLv4ZZdX1x22WYiPHvp8AtvRFgSMQwxdw'),
        lendingPoolCreditMint: new web3_js_1.PublicKey('74WmdEZG9rjvcAX2yczd5h54QeXnq62sVNLMz1DU2Abh'),
        lendingPoolCreditAccount: new web3_js_1.PublicKey('8kLAk9EiLV6qePNEtjnDLyWbRsySgMPSTihdWAEEv9H9'),
    },
    SAMO: {
        programId: exports.lendProgramId,
        tokenMint: tokens_1.TOKENS.SAMO.mintAddress,
        marketInfoAccount: new web3_js_1.PublicKey("4XNif294wbrxj6tJ8K5Rg7SuaEACnu9s2L27i28MQB6E"),
        marketOwner: new web3_js_1.PublicKey("7MBLg6oV5phip11YBbJPuq7u38kdzSi9PM3BifKSpLaR"),
        lendingMarketAuthority: new web3_js_1.PublicKey("sCDiYj7X7JmXg5fVq2nqED2q1Wqjo7PnqMgH3casMem"),
        lendingPoolInfoAccount: new web3_js_1.PublicKey('HJLkovTpxof1z9tVMuX1pWp1ePDx1zgmnGYfFMxFRAd'),
        lendingPoolTknAccount: new web3_js_1.PublicKey('J198uwytk8CPejCyNM9TMPJcLPCky6pZwGwP3n1CUfto'),
        lendingPoolFeeAccount: new web3_js_1.PublicKey('J198uwytk8CPejCyNM9TMPJcLPCky6pZwGwP3n1CUfto'),
        lendingPoolShareMint: new web3_js_1.PublicKey('A9H3fAqkWmRnnFzXXzydZHzyLQdzK5o9dMejCL27tqq8'),
        lendingPoolShareAccount: new web3_js_1.PublicKey('HP3sKy5xqLkoNnbuHRCPuFaQUtYskNnockBPcYo5PZut'),
        lendingPoolCreditMint: new web3_js_1.PublicKey('E5ejPudazVxEdQUsS37g8yUn5HD22FBHsGEKWkGRDR3F'),
        lendingPoolCreditAccount: new web3_js_1.PublicKey('23tLfFHXZR1tAimzVyEAdN7HWJWiqvjepaX5qBHkgQvT'),
    },
    POLIS: {
        programId: exports.lendProgramId,
        tokenMint: tokens_1.TOKENS.POLIS.mintAddress,
        marketInfoAccount: new web3_js_1.PublicKey("4XNif294wbrxj6tJ8K5Rg7SuaEACnu9s2L27i28MQB6E"),
        marketOwner: new web3_js_1.PublicKey("7MBLg6oV5phip11YBbJPuq7u38kdzSi9PM3BifKSpLaR"),
        lendingMarketAuthority: new web3_js_1.PublicKey("sCDiYj7X7JmXg5fVq2nqED2q1Wqjo7PnqMgH3casMem"),
        lendingPoolInfoAccount: new web3_js_1.PublicKey("BWJkfNQnYt817V6sKxEJ3sFcivcXZgfKGCe8Z9f4LcVj"),
        lendingPoolTknAccount: new web3_js_1.PublicKey("7Ad6djoPLzigyBnnWjoLh2gQ6Tbw2s6kAdsaVUBq6L8N"),
        lendingPoolFeeAccount: new web3_js_1.PublicKey("7Ad6djoPLzigyBnnWjoLh2gQ6Tbw2s6kAdsaVUBq6L8N"),
        lendingPoolShareMint: new web3_js_1.PublicKey("FenVvq6s6S3McD1BCm76Ktz1EvRNCB4qYKGFU76fB7Fj"),
        lendingPoolShareAccount: new web3_js_1.PublicKey("GKTqMGVCgXJaDzjYfPdgMbVfnzCKDj6KqRpykauw19do"),
        lendingPoolCreditMint: new web3_js_1.PublicKey("EkKFNt7PBRdWy8EpmZAbAZdvXZpSKvdwMfo8eotN1PEr"),
        lendingPoolCreditAccount: new web3_js_1.PublicKey("FGKikDsnBPQ5m7jgbNNbHaR1gh1T8GYJ5sPYWoYdoE4p"),
    },
    ATLAS: {
        programId: exports.lendProgramId,
        tokenMint: tokens_1.TOKENS.ATLAS.mintAddress,
        marketInfoAccount: new web3_js_1.PublicKey("4XNif294wbrxj6tJ8K5Rg7SuaEACnu9s2L27i28MQB6E"),
        marketOwner: new web3_js_1.PublicKey("7MBLg6oV5phip11YBbJPuq7u38kdzSi9PM3BifKSpLaR"),
        lendingMarketAuthority: new web3_js_1.PublicKey("sCDiYj7X7JmXg5fVq2nqED2q1Wqjo7PnqMgH3casMem"),
        lendingPoolInfoAccount: new web3_js_1.PublicKey("FnkmnUqazYmSm791rSTSXYbAzDzsuHJBgySgNZeCHGDJ"),
        lendingPoolTknAccount: new web3_js_1.PublicKey("E5ovYitzudRyUy7AS4U52eFgrq1rjr773rRGpYax1nmD"),
        lendingPoolFeeAccount: new web3_js_1.PublicKey("E5ovYitzudRyUy7AS4U52eFgrq1rjr773rRGpYax1nmD"),
        lendingPoolShareMint: new web3_js_1.PublicKey("4mygt5bFQrbXH9gNg75j1KVTrKGhvcYiQjir6FJ8afYH"),
        lendingPoolShareAccount: new web3_js_1.PublicKey("Eb4Ai2PXmRX3V9TX6awjNdBXZKGRyCgTmQu6Pni9NnTb"),
        lendingPoolCreditMint: new web3_js_1.PublicKey("3nAgm2XrSi3RNDWz4wCvUWwQW3QQE7s5i7MxNz8r8mGZ"),
        lendingPoolCreditAccount: new web3_js_1.PublicKey("3PwecBTLVt8zqKadSFDjeKH7Swzt1GxvU27urj66L89P"),
    },
    GENE: {
        programId: exports.lendProgramId,
        tokenMint: tokens_1.TOKENS.GENE.mintAddress,
        marketInfoAccount: new web3_js_1.PublicKey("4XNif294wbrxj6tJ8K5Rg7SuaEACnu9s2L27i28MQB6E"),
        marketOwner: new web3_js_1.PublicKey("7MBLg6oV5phip11YBbJPuq7u38kdzSi9PM3BifKSpLaR"),
        lendingMarketAuthority: new web3_js_1.PublicKey("sCDiYj7X7JmXg5fVq2nqED2q1Wqjo7PnqMgH3casMem"),
        lendingPoolInfoAccount: new web3_js_1.PublicKey('FrgyPPEPNC25ihob4ZobP3eYjh8uBoxLiobdMDX2EsMc'),
        lendingPoolTknAccount: new web3_js_1.PublicKey('A4UMfr97y35AvCaakKRAE5UBsb4UuvRQ2JiT23dyzDFu'),
        lendingPoolFeeAccount: new web3_js_1.PublicKey('A4UMfr97y35AvCaakKRAE5UBsb4UuvRQ2JiT23dyzDFu'),
        lendingPoolShareMint: new web3_js_1.PublicKey('9z3eDHueAMdUtym9Q2ku3hi5YXHTYjpFLp1YEEnxUHPV'),
        lendingPoolShareAccount: new web3_js_1.PublicKey('58teG2GNEoeYQUBa3VxQe5ZH5SPCZ6nUsrCuN5akbaQS'),
        lendingPoolCreditMint: new web3_js_1.PublicKey('2KR5Q6zCik6kFyiWPMMVCKC5HDAP6joGNhKhbEv6nFdY'),
        lendingPoolCreditAccount: new web3_js_1.PublicKey('7LBnkdaCZsM82t7G5MYKBoYAZetcnsqfU1cABxnPoquw'),
    },
    SONAR: {
        programId: exports.lendProgramId,
        tokenMint: tokens_1.TOKENS.SONAR.mintAddress,
        marketInfoAccount: new web3_js_1.PublicKey("4XNif294wbrxj6tJ8K5Rg7SuaEACnu9s2L27i28MQB6E"),
        marketOwner: new web3_js_1.PublicKey("7MBLg6oV5phip11YBbJPuq7u38kdzSi9PM3BifKSpLaR"),
        lendingMarketAuthority: new web3_js_1.PublicKey("sCDiYj7X7JmXg5fVq2nqED2q1Wqjo7PnqMgH3casMem"),
        lendingPoolInfoAccount: new web3_js_1.PublicKey('3USSy3sAZSCnUiCZJaTFt5hCHWSw87kg8tNPaEZh1p9D'),
        lendingPoolTknAccount: new web3_js_1.PublicKey('2F2yrLZQa3av7PxW7vdqiYBV4AwTbkCdswgN2BYGgtvs'),
        lendingPoolFeeAccount: new web3_js_1.PublicKey('2F2yrLZQa3av7PxW7vdqiYBV4AwTbkCdswgN2BYGgtvs'),
        lendingPoolShareMint: new web3_js_1.PublicKey('FvHrhBTTgo7q8uDX7gN5f7YzSXMrf3mwYtkay5NvkJzg'),
        lendingPoolShareAccount: new web3_js_1.PublicKey('DPMixAHcMsryStgytYBhvvtR2WbvyRxkMwG147NmXEPS'),
        lendingPoolCreditMint: new web3_js_1.PublicKey('HsKN1EC5zcArg1yoq1tMg7JKfxPW9m644mqCphrbLMC'),
        lendingPoolCreditAccount: new web3_js_1.PublicKey('7zqtZ8LhBYUUXoDCwDFcdLgNkYo8tZZqTdoFJqAL5ie8'),
    },
    DFL: {
        programId: exports.lendProgramId,
        tokenMint: tokens_1.TOKENS.DFL.mintAddress,
        marketInfoAccount: new web3_js_1.PublicKey("4XNif294wbrxj6tJ8K5Rg7SuaEACnu9s2L27i28MQB6E"),
        marketOwner: new web3_js_1.PublicKey("7MBLg6oV5phip11YBbJPuq7u38kdzSi9PM3BifKSpLaR"),
        lendingMarketAuthority: new web3_js_1.PublicKey("sCDiYj7X7JmXg5fVq2nqED2q1Wqjo7PnqMgH3casMem"),
        lendingPoolInfoAccount: new web3_js_1.PublicKey('74fWq2xnHBP1W6QhLDazCZAuZvp9XY3ck1PGNJg9hfhi'),
        lendingPoolTknAccount: new web3_js_1.PublicKey('5QNHLaujxMgFKYaJM55uYb3jyMGgvspizg7S5eVAD3d8'),
        lendingPoolFeeAccount: new web3_js_1.PublicKey('5QNHLaujxMgFKYaJM55uYb3jyMGgvspizg7S5eVAD3d8'),
        lendingPoolShareMint: new web3_js_1.PublicKey('CGhMaGeVLxxQGJh6Y2bEYfLazumuFyEXpxF1UgrcMRJi'),
        lendingPoolShareAccount: new web3_js_1.PublicKey('EHUF88xvJZyXzosAQ7Jzx5ozdvxCfZuwppSEo5b9pe2S'),
        lendingPoolCreditMint: new web3_js_1.PublicKey('26Eo8VcgDuNZSn8x5infdYxDKs7mYZj4JKxLUCkKP3DV'),
        lendingPoolCreditAccount: new web3_js_1.PublicKey('GYZgKQmYZBosdS3PsSRv83CdPFAtTRXRZaqdFVUnumiw'),
    },
    wbWBNB: {
        programId: exports.lendProgramId,
        tokenMint: tokens_1.TOKENS.wbWBNB.mintAddress,
        marketInfoAccount: new web3_js_1.PublicKey("4XNif294wbrxj6tJ8K5Rg7SuaEACnu9s2L27i28MQB6E"),
        marketOwner: new web3_js_1.PublicKey("7MBLg6oV5phip11YBbJPuq7u38kdzSi9PM3BifKSpLaR"),
        lendingMarketAuthority: new web3_js_1.PublicKey("sCDiYj7X7JmXg5fVq2nqED2q1Wqjo7PnqMgH3casMem"),
        lendingPoolInfoAccount: new web3_js_1.PublicKey('6RnKftFywvw8sceeWUM623BYyMeW216276G4atZ23CrF'),
        lendingPoolTknAccount: new web3_js_1.PublicKey('F7SQgbgunwx7PKc9i5zXJfSNsjU9xM5P8Sb18FCUBKyf'),
        lendingPoolFeeAccount: new web3_js_1.PublicKey('F7SQgbgunwx7PKc9i5zXJfSNsjU9xM5P8Sb18FCUBKyf'),
        lendingPoolShareMint: new web3_js_1.PublicKey('8EzbLKBQ8Rest1SWUtZR6zFvYTo4cCaQ73ni8sF9vtjA'),
        lendingPoolShareAccount: new web3_js_1.PublicKey('DgBnJvW9iQT5sP5YyVHeNbRjG8Xd2d53f19a2451eXkV'),
        lendingPoolCreditMint: new web3_js_1.PublicKey('6bMAnnzp3pim8wd8XypLGuqVFQ3LMuY3vkRRse1DkZ7o'),
        lendingPoolCreditAccount: new web3_js_1.PublicKey('CFsjid6WFmsyZozJkj43ssA5rRmYa97mzvfs8eRYqBeZ'),
    },
    'stSOL': {
        programId: exports.lendProgramId,
        tokenMint: tokens_1.TOKENS.stSOL.mintAddress,
        marketInfoAccount: new web3_js_1.PublicKey("4XNif294wbrxj6tJ8K5Rg7SuaEACnu9s2L27i28MQB6E"),
        marketOwner: new web3_js_1.PublicKey("7MBLg6oV5phip11YBbJPuq7u38kdzSi9PM3BifKSpLaR"),
        lendingMarketAuthority: new web3_js_1.PublicKey("sCDiYj7X7JmXg5fVq2nqED2q1Wqjo7PnqMgH3casMem"),
        lendingPoolInfoAccount: new web3_js_1.PublicKey('9ogtxmnUF94KaPn3dB3unT5yyu7rpzSwc5igysgqFxrU'),
        lendingPoolTknAccount: new web3_js_1.PublicKey('85pJTrAVdjHNvgCcUtefwkSe9RDKnHueyvs2uTocWmWs'),
        lendingPoolFeeAccount: new web3_js_1.PublicKey('85pJTrAVdjHNvgCcUtefwkSe9RDKnHueyvs2uTocWmWs'),
        lendingPoolShareMint: new web3_js_1.PublicKey('HyxfDg47HKS1rQUXWvJH2XQhwEJM3AqrHU7sK9bSNt5h'),
        lendingPoolShareAccount: new web3_js_1.PublicKey('3H64RCjRcHEJAJK9mZ3q1mBZJ2hWuv3tUvXVY7XgFL5L'),
        lendingPoolCreditMint: new web3_js_1.PublicKey('Hhh5mQQMuWjmyUNyYFDtuXKFp9wptPuLYvH9x98FSDq8'),
        lendingPoolCreditAccount: new web3_js_1.PublicKey('CRorEh6T7aXamCsjggckCMgF1ewVavo6CfXTPNkipson'),
    },
    SHDW: {
        programId: exports.lendProgramId,
        tokenMint: tokens_1.TOKENS.SHDW.mintAddress,
        marketInfoAccount: new web3_js_1.PublicKey("4XNif294wbrxj6tJ8K5Rg7SuaEACnu9s2L27i28MQB6E"),
        marketOwner: new web3_js_1.PublicKey("7MBLg6oV5phip11YBbJPuq7u38kdzSi9PM3BifKSpLaR"),
        lendingPoolInfoAccount: new web3_js_1.PublicKey('8NvPwUxtBohPg45bAmzzB45qLajpfptMVgqgu5vmCNgi'),
        lendingMarketAuthority: new web3_js_1.PublicKey("sCDiYj7X7JmXg5fVq2nqED2q1Wqjo7PnqMgH3casMem"),
        lendingPoolTknAccount: new web3_js_1.PublicKey('EoVqoTtczavGmATCagSQU11c74MkL8enp96fGkQz6uqK'),
        lendingPoolShareMint: new web3_js_1.PublicKey('AJu1s738dGsZ8mV2XKTqjuMiAiqNGaRsixTR4Czx4mJ2'),
        lendingPoolShareAccount: new web3_js_1.PublicKey('48YJFjo3zLGSkUDXpa8mte3EMFNSPvy6pE9wUhoPSLnH'),
        lendingPoolCreditMint: new web3_js_1.PublicKey('3fu3y5yGbBPqDpwKjN1PsVurm4af6uQnpkkuSo5SxZQa'),
        lendingPoolCreditAccount: new web3_js_1.PublicKey('2zxQFrSGMSWbHMzMjxDN8ceqgbJVUyyEEwR75R5zzh4M'),
        lendingPoolFeeAccount: new web3_js_1.PublicKey('3H3QXT9oEG5DaoWfQ727jgJdhVewgrTNc8jsGgBsXvQE'),
    },
    BASIS: {
        programId: exports.lendProgramId,
        tokenMint: tokens_1.TOKENS.BASIS.mintAddress,
        marketInfoAccount: new web3_js_1.PublicKey("4XNif294wbrxj6tJ8K5Rg7SuaEACnu9s2L27i28MQB6E"),
        marketOwner: new web3_js_1.PublicKey("7MBLg6oV5phip11YBbJPuq7u38kdzSi9PM3BifKSpLaR"),
        lendingPoolInfoAccount: new web3_js_1.PublicKey('499SnZR7dFzLU6BF9v9obfSCsmgui3FBtYtFDakD89zQ'),
        lendingMarketAuthority: new web3_js_1.PublicKey("sCDiYj7X7JmXg5fVq2nqED2q1Wqjo7PnqMgH3casMem"),
        lendingPoolTknAccount: new web3_js_1.PublicKey('E85Eyi4NmBp2Der1ntoNNx3AH5FHmjp4gGoUPW7qYVJ8'),
        lendingPoolShareMint: new web3_js_1.PublicKey('9s2iRZBzSNCNTUtuzSRHcv3q4Q9SaFFhSFZVnFXGtHUp'),
        lendingPoolShareAccount: new web3_js_1.PublicKey('5phQAZCwdfbYqj1tsv4xYWREpLZuxhqnMaZ126RAfVNb'),
        lendingPoolCreditMint: new web3_js_1.PublicKey('AVRoCxDDfx525L1XJJ5JbKrPPbuHs1JL6Z17vWS3HpKc'),
        lendingPoolCreditAccount: new web3_js_1.PublicKey('FeMzBdYfGJtqeRAadiC4fStY13cjxe6GutTyFCFsP1vL'),
        lendingPoolFeeAccount: new web3_js_1.PublicKey('CtdMN3hrU9yq2GJpojbU1oQZbMXJyuVZYMfNkAPbgaZN'),
    },
    wUST: {
        programId: exports.lendProgramId,
        tokenMint: tokens_1.TOKENS.wUST.mintAddress,
        marketInfoAccount: new web3_js_1.PublicKey("4XNif294wbrxj6tJ8K5Rg7SuaEACnu9s2L27i28MQB6E"),
        marketOwner: new web3_js_1.PublicKey("7MBLg6oV5phip11YBbJPuq7u38kdzSi9PM3BifKSpLaR"),
        lendingPoolInfoAccount: new web3_js_1.PublicKey('G1PYcFc56DQHDjSyt8zRKcnxbBrKJq4Vehurzrt2KUzs'),
        lendingMarketAuthority: new web3_js_1.PublicKey("sCDiYj7X7JmXg5fVq2nqED2q1Wqjo7PnqMgH3casMem"),
        lendingPoolTknAccount: new web3_js_1.PublicKey('DKYpvhtPyU9yWkGZx5VM4fA1isLwpcmQ8ttqhPJgMMrJ'),
        lendingPoolShareMint: new web3_js_1.PublicKey('FPnFwW1ASFLBReVz1EPWVGkbrsnrekVkEcQnrBRkvDXS'),
        lendingPoolShareAccount: new web3_js_1.PublicKey('MeLTPseZyHm2aWSk9Gda9d6xWM8DS5QySFKBS3Vyzsv'),
        lendingPoolCreditMint: new web3_js_1.PublicKey('AbPJaMzRetUebzRV3mtoXtyPDVwEBEHif3EibVKhfcTs'),
        lendingPoolCreditAccount: new web3_js_1.PublicKey('mYi2FARcy26GxkDuTneTybM75izMAR7swioxUMvyPRa'),
        lendingPoolFeeAccount: new web3_js_1.PublicKey('5wTjKzJyEJHbjw4fsZM51fKBDwiq5JSA3bSzzRE2PXob'),
    },
    'GST': {
        programId: exports.lendProgramId,
        tokenMint: tokens_1.TOKENS.GST.mintAddress,
        marketInfoAccount: new web3_js_1.PublicKey("4XNif294wbrxj6tJ8K5Rg7SuaEACnu9s2L27i28MQB6E"),
        marketOwner: new web3_js_1.PublicKey("7MBLg6oV5phip11YBbJPuq7u38kdzSi9PM3BifKSpLaR"),
        lendingMarketAuthority: new web3_js_1.PublicKey("sCDiYj7X7JmXg5fVq2nqED2q1Wqjo7PnqMgH3casMem"),
        lendingPoolInfoAccount: new web3_js_1.PublicKey('93zxnt1gPepvUjWYPP32uANnhAK6NNphGD6A2yQFhk4s'),
        lendingPoolTknAccount: new web3_js_1.PublicKey('4Y7VyaAWEDaQSkeYXbXRf1d82ZPViVRZ7Dt4fjHKgotL'),
        lendingPoolShareMint: new web3_js_1.PublicKey('4GDuSZwHUQvL6zga2UNdjgXSqXjcWaUYc4S36N9q4567'),
        lendingPoolShareAccount: new web3_js_1.PublicKey('GjmnPLwAAA5aZCfN5sCwcbartonu8iyVMPqSnfkyvtC9'),
        lendingPoolCreditMint: new web3_js_1.PublicKey('8Nq8YV7BYoe3dqdTX996TRDB2iYryAHbzZTXntt2A3Uo'),
        lendingPoolCreditAccount: new web3_js_1.PublicKey('8dzJgZxp7rBEbKhvFYXTMXKLmkqscn9UXwzcGA14fEgN'),
        lendingPoolFeeAccount: new web3_js_1.PublicKey('ERyCAJh59hn8187Kn6wiKpeXdnGw4hptrbn55vh9YJQ1'),
    },
    'GMT': {
        programId: exports.lendProgramId,
        tokenMint: tokens_1.TOKENS.GMT.mintAddress,
        marketInfoAccount: new web3_js_1.PublicKey("4XNif294wbrxj6tJ8K5Rg7SuaEACnu9s2L27i28MQB6E"),
        lendingMarketAuthority: new web3_js_1.PublicKey("sCDiYj7X7JmXg5fVq2nqED2q1Wqjo7PnqMgH3casMem"),
        marketOwner: new web3_js_1.PublicKey("7MBLg6oV5phip11YBbJPuq7u38kdzSi9PM3BifKSpLaR"),
        lendingPoolInfoAccount: new web3_js_1.PublicKey('9b5QWW2SqjmhRxtYTHXj6LFkmUkahrLAMQer3rEurd1k'),
        lendingPoolTknAccount: new web3_js_1.PublicKey('7ZosDG7JWwXFnzhQ2aQ1ghCSQ9mrQzvH3ZhduhRPHgJB'),
        lendingPoolShareMint: new web3_js_1.PublicKey('9R6toP3xTFBWoqwBY8bfjUj6Tyd5hyR192jY4NeMmkNg'),
        lendingPoolShareAccount: new web3_js_1.PublicKey('BPPtMXgGkpo5HuDbiFyQWiA5r4BfxRWfwUA8XoKPixb7'),
        lendingPoolCreditMint: new web3_js_1.PublicKey('B1yWMWp7nvSV2upJxFUxpUkDiZAVD2MabnRgaBtR14Dm'),
        lendingPoolCreditAccount: new web3_js_1.PublicKey('6NvFHm9p6oSK8DJ9jurTMkC8KBo2hCk8UQm19Nub7R6X'),
        lendingPoolFeeAccount: new web3_js_1.PublicKey('ejLn4fFkVS29aYksC87TQYwcu8wy1tvc2AySpfHGabQ'),
    },
    ZBC: {
        programId: exports.lendProgramId,
        tokenMint: tokens_1.TOKENS.ZBC.mintAddress,
        marketInfoAccount: new web3_js_1.PublicKey("4XNif294wbrxj6tJ8K5Rg7SuaEACnu9s2L27i28MQB6E"),
        lendingMarketAuthority: new web3_js_1.PublicKey("sCDiYj7X7JmXg5fVq2nqED2q1Wqjo7PnqMgH3casMem"),
        marketOwner: new web3_js_1.PublicKey("7MBLg6oV5phip11YBbJPuq7u38kdzSi9PM3BifKSpLaR"),
        lendingPoolInfoAccount: new web3_js_1.PublicKey('6uPerEUyomCC5qn4z9xF3fbVa7SCytcKoCVannPDez5R'),
        lendingPoolTknAccount: new web3_js_1.PublicKey('7quTFi1N9EZfz7jUx7ui9wx9rk1rb3F1Enjbh8Me2tSs'),
        lendingPoolShareMint: new web3_js_1.PublicKey('HUry6RSbDasMfU7LXZ5QR42YajkbMuLBQkyV742xgFKx'),
        lendingPoolShareAccount: new web3_js_1.PublicKey('BHuVdzrDTssFdWTqt15fsdJzJuHESQrMhSnc3jsJ8X2Y'),
        lendingPoolCreditMint: new web3_js_1.PublicKey('ADWAgFjannL985LDKQsQ9B833BFDMj4q8qQappe2F3X9'),
        lendingPoolCreditAccount: new web3_js_1.PublicKey('5xqkszmojpTFy3wGQTXFB7XoLyWXxw7ymmHN9iNYVdUK'),
        lendingPoolFeeAccount: new web3_js_1.PublicKey('qt8KMibr5UcB1WZGVdgyFNGvmzEtp3faUqfcKDytLzq'),
    }
};
exports.lendingPoolList = [
    {
        pool: 'USDC',
        scale: 6
    },
    {
        pool: 'USDT',
        scale: 6
    },
    {
        pool: 'wUST',
        scale: 6,
    },
    {
        pool: 'SOL',
        scale: 9
    },
    {
        pool: 'mSOL',
        scale: 9
    },
    {
        pool: 'stSOL',
        scale: tokens_1.TOKENS.stSOL.decimals
    },
    {
        pool: 'BTC',
        scale: 6
    },
    {
        pool: 'ETH',
        scale: 6
    },
    {
        pool: 'SRM',
        scale: 6
    },
    {
        pool: 'ORCA',
        scale: 6
    },
    {
        pool: 'whETH',
        scale: 8
    },
    {
        pool: 'weUNI',
        scale: tokens_1.TOKENS.weUNI.decimals
    },
    {
        pool: 'weSUSHI',
        scale: tokens_1.TOKENS.weSUSHI.decimals
    },
    {
        pool: 'RAY',
        scale: 6
    },
    {
        pool: 'SAMO',
        scale: tokens_1.TOKENS.SAMO.decimals
    },
    {
        pool: 'POLIS',
        scale: tokens_1.TOKENS.POLIS.decimals
    },
    {
        pool: 'ATLAS',
        scale: tokens_1.TOKENS.ATLAS.decimals
    },
    {
        pool: 'GENE',
        scale: tokens_1.TOKENS.GENE.decimals
    },
    {
        pool: 'DFL',
        scale: tokens_1.TOKENS.DFL.decimals
    },
    {
        pool: 'wbWBNB',
        scale: tokens_1.TOKENS.wbWBNB.decimals
    },
    {
        pool: 'SHDW',
        scale: tokens_1.TOKENS.SHDW.decimals
    },
    {
        pool: 'BASIS',
        scale: tokens_1.TOKENS.BASIS.decimals
    },
    {
        pool: 'GST',
        scale: tokens_1.TOKENS.GST.decimals
    },
    {
        pool: 'GMT',
        scale: tokens_1.TOKENS.GMT.decimals
    },
    {
        pool: 'ZBC',
        scale: tokens_1.TOKENS.ZBC.decimals
    }
];
