// francium-sdk/dist/constants/lend/rewards.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.lendRewardInfo = exports.lendRewardProgramId = exports.RewardUserLayout = exports.RewardPoolLayout = void 0;
const web3_js_1 = require("@solana/web3.js");
const BufferLayout = require("buffer-layout");
exports.RewardPoolLayout = BufferLayout.struct([
    BufferLayout.u8("version"),
    BufferLayout.u8("is_dual_rewards"),
    BufferLayout.blob(32, "admin"),
    BufferLayout.blob(32, "token_program_id"),
    BufferLayout.blob(32, "pool_authority"),
    BufferLayout.blob(32, "staked_token_mint"),
    BufferLayout.blob(32, "staked_token_account"),
    BufferLayout.blob(32, "rewards_token_mint"),
    BufferLayout.blob(32, "rewards_token_account"),
    BufferLayout.blob(32, "rewards_token_mint_b"),
    BufferLayout.blob(32, "rewards_token_account_b"),
    BufferLayout.nu64("pool_stake_cap"),
    BufferLayout.nu64("user_stake_cap"),
    BufferLayout.nu64("rewards_start_slot"),
    BufferLayout.nu64("rewards_end_slot"),
    BufferLayout.nu64("rewards_per_day"),
    BufferLayout.nu64("rewards_start_slot_b"),
    BufferLayout.nu64("rewards_end_slot_b"),
    BufferLayout.nu64("rewards_per_day_b"),
    BufferLayout.nu64("total_staked_amount"),
    BufferLayout.nu64("last_update_slot"),
    BufferLayout.blob(16, "accumulated_rewards_per_share"),
    BufferLayout.blob(16, "accumulated_rewards_per_share_b"),
    BufferLayout.blob(128, "padding")
]);
exports.RewardUserLayout = BufferLayout.struct([
    BufferLayout.u8("version"),
    BufferLayout.nu64("staked_amount"),
    BufferLayout.nu64("rewards_debt"),
    BufferLayout.nu64("rewards_debt_b"),
    BufferLayout.blob(32, "farming_pool"),
    BufferLayout.blob(32, "user_main"),
    BufferLayout.blob(32, "stake_token_account"),
    BufferLayout.blob(32, "rewards_token_accont"),
    BufferLayout.blob(32, "rewards_token_account_b"),
    BufferLayout.blob(128, "padding")
]);
exports.lendRewardProgramId = new web3_js_1.PublicKey("3Katmm9dhvLQijAvomteYMo6rfVbY5NaCRNq9ZBqBgr6");
exports.lendRewardInfo = {
    USDC: {
        programId: exports.lendRewardProgramId,
        farmingPoolAccount: new web3_js_1.PublicKey('8Eq2XZRQe2EjYiNmu7Lhgb2xVHqJ5wFvcVU6yH3CUn34'),
        farmingPoolAuthority: new web3_js_1.PublicKey('4NWwKzVvEfKCsMeauE4cZHRR9K91FsFauxnrW6pK8H2E'),
        farmingPoolStakeTknMint: new web3_js_1.PublicKey('62fDf5daUJ9jBz8Xtj6Bmw1bh1DvHn8AG4L9hMmxCzpu'),
        farmingPoolStakeTknAccount: new web3_js_1.PublicKey('3yNu5pg2DhtaxZbAwgUSsVnemqMn1WqxnBn6tgKGj7R2'),
        farmingPoolRewardsTknMint: new web3_js_1.PublicKey('Hjibp1cn2bSk1dkTdpbxez3YAiBGTLjzc8xZ8LbCCUHS'),
        farmingPoolRewardsTknAccount: new web3_js_1.PublicKey('34R2ZVwg6uvJWFYjQ2LrrKFFaZ7CgsyZbMKwvfjxkvip'),
        farmingPoolRewardsTknMintB: new web3_js_1.PublicKey('EgiD69Uhf8t13CRPKz1btmtHj7SogeEjyPHfnT4d13XN'),
        farmingPoolRewardsTknAccountB: new web3_js_1.PublicKey('FGAh5YjdcyzQ841skvGQGWyejK3uPiwpEdtMncJqe7f9')
    },
    PAI: {
        programId: exports.lendRewardProgramId,
        farmingPoolAccount: new web3_js_1.PublicKey('2Dp5WMATsyfHtMEkU2JTrffc5onPCLqZFybcYvPAPxNP'),
        farmingPoolAuthority: new web3_js_1.PublicKey('7A88TMr5kQpHWidb5w9zeZr4dXMse6DsWTELbLe1WEyV'),
        farmingPoolStakeTknMint: new web3_js_1.PublicKey('HDvD8a4VWbkHNG7hb4CBumNhn41DyKL51qVYBNH73o23'),
        farmingPoolStakeTknAccount: new web3_js_1.PublicKey('8QaCpnJP61u8qm9ZJYdXgQWGNacoefmU2bDKfmLPHSmE'),
        farmingPoolRewardsTknMint: new web3_js_1.PublicKey('Hjibp1cn2bSk1dkTdpbxez3YAiBGTLjzc8xZ8LbCCUHS'),
        farmingPoolRewardsTknAccount: new web3_js_1.PublicKey('7B2BnzmT779H3KzAn8raQQEcRKSw4GhPPDS657X7NRDs'),
        farmingPoolRewardsTknMintB: new web3_js_1.PublicKey('Ea5SjE2Y6yvCeW5dYTn7PYMuW5ikXkvbGdcmSnXeaLjS'),
        farmingPoolRewardsTknAccountB: new web3_js_1.PublicKey('GGwsqBzz6L7Wvq7ZKdn7yQBCU5qLdqwMfVZZQWp8uYZn')
    },
    RAY: {
        programId: exports.lendRewardProgramId,
        farmingPoolAccount: new web3_js_1.PublicKey('9G2EXWwSaqJNgk9AWkx5u6JsUpdCaeh7ZUzsjJsSJkYz'),
        farmingPoolAuthority: new web3_js_1.PublicKey('gbPTC9F2tnP9z9xzjeQWjNXBsuHw9ZjkG9NahZ7wDPa'),
        farmingPoolStakeTknMint: new web3_js_1.PublicKey('5uZERkJVyhCABEdocEssPZyMXYa8GJpLhngi95yXr9jc'),
        farmingPoolStakeTknAccount: new web3_js_1.PublicKey('GKV52HXiEDR8Qqazdp9LA5Aiwjny94v9CavCVAt7TvxS'),
        farmingPoolRewardsTknMint: new web3_js_1.PublicKey('4k3Dyjzvzp8eMZWUXbBCjEvwSkkk59S5iCNLY3QrkX6R'),
        farmingPoolRewardsTknAccount: new web3_js_1.PublicKey('9dvYPm8LjhTrKF3bJ4Lt8YwQqxhxzWWXWGnbqfamNVQK'),
        farmingPoolRewardsTknMintB: new web3_js_1.PublicKey('4k3Dyjzvzp8eMZWUXbBCjEvwSkkk59S5iCNLY3QrkX6R'),
        farmingPoolRewardsTknAccountB: new web3_js_1.PublicKey('9dvYPm8LjhTrKF3bJ4Lt8YwQqxhxzWWXWGnbqfamNVQK'),
    },
    USDT: {
        programId: exports.lendRewardProgramId,
        farmingPoolAccount: new web3_js_1.PublicKey('AKi1GE5Xa5boTwNEy4u2PaHqtQBVFcKXVU5oqXFFYST8'),
        farmingPoolAuthority: new web3_js_1.PublicKey('CQ83mRhE3AhM1yLiGkgKut1Fsy9dZPmzCjNhLu1vKE8Y'),
        farmingPoolStakeTknMint: new web3_js_1.PublicKey('8TtsTZQ6mU4YSW5jw4sDjSDEG1CW8QLpK7C9g5TVgBvn'),
        farmingPoolStakeTknAccount: new web3_js_1.PublicKey('CiwYeQcjMyvt7gvkwNAQRq8DEbtAh7J3K3Zrfsb54got'),
        farmingPoolRewardsTknMint: new web3_js_1.PublicKey('Hjibp1cn2bSk1dkTdpbxez3YAiBGTLjzc8xZ8LbCCUHS'),
        farmingPoolRewardsTknAccount: new web3_js_1.PublicKey('9QsFN5L8Ld9aR8FFZ8gTM5euG6us7Rw7QWZUUthptoib'),
        farmingPoolRewardsTknMintB: new web3_js_1.PublicKey('Hjibp1cn2bSk1dkTdpbxez3YAiBGTLjzc8xZ8LbCCUHS'),
        farmingPoolRewardsTknAccountB: new web3_js_1.PublicKey('BC3GujgFNpbo2rYRfit7Rr3Af769Jo6ApN3vr58BQazw')
    },
    SRM: {
        programId: exports.lendRewardProgramId,
        farmingPoolAccount: new web3_js_1.PublicKey('63Xrt9Z5jC3D9pR2RVL2D57z5bspaX6pqdDVW9XnAMU8'),
        farmingPoolAuthority: new web3_js_1.PublicKey('D1QkTSH7UaNeCsiTsB5PGF27BAeAwu9TBNiCX4upARzg'),
        farmingPoolStakeTknMint: new web3_js_1.PublicKey('6CZhCFwA7hbqDEoVUQdFpHaCGQU7fSSwhaJWdG4DmsQ6'),
        farmingPoolStakeTknAccount: new web3_js_1.PublicKey('6E7dXsDkdErDuG4z7ED8knoAR9EFtnKApVvYQyytm1kU'),
        farmingPoolRewardsTknMint: new web3_js_1.PublicKey('Hjibp1cn2bSk1dkTdpbxez3YAiBGTLjzc8xZ8LbCCUHS'),
        farmingPoolRewardsTknAccount: new web3_js_1.PublicKey('gw2RdaKNv8TW7Epvz41vCZZn2AppMRoc9cZXsVtKZ1k'),
        farmingPoolRewardsTknMintB: new web3_js_1.PublicKey('Hjibp1cn2bSk1dkTdpbxez3YAiBGTLjzc8xZ8LbCCUHS'),
        farmingPoolRewardsTknAccountB: new web3_js_1.PublicKey('gw2RdaKNv8TW7Epvz41vCZZn2AppMRoc9cZXsVtKZ1k')
    },
    ETH: {
        programId: exports.lendRewardProgramId,
        farmingPoolAccount: new web3_js_1.PublicKey('7rKAmXDUhqZrzM2nZbM2PxbxsErAjGPxbRgr2fFxQ9Y'),
        farmingPoolAuthority: new web3_js_1.PublicKey('FzoCLipUnS18zMvag9ffr3ATPJicrphYZyEZeB9Acp4G'),
        farmingPoolStakeTknMint: new web3_js_1.PublicKey('B8QXcUv5FFwyHH5V88g6PhYBc8fQvwSpza4C9PsiRpQD'),
        farmingPoolStakeTknAccount: new web3_js_1.PublicKey('BR1gbF7jjcTaXTJfASSm1hz1h7A8My7F4meCgB9dKXwq'),
        farmingPoolRewardsTknMint: new web3_js_1.PublicKey('Hjibp1cn2bSk1dkTdpbxez3YAiBGTLjzc8xZ8LbCCUHS'),
        farmingPoolRewardsTknAccount: new web3_js_1.PublicKey('H1aJwXV3LwEghvPKJpNkjdedH7dmDyDLjrTChtomhJKQ'),
        farmingPoolRewardsTknMintB: new web3_js_1.PublicKey('Hjibp1cn2bSk1dkTdpbxez3YAiBGTLjzc8xZ8LbCCUHS'),
        farmingPoolRewardsTknAccountB: new web3_js_1.PublicKey('H1aJwXV3LwEghvPKJpNkjdedH7dmDyDLjrTChtomhJKQ')
    },
    SOL: {
        programId: exports.lendRewardProgramId,
        farmingPoolAccount: new web3_js_1.PublicKey('G8MpQfTa18RcGK6xdKxKqG3u9NXWwqEMN3YM4MAFSukS'),
        farmingPoolAuthority: new web3_js_1.PublicKey('9EkVgtRxTWJWtksC2maF7YJqKyDR45z3PrxmzUMZwGpm'),
        farmingPoolStakeTknMint: new web3_js_1.PublicKey('92Zst3rEoJsyZUW2yNc2811GtkBwFrm3tqgxMTtFWuSB'),
        farmingPoolStakeTknAccount: new web3_js_1.PublicKey('5Dwb6bW6nJ7tawnXXuJLRRsHnpDWrsQGMBNePsi2HHNc'),
        farmingPoolRewardsTknMint: new web3_js_1.PublicKey('Hjibp1cn2bSk1dkTdpbxez3YAiBGTLjzc8xZ8LbCCUHS'),
        farmingPoolRewardsTknAccount: new web3_js_1.PublicKey('CRmaS1yBemFcHAooAfRUM41ZCBpnM7FCRKi7x71YXnTj'),
        farmingPoolRewardsTknMintB: new web3_js_1.PublicKey('Hjibp1cn2bSk1dkTdpbxez3YAiBGTLjzc8xZ8LbCCUHS'),
        farmingPoolRewardsTknAccountB: new web3_js_1.PublicKey('CRmaS1yBemFcHAooAfRUM41ZCBpnM7FCRKi7x71YXnTj')
    },
    mSOL: {
        programId: exports.lendRewardProgramId,
        farmingPoolAccount: new web3_js_1.PublicKey('Hu2QsZ57NHmgKEReXYPfS79amdqyyuxzbwSdxvd1UTwv'),
        farmingPoolAuthority: new web3_js_1.PublicKey('7af5t7Ms526TYXEh5Y4nvot6xbWyGaiQnNJf3D9dygsb'),
        farmingPoolStakeTknMint: new web3_js_1.PublicKey('4mntRxKJWib1YTWNV4aLfszF6FeziBHdwnNAwGSfKKMX'),
        farmingPoolStakeTknAccount: new web3_js_1.PublicKey('39bnyhokuj3RCsLZnZsjbctRtx7rsYV4hQNoxQEKx9xy'),
        farmingPoolRewardsTknMint: new web3_js_1.PublicKey('MNDEFzGvMt87ueuHvVU9VcTqsAP5b3fTGPsHuuPA5ey'),
        farmingPoolRewardsTknAccount: new web3_js_1.PublicKey('2JEph4UfH6H49fmsYBZg7zSXKTUENszypmPC5qMmefkc'),
        farmingPoolRewardsTknMintB: new web3_js_1.PublicKey('MNDEFzGvMt87ueuHvVU9VcTqsAP5b3fTGPsHuuPA5ey'),
        farmingPoolRewardsTknAccountB: new web3_js_1.PublicKey('BFBh633oDNn7hV89mA6TSUPGU6JeMdKzXVBHUi3Ab6Yd'),
    },
    BTC: {
        programId: exports.lendRewardProgramId,
        farmingPoolAccount: new web3_js_1.PublicKey('9LPNrtfkDgNTf7hex7PvufE1KypP54eU6AmStgn5Bo8g'),
        farmingPoolAuthority: new web3_js_1.PublicKey('DaB9HtEeWGfH25dFXMbYpxsUkZtqvbxQPjdsoxmpNCHK'),
        farmingPoolStakeTknMint: new web3_js_1.PublicKey('2G9iwy9zfLaXB2bFiqSA7YbKEvtAEXVmdvGTF28jQVgg'),
        farmingPoolStakeTknAccount: new web3_js_1.PublicKey('DNuEdaUXrYCRpELZssM9t2ZUQ6V4GJiijCEaXRr54Gpp'),
        farmingPoolRewardsTknMint: new web3_js_1.PublicKey('Hjibp1cn2bSk1dkTdpbxez3YAiBGTLjzc8xZ8LbCCUHS'),
        farmingPoolRewardsTknAccount: new web3_js_1.PublicKey('GAeyyaFDBwX7LPSp5oQBRdZDCVUePeaJFDujCVkj5NDV'),
        farmingPoolRewardsTknMintB: new web3_js_1.PublicKey('Hjibp1cn2bSk1dkTdpbxez3YAiBGTLjzc8xZ8LbCCUHS'),
        farmingPoolRewardsTknAccountB: new web3_js_1.PublicKey('4uL23rEj7fzNCFZ8Bv3SunLhtMamX3nAciYCSdSmAP1R'),
    },
    whETH: {
        programId: exports.lendRewardProgramId,
        farmingPoolAccount: new web3_js_1.PublicKey('4LCHfdGYZVkNzmPrVt6LyvzEPyJ9uWgnna9Fm4GU3Mib'),
        farmingPoolAuthority: new web3_js_1.PublicKey('GPMxFBWur7UWwpyCbU29jTvWgaoJJKVke2nwd6azJEyQ'),
        farmingPoolStakeTknMint: new web3_js_1.PublicKey('Hk8b6i2C7PxFBPdd1TCrGssuZpCDEg4fh8yFuyHNJuJe'),
        farmingPoolStakeTknAccount: new web3_js_1.PublicKey('4uW5PiPXRBQmcdU4Sreg1uJmfqBGazeH5ah2MnC1NSDL'),
        farmingPoolRewardsTknMint: new web3_js_1.PublicKey("7vfCXTUXx5WJV5JADk17DUJ4ksgau7utNKj4b963voxs"),
        farmingPoolRewardsTknAccount: new web3_js_1.PublicKey("9unatyGE1mM46TnWpwhqgr7n176m4pZ8j1vtVhF84i83"),
        farmingPoolRewardsTknMintB: new web3_js_1.PublicKey('Hjibp1cn2bSk1dkTdpbxez3YAiBGTLjzc8xZ8LbCCUHS'),
        farmingPoolRewardsTknAccountB: new web3_js_1.PublicKey('7LyyPUnM2MiAH1k8xN8pN2pQJZReCprFKS78e16tweud'),
    },
    weUNI: {
        programId: exports.lendRewardProgramId,
        farmingPoolAccount: new web3_js_1.PublicKey('ArRauaQ8xnFMdNdjdJe5vCp1h2v8iWWeHSNjf2dhMGdo'),
        farmingPoolAuthority: new web3_js_1.PublicKey('BuujLrfraC8o5c6257Fv5jsWkoWwAn4W66eKb121mU6C'),
        farmingPoolStakeTknMint: new web3_js_1.PublicKey('CGj7e1g4ojW1RhSocF8AKWxHFaqCQjK8kBsbHqNB4BxP'),
        farmingPoolStakeTknAccount: new web3_js_1.PublicKey('5T5YxRxXbTidTpbcJqVdGu3UBcSc6x2W446tLSEfSEBA'),
        farmingPoolRewardsTknMint: new web3_js_1.PublicKey('Hjibp1cn2bSk1dkTdpbxez3YAiBGTLjzc8xZ8LbCCUHS'),
        farmingPoolRewardsTknAccount: new web3_js_1.PublicKey('3S6kQMUYhGKvXMRM74nsPkhNbx8PrUEsw4NdjJMbaF9K'),
        farmingPoolRewardsTknMintB: new web3_js_1.PublicKey('Hjibp1cn2bSk1dkTdpbxez3YAiBGTLjzc8xZ8LbCCUHS'),
        farmingPoolRewardsTknAccountB: new web3_js_1.PublicKey('hbh6ziFx6FRWdoRGqtTuRVFwr9i7iXhFNtuN66ufK4B'),
    },
    weSUSHI: {
        programId: exports.lendRewardProgramId,
        farmingPoolAccount: new web3_js_1.PublicKey('21WzBAky2CbcmHvrxgyFhCytdF3xQA6YdzWXN7JYk7Ga'),
        farmingPoolAuthority: new web3_js_1.PublicKey('GE7jFYgkUc533u1V38Y5hN1icrdAgCTgftHW6q2tu9ZV'),
        farmingPoolStakeTknMint: new web3_js_1.PublicKey('EEhiV55jAt5JDpeH3GF4VGrStiPn5gCeWmqffyTp9B4E'),
        farmingPoolStakeTknAccount: new web3_js_1.PublicKey('ES5w4AaFNfjartY74tw3dr8XjLc2DLbFpD9qmXagnE83'),
        farmingPoolRewardsTknMint: new web3_js_1.PublicKey('Hjibp1cn2bSk1dkTdpbxez3YAiBGTLjzc8xZ8LbCCUHS'),
        farmingPoolRewardsTknAccount: new web3_js_1.PublicKey('DWy2TzewHtoYw6QEczXvJoDRk6rwNrdyGkSo7t5hTzCM'),
        farmingPoolRewardsTknMintB: new web3_js_1.PublicKey('Hjibp1cn2bSk1dkTdpbxez3YAiBGTLjzc8xZ8LbCCUHS'),
        farmingPoolRewardsTknAccountB: new web3_js_1.PublicKey('5R1NiQM9U99DzMN21Kx3hYxfQAupEVeQ7tPsmNNa5KVo'),
    },
    SAMO: {
        programId: exports.lendRewardProgramId,
        farmingPoolAccount: new web3_js_1.PublicKey('J5vPqifPvVYcYfgwhAkrjFq8xnsJBjFhvSwVpTbpqs5G'),
        farmingPoolAuthority: new web3_js_1.PublicKey('3ocZg6J44UVx6gWFC8PuTDumFSsVNh7W8Md3Cf5jXR6P'),
        farmingPoolStakeTknMint: new web3_js_1.PublicKey('A9H3fAqkWmRnnFzXXzydZHzyLQdzK5o9dMejCL27tqq8'),
        farmingPoolStakeTknAccount: new web3_js_1.PublicKey('7zLG1aWBvgwCgkDEsKHAkPgyuUGxdYPSDUPVoXi4LuVi'),
        farmingPoolRewardsTknMint: new web3_js_1.PublicKey('Hjibp1cn2bSk1dkTdpbxez3YAiBGTLjzc8xZ8LbCCUHS'),
        farmingPoolRewardsTknAccount: new web3_js_1.PublicKey('CYHmcVin4cQzdjcEcAYJki3qQuHMWq2NR15ienvRnXwc'),
        farmingPoolRewardsTknMintB: new web3_js_1.PublicKey('Hjibp1cn2bSk1dkTdpbxez3YAiBGTLjzc8xZ8LbCCUHS'),
        farmingPoolRewardsTknAccountB: new web3_js_1.PublicKey('GHM6YvgVKuoBEwcaMRJjQ2yEDfyCJg7CcK2eKTZ1k7pn'),
    },
    POLIS: {
        programId: exports.lendRewardProgramId,
        farmingPoolAccount: new web3_js_1.PublicKey('9fdYMtX2xuqUTUVFrPR2fegRHh7qakLkXwX1RVw9aTHH'),
        farmingPoolAuthority: new web3_js_1.PublicKey('G5BxwBHd1j7KdWbHo2HwZCmEWtXjvQ8uHRVLo2g1cikM'),
        farmingPoolStakeTknMint: new web3_js_1.PublicKey('FenVvq6s6S3McD1BCm76Ktz1EvRNCB4qYKGFU76fB7Fj'),
        farmingPoolStakeTknAccount: new web3_js_1.PublicKey('GhbQAXhtKt31BTCLdkP92CJJRJ1wDoRnjuSBbVoz4AV2'),
        farmingPoolRewardsTknMint: new web3_js_1.PublicKey('Hjibp1cn2bSk1dkTdpbxez3YAiBGTLjzc8xZ8LbCCUHS'),
        farmingPoolRewardsTknAccount: new web3_js_1.PublicKey('FjFMXsVPA4KJxCx3ARjuTLJH1c6UwMvSHXLVZD5UcnR3'),
        farmingPoolRewardsTknMintB: new web3_js_1.PublicKey('Hjibp1cn2bSk1dkTdpbxez3YAiBGTLjzc8xZ8LbCCUHS'),
        farmingPoolRewardsTknAccountB: new web3_js_1.PublicKey('AG3sYd5HhxeWQ2CSzZsrK6XF6Zg9UXZh4uGUMgDW8VZp'),
    },
    ATLAS: {
        programId: exports.lendRewardProgramId,
        farmingPoolAccount: new web3_js_1.PublicKey('DWSsuov4Jgp8d2h5f9vx142tmJMeKBqCwyzgp8YNpXUw'),
        farmingPoolAuthority: new web3_js_1.PublicKey('D6VwvNE8Nt4QcotuUVW1as51spadXNVBD9PXidZkvDWq'),
        farmingPoolStakeTknMint: new web3_js_1.PublicKey('4mygt5bFQrbXH9gNg75j1KVTrKGhvcYiQjir6FJ8afYH'),
        farmingPoolStakeTknAccount: new web3_js_1.PublicKey('2EFNPCYTTy6AoAgUmeQiNNgHRnmP2zQE7pWhH8vvnPCJ'),
        farmingPoolRewardsTknMint: new web3_js_1.PublicKey('Hjibp1cn2bSk1dkTdpbxez3YAiBGTLjzc8xZ8LbCCUHS'),
        farmingPoolRewardsTknAccount: new web3_js_1.PublicKey('DWsBfN9jYzXaJ7CiprCXMbzvSWrmRvWYWSBYU2b84oGe'),
        farmingPoolRewardsTknMintB: new web3_js_1.PublicKey('Hjibp1cn2bSk1dkTdpbxez3YAiBGTLjzc8xZ8LbCCUHS'),
        farmingPoolRewardsTknAccountB: new web3_js_1.PublicKey('4EGXGVXAwzw4dhQPdTb5qdg9Bas99fNVgt1ZC4nDTJvZ'),
    },
    GENE: {
        programId: exports.lendRewardProgramId,
        farmingPoolAccount: new web3_js_1.PublicKey('3EhxTvGjycQSKBY4EFz7MGA5Ke7rf39oUU2nM9qBP6Cj'),
        farmingPoolAuthority: new web3_js_1.PublicKey('4hqgkEYCFb736sSpxvrxD5Vvq1m7FdJKzfjinrr7xHSd'),
        farmingPoolStakeTknMint: new web3_js_1.PublicKey('9z3eDHueAMdUtym9Q2ku3hi5YXHTYjpFLp1YEEnxUHPV'),
        farmingPoolStakeTknAccount: new web3_js_1.PublicKey('DBrkDtmFWuXqpS8svgpXt2xLpQoJQQz5QUNeMWtnzh2A'),
        farmingPoolRewardsTknMint: new web3_js_1.PublicKey('Hjibp1cn2bSk1dkTdpbxez3YAiBGTLjzc8xZ8LbCCUHS'),
        farmingPoolRewardsTknAccount: new web3_js_1.PublicKey('EfTo3cTDREravKZQtcG7cYSHkLfbHP6TLfxFnFMcFYTE'),
        farmingPoolRewardsTknMintB: new web3_js_1.PublicKey('Hjibp1cn2bSk1dkTdpbxez3YAiBGTLjzc8xZ8LbCCUHS'),
        farmingPoolRewardsTknAccountB: new web3_js_1.PublicKey('5GhnJNhPU4FeAeX8zPjEGboSHWJcnu9JsqSzNDeYMb4q'),
    },
    DFL: {
        programId: exports.lendRewardProgramId,
        farmingPoolAccount: new web3_js_1.PublicKey('GiKoJhe5TN2DN2H9iW57P1ewtzZmi9vb8YhLMEVbweAH'),
        farmingPoolAuthority: new web3_js_1.PublicKey('Cbf8HLvkmch1HBb2XQf8LHhtPtECLrBDxsR3bCXZ5tA'),
        farmingPoolStakeTknMint: new web3_js_1.PublicKey('CGhMaGeVLxxQGJh6Y2bEYfLazumuFyEXpxF1UgrcMRJi'),
        farmingPoolStakeTknAccount: new web3_js_1.PublicKey('bMcNaYHuVVCL6PFqXGECDD2EgKSKUeVNzZcEKYUMe3T'),
        farmingPoolRewardsTknMint: new web3_js_1.PublicKey('Hjibp1cn2bSk1dkTdpbxez3YAiBGTLjzc8xZ8LbCCUHS'),
        farmingPoolRewardsTknAccount: new web3_js_1.PublicKey('7DzVmRce9QgooSrLpN1khQXtYv1AqUncc3BoBWvQi1ZV'),
        farmingPoolRewardsTknMintB: new web3_js_1.PublicKey('Hjibp1cn2bSk1dkTdpbxez3YAiBGTLjzc8xZ8LbCCUHS'),
        farmingPoolRewardsTknAccountB: new web3_js_1.PublicKey('3sZNEj3a6pkCVvZBSGYs9L4Komyd66vbcStckQSpZ32A'),
    },
    wbWBNB: {
        programId: exports.lendRewardProgramId,
        farmingPoolAccount: new web3_js_1.PublicKey('DyB3NGRQRAQcRowaxFhE8tBzB2kV6JcGCHJWADwVjFw9'),
        farmingPoolAuthority: new web3_js_1.PublicKey('FFVAs3tsHUjmGBKeHGex6b1bfyAfcJbwTPhrxtubMMDo'),
        farmingPoolStakeTknMint: new web3_js_1.PublicKey('8EzbLKBQ8Rest1SWUtZR6zFvYTo4cCaQ73ni8sF9vtjA'),
        farmingPoolStakeTknAccount: new web3_js_1.PublicKey('DuBFjRWMaNUbvfr2FgBNRuqMuRYuPcYZFRVFXHyR5wjx'),
        farmingPoolRewardsTknMint: new web3_js_1.PublicKey('Hjibp1cn2bSk1dkTdpbxez3YAiBGTLjzc8xZ8LbCCUHS'),
        farmingPoolRewardsTknAccount: new web3_js_1.PublicKey('FaUXYgAiMe3eSX11WyXZSx47Z6SoWnHvvMRPYR4WPXea'),
        farmingPoolRewardsTknMintB: new web3_js_1.PublicKey('Hjibp1cn2bSk1dkTdpbxez3YAiBGTLjzc8xZ8LbCCUHS'),
        farmingPoolRewardsTknAccountB: new web3_js_1.PublicKey('6qvfR54nRwWMXSP18D2Y5qKnwmd6AVJ9TKdjJNmS7yGr'),
    },
    stSOL: {
        programId: exports.lendRewardProgramId,
        farmingPoolAccount: new web3_js_1.PublicKey('J7Ny8VBzKHqVxbi8RAeNHatxELF61KhPCeV7S6m3AVa3'),
        farmingPoolAuthority: new web3_js_1.PublicKey('2q2cWnsuSBr8AukmhPj6memCY4qEqaajYfpjQhofGhYP'),
        farmingPoolStakeTknMint: new web3_js_1.PublicKey('HyxfDg47HKS1rQUXWvJH2XQhwEJM3AqrHU7sK9bSNt5h'),
        farmingPoolStakeTknAccount: new web3_js_1.PublicKey('8Ki5bPSCjEMxDdzjHNmZyowtSEXFjb5YXEPCysWsudo'),
        farmingPoolRewardsTknMint: new web3_js_1.PublicKey("HZRCwxP2Vq9PCpPXooayhJ2bxTpo5xfpQrwB1svh332p"),
        farmingPoolRewardsTknAccount: new web3_js_1.PublicKey("9TvniKw9GPczkYZLA5Ba43LmvkfgbRmuPJnVMJwGxgPb"),
        farmingPoolRewardsTknMintB: new web3_js_1.PublicKey('Hjibp1cn2bSk1dkTdpbxez3YAiBGTLjzc8xZ8LbCCUHS'),
        farmingPoolRewardsTknAccountB: new web3_js_1.PublicKey('AXbZA7hRj7uyNUVo6Bo2UYSdTET4V6Lck8guWYudW85d'),
    },
    SHDW: {
        programId: exports.lendRewardProgramId,
        farmingPoolAccount: new web3_js_1.PublicKey('Qy66gmV7hqvTG8FtwfLjbYaKrxTD953xMWw2tpLfyHs'),
        farmingPoolAuthority: new web3_js_1.PublicKey('3b7sreFukzQB7bWBVcFrQgxxoCiD34HPKupzh7gNrRi3'),
        farmingPoolStakeTknMint: new web3_js_1.PublicKey('AJu1s738dGsZ8mV2XKTqjuMiAiqNGaRsixTR4Czx4mJ2'),
        farmingPoolStakeTknAccount: new web3_js_1.PublicKey('FQo71LhxWHmjNiLoV19RDcoFS4gTMnHGRe6NixjgjsAp'),
        farmingPoolRewardsTknMint: new web3_js_1.PublicKey('Hjibp1cn2bSk1dkTdpbxez3YAiBGTLjzc8xZ8LbCCUHS'),
        farmingPoolRewardsTknAccount: new web3_js_1.PublicKey('7Q7gufPTfvUkTuVTFsEZAnL1B8gAFzsUgVD2bSYimG51'),
        farmingPoolRewardsTknMintB: new web3_js_1.PublicKey('Hjibp1cn2bSk1dkTdpbxez3YAiBGTLjzc8xZ8LbCCUHS'),
        farmingPoolRewardsTknAccountB: new web3_js_1.PublicKey('DVAPvhBkHEWhE4w9T84KF5rySu7cyQiFHLfsKDMvcT7m'),
    },
    BASIS: {
        programId: exports.lendRewardProgramId,
        farmingPoolAccount: new web3_js_1.PublicKey('GD2syrTphVNFhgWPfX52JxsqTLFXpFKHsFuHFuyH6SCa'),
        farmingPoolAuthority: new web3_js_1.PublicKey('9SMowoWcvciMBVTD4K3fDhMqjTQ5CRHCpGpT15R6sjBx'),
        farmingPoolStakeTknMint: new web3_js_1.PublicKey('9s2iRZBzSNCNTUtuzSRHcv3q4Q9SaFFhSFZVnFXGtHUp'),
        farmingPoolStakeTknAccount: new web3_js_1.PublicKey('9T6pP9xrEa3sNoaiDv2rRoViKWehnYv11Mpy3KAPqp25'),
        farmingPoolRewardsTknMint: new web3_js_1.PublicKey('Hjibp1cn2bSk1dkTdpbxez3YAiBGTLjzc8xZ8LbCCUHS'),
        farmingPoolRewardsTknAccount: new web3_js_1.PublicKey('3ons5X9YkRgccyEX7CVC9FSCeTzkkKrDouyAejZNeXfN'),
        farmingPoolRewardsTknMintB: new web3_js_1.PublicKey('Hjibp1cn2bSk1dkTdpbxez3YAiBGTLjzc8xZ8LbCCUHS'),
        farmingPoolRewardsTknAccountB: new web3_js_1.PublicKey('AAYSvVfYigFoSZHYgrZ78sviYsSpLGuP9gzuTkMQCQNq'),
    },
    wUST: {
        programId: exports.lendRewardProgramId,
        farmingPoolAccount: new web3_js_1.PublicKey('6Y5Y3mc6byPdXLhb8twQM2Ei4L3YfceJpLFzFDNsZSBi'),
        farmingPoolAuthority: new web3_js_1.PublicKey('9a3QEmKhwyzqPY3SCXSmoTBfEgyzb4w96VP44XDM9p3V'),
        farmingPoolStakeTknMint: new web3_js_1.PublicKey('FPnFwW1ASFLBReVz1EPWVGkbrsnrekVkEcQnrBRkvDXS'),
        farmingPoolStakeTknAccount: new web3_js_1.PublicKey('9fZhscUjooAhovtGDn7xvtYabs35aqj8uN1H45ERZiQa'),
        farmingPoolRewardsTknMint: new web3_js_1.PublicKey('Hjibp1cn2bSk1dkTdpbxez3YAiBGTLjzc8xZ8LbCCUHS'),
        farmingPoolRewardsTknAccount: new web3_js_1.PublicKey('BTgLyeGfjYzA3joa62k78wvsEE3a8eFHhzVoLMTDVPhM'),
        farmingPoolRewardsTknMintB: new web3_js_1.PublicKey('Hjibp1cn2bSk1dkTdpbxez3YAiBGTLjzc8xZ8LbCCUHS'),
        farmingPoolRewardsTknAccountB: new web3_js_1.PublicKey('CqsbNDh1im9bJx8rFUCdi72o26sZFNrbvHET3i7mFkzj'),
    },
    GST: {
        programId: exports.lendRewardProgramId,
        farmingPoolAccount: new web3_js_1.PublicKey('FeWUu6GVAvbpvcYqxymDPLsE4bWjtfLsJNg9FFVjQHwb'),
        farmingPoolAuthority: new web3_js_1.PublicKey('AqTdNkV6ZcpgN3QyBH1cSQZx8DHm6AvoHWusrGT9cPiw'),
        farmingPoolStakeTknMint: new web3_js_1.PublicKey('4GDuSZwHUQvL6zga2UNdjgXSqXjcWaUYc4S36N9q4567'),
        farmingPoolStakeTknAccount: new web3_js_1.PublicKey('A2d7G4aNKt4emJHdKpz7fb1C8GCw6gZxA5d9pUF1hxEW'),
        farmingPoolRewardsTknMint: new web3_js_1.PublicKey('Hjibp1cn2bSk1dkTdpbxez3YAiBGTLjzc8xZ8LbCCUHS'),
        farmingPoolRewardsTknAccount: new web3_js_1.PublicKey('BVRKUs86o7b1tiyGWByag6DtpejxZdvz9BkBRkZv1auh'),
        farmingPoolRewardsTknMintB: new web3_js_1.PublicKey('Hjibp1cn2bSk1dkTdpbxez3YAiBGTLjzc8xZ8LbCCUHS'),
        farmingPoolRewardsTknAccountB: new web3_js_1.PublicKey('E81P1ip8ynf8q9Y3d365fghEhNgHK5yPcYkiuTF7WjVs'),
    },
    GMT: {
        programId: exports.lendRewardProgramId,
        farmingPoolAccount: new web3_js_1.PublicKey('BvZ4vUfiXRvxcKV6ZiPcZdWJRtRpovMV19nZ1S8WV7JU'),
        farmingPoolAuthority: new web3_js_1.PublicKey('5G8BjukrA9tvU3dsafR95wywdj8vq44pg8TcMSbN1r2z'),
        farmingPoolStakeTknMint: new web3_js_1.PublicKey('9R6toP3xTFBWoqwBY8bfjUj6Tyd5hyR192jY4NeMmkNg'),
        farmingPoolStakeTknAccount: new web3_js_1.PublicKey('DPXP9bkB5A183bfbpHrCYdXpRSEFehWqF4PhpG1P5NtJ'),
        farmingPoolRewardsTknMint: new web3_js_1.PublicKey('Hjibp1cn2bSk1dkTdpbxez3YAiBGTLjzc8xZ8LbCCUHS'),
        farmingPoolRewardsTknAccount: new web3_js_1.PublicKey('2Bo8GWQoysSTswVPxsMdCFC8fmVP84D7CJPuqfSiKPJd'),
        farmingPoolRewardsTknMintB: new web3_js_1.PublicKey('Hjibp1cn2bSk1dkTdpbxez3YAiBGTLjzc8xZ8LbCCUHS'),
        farmingPoolRewardsTknAccountB: new web3_js_1.PublicKey('29TsFVuj7ZdFwLUvfBeBt7Wcw5jBGguGi8LGRau9TRaz'),
    },
    ZBC: {
        programId: exports.lendRewardProgramId,
        farmingPoolAccount: new web3_js_1.PublicKey('FPKSH61ciN4YetqKvW7KMWVc1icKDqtPogQR1tQqZrWF'),
        farmingPoolAuthority: new web3_js_1.PublicKey('4v2FpU7sTv8BoQdu9LXY65EK4AMHheuDyVDp3viuzh5f'),
        farmingPoolStakeTknMint: new web3_js_1.PublicKey('HUry6RSbDasMfU7LXZ5QR42YajkbMuLBQkyV742xgFKx'),
        farmingPoolStakeTknAccount: new web3_js_1.PublicKey('4TMJwumst5n67zuFFj12zBVe1U4WX1iApjni6FMNPrwP'),
        farmingPoolRewardsTknMint: new web3_js_1.PublicKey('Hjibp1cn2bSk1dkTdpbxez3YAiBGTLjzc8xZ8LbCCUHS'),
        farmingPoolRewardsTknAccount: new web3_js_1.PublicKey('33qA8PTrZuzMe1ixw1ABeSCDykGEcaBRcDbYcySjY66E'),
        farmingPoolRewardsTknMintB: new web3_js_1.PublicKey('Hjibp1cn2bSk1dkTdpbxez3YAiBGTLjzc8xZ8LbCCUHS'),
        farmingPoolRewardsTknAccountB: new web3_js_1.PublicKey('8CHYivAkcJo8V56bcVaBmKQjLAydtLc9aXiyiyWCfk1U'),
    }
};
