// @mercurial-finance/dynamic-amm-sdk/dist/cjs/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.VaultIdl = exports.AmmIdl = exports.DEVNET_POOL = exports.MAINNET_POOL = exports.PROGRAM_ID = exports.calculatePoolInfo = exports.calculateSwapQuote = exports.calculateMaxSwapOutAmount = exports.getOnchainTime = exports.getDepegAccounts = exports.StableSwap = exports.ConstantProductSwap = void 0;
const amm_1 = __importDefault(require("./src/amm"));
const curve_1 = require("./src/amm/curve");
Object.defineProperty(exports, "StableSwap", { enumerable: true, get: function () { return curve_1.StableSwap; } });
Object.defineProperty(exports, "ConstantProductSwap", { enumerable: true, get: function () { return curve_1.ConstantProductSwap; } });
const constants_1 = require("./src/amm/constants");
Object.defineProperty(exports, "PROGRAM_ID", { enumerable: true, get: function () { return constants_1.PROGRAM_ID; } });
Object.defineProperty(exports, "MAINNET_POOL", { enumerable: true, get: function () { return constants_1.MAINNET_POOL; } });
Object.defineProperty(exports, "DEVNET_POOL", { enumerable: true, get: function () { return constants_1.DEVNET_POOL; } });
const utils_1 = require("./src/amm/utils");
Object.defineProperty(exports, "getOnchainTime", { enumerable: true, get: function () { return utils_1.getOnchainTime; } });
Object.defineProperty(exports, "calculateMaxSwapOutAmount", { enumerable: true, get: function () { return utils_1.calculateMaxSwapOutAmount; } });
Object.defineProperty(exports, "calculateSwapQuote", { enumerable: true, get: function () { return utils_1.calculateSwapQuote; } });
Object.defineProperty(exports, "calculatePoolInfo", { enumerable: true, get: function () { return utils_1.calculatePoolInfo; } });
Object.defineProperty(exports, "getDepegAccounts", { enumerable: true, get: function () { return utils_1.getDepegAccounts; } });
const idl_1 = require("./src/amm/idl");
Object.defineProperty(exports, "AmmIdl", { enumerable: true, get: function () { return idl_1.IDL; } });
const vault_idl_1 = require("./src/amm/vault-idl");
Object.defineProperty(exports, "VaultIdl", { enumerable: true, get: function () { return vault_idl_1.IDL; } });
exports.default = amm_1.default;
