// @port.finance/port-sdk/dist/esm/Environment.js
import { PublicKey } from "@solana/web3.js";
import { AssetContext } from "./models";
import { PORT_LENDING, PORT_STAKING } from "./constants";
import { MAINNET_ASSETS } from "./utils/AssetConfigs";
import { ENV } from "./utils/env";
export class Environment {
    constructor(env, lendingProgramPk, stakingProgramPk, tokenProgramPk, assetConfigs) {
        this.env = env;
        this.lendingProgramPk = lendingProgramPk;
        this.stakingProgramPk = stakingProgramPk;
        this.tokenProgramPk = tokenProgramPk;
        this.assetContext = AssetContext.index(assetConfigs);
    }
    static forMainNet() {
        return new Environment(ENV.MainnetBeta, PORT_LENDING, PORT_STAKING, new PublicKey("TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA"), MAINNET_ASSETS);
    }
    getEnv() {
        return this.env;
    }
    getLendingProgramPk() {
        return this.lendingProgramPk;
    }
    getStakingProgramPk() {
        return this.stakingProgramPk;
    }
    getTokenProgramPk() {
        return this.tokenProgramPk;
    }
    getAssetContext() {
        return this.assetContext;
    }
}
