// @port.finance/port-sdk/dist/esm/Port.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { Connection, Keypair, PublicKey, SystemProgram } from "@solana/web3.js";
import { RESERVE_DATA_SIZE, ReserveLayout, STAKING_POOL_DATA_SIZE, PORT_PROFILE_DATA_SIZE, ObligationLayout, } from "./structs";
import { ReserveInfo, ReserveContext, QuoteValue, TokenAccount, PortProfile, StakingPool, StakingPoolContext, } from "./models";
import { Environment } from "./Environment";
import { DEFAULT_PORT_LENDING_MARKET } from "./constants";
import { AccessType } from "./utils/Instructions";
import { BN } from "@project-serum/anchor";
import { initLendingMarketInstruction, initReserveInstruction, PORT_LENDING, } from ".";
import { AccountLayout, MintLayout, TOKEN_PROGRAM_ID } from "@solana/spl-token";
import { TransactionEnvelope } from "@saberhq/solana-contrib";
import { getTokenAccount } from "@saberhq/token-utils";
import invariant from "tiny-invariant";
export const LENDING_MARKET_LEN = 258;
export class Port {
    constructor(connection, environment, lendingMarket) {
        this.connection = connection;
        this.environment = environment;
        this.lendingMarket = lendingMarket;
    }
    setConnection(connection) {
        this.connection = connection;
    }
    setLendingMarket(lendingMarket) {
        this.lendingMarket = lendingMarket;
    }
    static forMainNet({ connection = new Connection("https://api.mainnet-beta.solana.com"), profile = Environment.forMainNet(), lendingMarket = DEFAULT_PORT_LENDING_MARKET, }) {
        return new Port(connection, profile, lendingMarket);
    }
    getEnvironment() {
        return this.environment;
    }
    load() {
        return __awaiter(this, void 0, void 0, function* () {
            this.reserveContext = yield this.getReserveContext();
        });
    }
    getTotalMarketCap() {
        return __awaiter(this, void 0, void 0, function* () {
            const context = yield this.getReserveContext();
            return context
                .getAllReserves()
                .map((r) => r.getMarketCap())
                .map((c) => c.getValue())
                .reduce(QuoteValue.sum, QuoteValue.zero());
        });
    }
    getShareAccount(walletId, context) {
        return __awaiter(this, void 0, void 0, function* () {
            const shareMintPks = context
                .getAllReserves()
                .map((r) => r.getShareMintId())
                .map((s) => s.getAccess(AccessType.READ).pubkey);
            const programId = this.environment.getTokenProgramPk();
            const result = yield this.connection.getTokenAccountsByOwner(walletId.getAccess(AccessType.READ).pubkey, {
                programId,
            });
            const raw = result.value;
            return raw
                .map((a) => TokenAccount.fromRaw(a))
                .filter((p) => p &&
                shareMintPks.find((k) => k.equals(p.getMintId().getAccess(AccessType.READ).pubkey)));
        });
    }
    getPortProfile(walletId) {
        return __awaiter(this, void 0, void 0, function* () {
            const raw = yield this.connection.getProgramAccounts(this.environment.getLendingProgramPk(), {
                filters: [
                    {
                        memcmp: {
                            // eslint-disable-next-line
                            offset: ObligationLayout.offsetOf("owner"),
                            bytes: walletId.toBase58(),
                        },
                    },
                    {
                        memcmp: {
                            // eslint-disable-next-line
                            offset: ObligationLayout.offsetOf("lendingMarket"),
                            bytes: this.lendingMarket.toBase58(),
                        },
                    },
                    {
                        dataSize: PORT_PROFILE_DATA_SIZE,
                    },
                ],
            });
            const parsed = raw.map((a) => PortProfile.fromRaw(a)).filter((p) => !!p);
            return parsed.length > 0 ? parsed[0] : undefined;
        });
    }
    getReserveContext() {
        return __awaiter(this, void 0, void 0, function* () {
            const raw = yield this.connection.getProgramAccounts(this.environment.getLendingProgramPk(), {
                filters: [
                    {
                        dataSize: RESERVE_DATA_SIZE,
                    },
                    {
                        memcmp: {
                            // eslint-disable-next-line
                            offset: ReserveLayout.offsetOf("lendingMarket"),
                            bytes: this.lendingMarket.toBase58(),
                        },
                    },
                ],
            });
            const parsed = raw.map((a) => ReserveInfo.fromRaw(a)).filter((p) => !!p);
            return ReserveContext.index(parsed);
        });
    }
    getStakingPoolContext() {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.environment.getStakingProgramPk() === undefined) {
                Promise.resolve();
            }
            const raw = yield this.connection.getProgramAccounts(
            // eslint-disable-next-line
            this.environment.getStakingProgramPk(), {
                filters: [
                    {
                        dataSize: STAKING_POOL_DATA_SIZE,
                    },
                ],
            });
            const parsed = raw.map((a) => StakingPool.fromRaw(a)).filter((p) => !!p);
            return StakingPoolContext.index(parsed);
        });
    }
    getAllPortProfiles() {
        return __awaiter(this, void 0, void 0, function* () {
            const raw = yield this.connection.getProgramAccounts(this.environment.getLendingProgramPk(), {
                filters: [
                    {
                        dataSize: PORT_PROFILE_DATA_SIZE,
                    },
                    {
                        memcmp: {
                            // eslint-disable-next-line
                            offset: ObligationLayout.offsetOf("lendingMarket"),
                            bytes: this.lendingMarket.toBase58(),
                        },
                    },
                ],
            });
            const parsed = raw.map((p) => PortProfile.fromRaw(p)).filter((p) => !!p);
            return parsed;
        });
    }
    getStakingPool(stakingPoolKey) {
        return __awaiter(this, void 0, void 0, function* () {
            const raw = yield this.connection.getAccountInfo(stakingPoolKey);
            if (!raw) {
                return Promise.reject(new Error("no reserve found"));
            }
            return StakingPool.fromRaw({
                pubkey: stakingPoolKey,
                account: raw,
            });
        });
    }
    getReserve(reserveKey) {
        return __awaiter(this, void 0, void 0, function* () {
            const raw = yield this.connection.getAccountInfo(reserveKey);
            if (!raw) {
                return Promise.reject(new Error("no reserve found"));
            }
            return ReserveInfo.fromRaw({
                pubkey: reserveKey,
                account: raw,
            });
        });
    }
    createLendingMarket({ provider, owner = provider.wallet.publicKey, }) {
        return __awaiter(this, void 0, void 0, function* () {
            let tx = new TransactionEnvelope(provider, []);
            const [createTx, lendingMarketPubkey] = yield this.createAccount({
                provider,
                space: LENDING_MARKET_LEN,
                owner: PORT_LENDING,
            });
            const createLendingMarketIx = initLendingMarketInstruction(owner, Buffer.from("USD\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0", "ascii"), lendingMarketPubkey);
            tx = tx.combine(createTx);
            tx.addInstructions(createLendingMarketIx);
            return [tx, lendingMarketPubkey];
        });
    }
    createReserve({ provider, reserveConfig, transferAuthority, sourceTokenWallet, initialLiquidity, oracle, price, }) {
        return __awaiter(this, void 0, void 0, function* () {
            invariant(!!oracle !== !!price, "Oracle and price can't both be present");
            const [createReserveAccountIx, reservePubKey] = yield this.createAccount({
                provider,
                space: ReserveLayout.span,
                owner: PORT_LENDING,
            });
            const [collateralMintIx, collateralMintPubKey] = yield this.createAccount({
                provider,
                space: MintLayout.span,
                owner: TOKEN_PROGRAM_ID,
            });
            const [liquiditySupplyIx, liquiditySupplyPubKey] = yield this.createAccount({
                provider,
                space: AccountLayout.span,
                owner: TOKEN_PROGRAM_ID,
            });
            const [collateralSupplyIx, collateralSupplyPubKey] = yield this.createAccount({
                provider,
                space: AccountLayout.span,
                owner: TOKEN_PROGRAM_ID,
            });
            const [userCollateralIx, userCollateralPubKey] = yield this.createAccount({
                provider,
                space: AccountLayout.span,
                owner: TOKEN_PROGRAM_ID,
            });
            const [feeReceiverIx, feeReceiverPubkey] = yield this.createAccount({
                provider,
                space: AccountLayout.span,
                owner: TOKEN_PROGRAM_ID,
            });
            const tokenAccount = yield getTokenAccount(provider, sourceTokenWallet);
            const initReserveIx = initReserveInstruction(initialLiquidity, oracle ? 0 : 1, // price Option
            price !== null && price !== void 0 ? price : new BN(1), reserveConfig, sourceTokenWallet, collateralSupplyPubKey, reservePubKey, tokenAccount.mint, liquiditySupplyPubKey, feeReceiverPubkey, oracle !== null && oracle !== void 0 ? oracle : Keypair.generate().publicKey, collateralMintPubKey, userCollateralPubKey, this.lendingMarket, (yield this.getLendingMarketAuthority())[0], provider.wallet.publicKey, transferAuthority);
            let tx1 = new TransactionEnvelope(provider, []);
            tx1 = tx1.combine(createReserveAccountIx);
            tx1 = tx1.combine(collateralMintIx);
            tx1 = tx1.combine(liquiditySupplyIx);
            tx1 = tx1.combine(collateralSupplyIx);
            tx1 = tx1.combine(userCollateralIx);
            let tx2 = new TransactionEnvelope(provider, []);
            tx2 = tx2.combine(feeReceiverIx);
            tx2 = tx2.addInstructions(initReserveIx);
            return [[tx1, tx2], reservePubKey];
        });
    }
    getLendingMarketAuthority() {
        return __awaiter(this, void 0, void 0, function* () {
            return yield PublicKey.findProgramAddress([this.lendingMarket.toBuffer()], PORT_LENDING);
        });
    }
    createAccount({ provider, space, owner, }) {
        return __awaiter(this, void 0, void 0, function* () {
            const newAccount = Keypair.generate();
            const tx = new TransactionEnvelope(provider, [
                SystemProgram.createAccount({
                    fromPubkey: provider.wallet.publicKey,
                    newAccountPubkey: newAccount.publicKey,
                    programId: owner,
                    lamports: yield provider.connection.getMinimumBalanceForRentExemption(space),
                    space,
                }),
            ], [newAccount]);
            return [tx, newAccount.publicKey];
        });
    }
}
