// francium-sdk/dist/constants/tokens/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TOKENS = void 0;
const spl_token_1 = require("@solana/spl-token");
const web3_js_1 = require("@solana/web3.js");
exports.TOKENS = {
    SOL: {
        symbol: 'SOL',
        name: 'solana',
        mintAddress: spl_token_1.NATIVE_MINT,
        decimals: 9
    },
    WSOL: {
        symbol: 'WSOL',
        name: 'Wrapped Solana',
        mintAddress: new web3_js_1.PublicKey('So11111111111111111111111111111111111111112'),
        decimals: 9
    },
    BTC: {
        symbol: 'BTC',
        name: 'bitcoin',
        mintAddress: new web3_js_1.PublicKey('9n4nbM75f5Ui33ZbPYXn59EwSgE8CGsHtAeTH5YFeJ9E'),
        decimals: 6
    },
    ETH: {
        symbol: 'ETH',
        name: 'ethereum',
        mintAddress: new web3_js_1.PublicKey('2FPyTwcZLUg1MDrwsyoP4D6s1tM7hAkHYRjkNb5w6Pxk'),
        decimals: 6
    },
    USDT: {
        symbol: 'USDT',
        name: 'USDT',
        mintAddress: new web3_js_1.PublicKey('Es9vMFrzaCERmJfrF4H2FYD4KCoNkY11McCe8BenwNYB'),
        decimals: 6
    },
    WUSDT: {
        symbol: 'WUSDT',
        name: 'Wrapped USDT',
        mintAddress: new web3_js_1.PublicKey('BQcdHdAQW1hczDbBi9hiegXAR7A98Q9jx3X3iBBBDiq4'),
        decimals: 6
    },
    USDC: {
        symbol: 'USDC',
        name: 'USDC',
        mintAddress: new web3_js_1.PublicKey('EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v'),
        decimals: 6
    },
    WUSDC: {
        symbol: 'WUSDC',
        name: 'Wrapped USDC',
        mintAddress: new web3_js_1.PublicKey('BXXkv6z8ykpG1yuvUDPgh732wzVHB69RnB9YgSYh3itW'),
        decimals: 6
    },
    YFI: {
        symbol: 'YFI',
        name: 'Wrapped YFI',
        mintAddress: new web3_js_1.PublicKey('3JSf5tPeuscJGtaCp5giEiDhv51gQ4v3zWg8DGgyLfAB'),
        decimals: 6
    },
    LINK: {
        symbol: 'LINK',
        name: 'Wrapped Chainlink',
        mintAddress: new web3_js_1.PublicKey('CWE8jPTUYhdCTZYWPTe1o5DFqfdjzWKc9WKz6rSjQUdG'),
        decimals: 6
    },
    ALEPH: {
        symbol: 'ALEPH',
        name: 'Wrapped ALEPH',
        mintAddress: new web3_js_1.PublicKey('CsZ5LZkDS7h9TDKjrbL7VAwQZ9nsRu8vJLhRYfmGaN8K'),
        decimals: 6
    },
    SXP: {
        symbol: 'SXP',
        name: 'Wrapped SXP',
        mintAddress: new web3_js_1.PublicKey('SF3oTvfWzEP3DTwGSvUXRrGTvr75pdZNnBLAH9bzMuX'),
        decimals: 6
    },
    HGET: {
        symbol: 'HGET',
        name: 'Wrapped HGET',
        mintAddress: new web3_js_1.PublicKey('BtZQfWqDGbk9Wf2rXEiWyQBdBY1etnUUn6zEphvVS7yN'),
        decimals: 6
    },
    CREAM: {
        symbol: 'CREAM',
        name: 'Wrapped CREAM',
        mintAddress: new web3_js_1.PublicKey('5Fu5UUgbjpUvdBveb3a1JTNirL8rXtiYeSMWvKjtUNQv'),
        decimals: 6
    },
    UBXT: {
        symbol: 'UBXT',
        name: 'Wrapped UBXT',
        mintAddress: new web3_js_1.PublicKey('873KLxCbz7s9Kc4ZzgYRtNmhfkQrhfyWGZJBmyCbC3ei'),
        decimals: 6
    },
    HNT: {
        symbol: 'HNT',
        name: 'Wrapped HNT',
        mintAddress: new web3_js_1.PublicKey('HqB7uswoVg4suaQiDP3wjxob1G5WdZ144zhdStwMCq7e'),
        decimals: 6
    },
    FRONT: {
        symbol: 'FRONT',
        name: 'Wrapped FRONT',
        mintAddress: new web3_js_1.PublicKey('9S4t2NEAiJVMvPdRYKVrfJpBafPBLtvbvyS3DecojQHw'),
        decimals: 6
    },
    AKRO: {
        symbol: 'AKRO',
        name: 'Wrapped AKRO',
        mintAddress: new web3_js_1.PublicKey('6WNVCuxCGJzNjmMZoKyhZJwvJ5tYpsLyAtagzYASqBoF'),
        decimals: 6
    },
    HXRO: {
        symbol: 'HXRO',
        name: 'Wrapped HXRO',
        mintAddress: new web3_js_1.PublicKey('DJafV9qemGp7mLMEn5wrfqaFwxsbLgUsGVS16zKRk9kc'),
        decimals: 6
    },
    UNI: {
        symbol: 'UNI',
        name: 'Wrapped UNI',
        mintAddress: new web3_js_1.PublicKey('DEhAasscXF4kEGxFgJ3bq4PpVGp5wyUxMRvn6TzGVHaw'),
        decimals: 6
    },
    SRM: {
        symbol: 'SRM',
        name: 'serum',
        mintAddress: new web3_js_1.PublicKey('SRMuApVNdxXokk5GT7XD5cUUgXMBCoAz2LHeuAoKWRt'),
        decimals: 6
    },
    FTT: {
        symbol: 'FTT',
        name: 'Wrapped FTT',
        mintAddress: new web3_js_1.PublicKey('AGFEad2et2ZJif9jaGpdMixQqvW5i81aBdvKe7PHNfz3'),
        decimals: 6
    },
    TOMO: {
        symbol: 'TOMO',
        name: 'Wrapped TOMO',
        mintAddress: new web3_js_1.PublicKey('GXMvfY2jpQctDqZ9RoU3oWPhufKiCcFEfchvYumtX7jd'),
        decimals: 6
    },
    KARMA: {
        symbol: 'KARMA',
        name: 'Wrapped KARMA',
        mintAddress: new web3_js_1.PublicKey('EcqExpGNFBve2i1cMJUTR4bPXj4ZoqmDD2rTkeCcaTFX'),
        decimals: 4
    },
    LUA: {
        symbol: 'LUA',
        name: 'Wrapped LUA',
        mintAddress: new web3_js_1.PublicKey('EqWCKXfs3x47uVosDpTRgFniThL9Y8iCztJaapxbEaVX'),
        decimals: 6
    },
    MATH: {
        symbol: 'MATH',
        name: 'Wrapped MATH',
        mintAddress: new web3_js_1.PublicKey('GeDS162t9yGJuLEHPWXXGrb1zwkzinCgRwnT8vHYjKza'),
        decimals: 6
    },
    KEEP: {
        symbol: 'KEEP',
        name: 'Wrapped KEEP',
        mintAddress: new web3_js_1.PublicKey('GUohe4DJUA5FKPWo3joiPgsB7yzer7LpDmt1Vhzy3Zht'),
        decimals: 6
    },
    SWAG: {
        symbol: 'SWAG',
        name: 'Wrapped SWAG',
        mintAddress: new web3_js_1.PublicKey('9F9fNTT6qwjsu4X4yWYKZpsbw5qT7o6yR2i57JF2jagy'),
        decimals: 6
    },
    FIDA: {
        symbol: 'FIDA',
        name: 'Bonfida',
        mintAddress: new web3_js_1.PublicKey('EchesyfXePKdLtoiZSL8pBe8Myagyy8ZRqsACNCFGnvp'),
        decimals: 6
    },
    KIN: {
        symbol: 'KIN',
        name: 'KIN',
        mintAddress: new web3_js_1.PublicKey('kinXdEcpDQeHPEuQnqmUgtYykqKGVFq6CeVX5iAHJq6'),
        decimals: 5
    },
    MAPS: {
        symbol: 'MAPS',
        name: 'MAPS',
        mintAddress: new web3_js_1.PublicKey('MAPS41MDahZ9QdKXhVa4dWB9RuyfV4XqhyAZ8XcYepb'),
        decimals: 6
    },
    OXY: {
        symbol: 'OXY',
        name: 'OXY',
        mintAddress: new web3_js_1.PublicKey('z3dn17yLaGMKffVogeFHQ9zWVcXgqgf3PQnDsNs2g6M'),
        decimals: 6
    },
    RAY: {
        symbol: 'RAY',
        name: 'raydium',
        mintAddress: new web3_js_1.PublicKey('4k3Dyjzvzp8eMZWUXbBCjEvwSkkk59S5iCNLY3QrkX6R'),
        decimals: 6
    },
    COPE: {
        symbol: 'COPE',
        name: 'cope',
        mintAddress: new web3_js_1.PublicKey('8HGyAAB1yoM1ttS7pXjHMa3dukTFGQggnFFH3hJZgzQh'),
        decimals: 6,
    },
    STEP: {
        symbol: 'STEP',
        name: 'step-finance',
        mintAddress: new web3_js_1.PublicKey('StepAscQoEioFxxWGnh2sLBDFp9d8rvKz2Yp39iDpyT'),
        decimals: 9,
    },
    MEDIA: {
        symbol: 'MEDIA',
        name: 'media-network',
        mintAddress: new web3_js_1.PublicKey('ETAtLmCmsoiEEKfNrHKJ2kYy3MoABhU6NQvpSfij5tDs'),
        decimals: 6
    },
    ROPE: {
        symbol: 'ROPE',
        name: 'ROPE',
        mintAddress: new web3_js_1.PublicKey('8PMHT4swUMtBzgHnh5U564N5sjPSiUz2cjEQzFnnP1Fo'),
        decimals: 9
    },
    MER: {
        symbol: 'MER',
        name: 'Mercurial',
        mintAddress: new web3_js_1.PublicKey('MERt85fc5boKw3BW1eYdxonEuJNvXbiMbs6hvheau5K'),
        decimals: 6,
    },
    LIKE: {
        symbol: 'LIKE',
        name: 'only1',
        mintAddress: new web3_js_1.PublicKey('3bRTivrVsitbmCTGtqwp7hxXPsybkjn4XLNtPsHqa3zR'),
        decimals: 9
    },
    ATLAS: {
        symbol: 'ATLAS',
        name: 'star-atlas',
        decimals: 8,
        mintAddress: new web3_js_1.PublicKey('ATLASXmbPQxBUYbxPsV97usA3fPQYEqzQBUHgiFCUsXx')
    },
    POLIS: {
        symbol: 'POLIS',
        name: 'star-atlas-dao',
        decimals: 8,
        mintAddress: new web3_js_1.PublicKey('poLisWXnNRwC6oBu1vHiuKQzFjGL4XDSu4g9qjz9qVk'),
    },
    ORCA: {
        symbol: 'ORCA',
        name: 'orca',
        decimals: 6,
        mintAddress: new web3_js_1.PublicKey('orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE')
    },
    mSOL: {
        symbol: 'mSOL',
        name: 'msol',
        decimals: 9,
        mintAddress: new web3_js_1.PublicKey('mSoLzYCxHdYgdzU16g5QSh3i5K3z3KZK7ytfqcJm7So')
    },
    MNDE: {
        symbol: 'MNDE',
        name: 'marinade',
        decimals: 9,
        mintAddress: new web3_js_1.PublicKey('MNDEFzGvMt87ueuHvVU9VcTqsAP5b3fTGPsHuuPA5ey')
    },
    NINJA: {
        symbol: 'NINJA',
        name: 'ninja-protocol',
        decimals: 6,
        mintAddress: new web3_js_1.PublicKey('FgX1WD9WzMU3yLwXaFSarPfkgzjLb2DZCqmkx9ExpuvJ')
    },
    whETH: {
        symbol: 'whETH',
        name: 'ethereum',
        mintAddress: new web3_js_1.PublicKey('7vfCXTUXx5WJV5JADk17DUJ4ksgau7utNKj4b963voxs'),
        decimals: 8
    },
    SBR: {
        symbol: 'SBR',
        name: 'saber',
        mintAddress: new web3_js_1.PublicKey('Saber2gLauYim4Mvftnrasomsv6NvAuncvMEZwcLpD1'),
        decimals: 6
    },
    LIQ: {
        symbol: 'LIQ',
        name: 'liq-protocol',
        mintAddress: new web3_js_1.PublicKey('4wjPQJ6PrkC4dHhYghwJzGBVP78DkBzA2U3kHoFNBuhj'),
        decimals: 6
    },
    PORT: {
        symbol: 'PORT',
        name: 'port-finance',
        mintAddress: new web3_js_1.PublicKey('PoRTjZMPXb9T7dyU7tpLEZRQj7e6ssfAE62j2oQuc6y'),
        decimals: 6
    },
    SAMO: {
        symbol: 'SAMO',
        name: 'samoyedcoin',
        mintAddress: new web3_js_1.PublicKey('7xKXtg2CW87d97TXJSDpbD5jBkheTqA83TZRuJosgAsU'),
        decimals: 9
    },
    ABR: {
        symbol: 'ABR',
        name: 'allbridge',
        mintAddress: new web3_js_1.PublicKey('a11bdAAuV8iB2fu7X6AxAvDTo1QZ8FXB3kk5eecdasp'),
        decimals: 9
    },
    weSUSHI: {
        symbol: 'weSUSHI',
        name: 'allbridge',
        mintAddress: new web3_js_1.PublicKey('ChVzxWRmrTeSgwd3Ui3UumcN8KX7VK3WaD4KGeSKpypj'),
        decimals: 8
    },
    weUNI: {
        symbol: 'weUNI',
        name: 'allbridge',
        mintAddress: new web3_js_1.PublicKey('8FU95xFJhUUkyyCLU13HSzDLs7oC4QZdXQHL6SCeab36'),
        decimals: 8
    },
    STARS: {
        symbol: 'STARS',
        name: 'starlaunch',
        mintAddress: new web3_js_1.PublicKey('HCgybxq5Upy8Mccihrp7EsmwwFqYZtrHrsmsKwtGXLgW'),
        decimals: 6
    },
    weDYDX: {
        symbol: 'weDYDX',
        name: 'dydx',
        mintAddress: new web3_js_1.PublicKey('4Hx6Bj56eGyw8EJrrheM6LBQAvVYRikYCWsALeTrwyRU'),
        decimals: 8
    },
    weAXS: {
        symbol: 'weAXS',
        name: 'axie-infinity',
        mintAddress: new web3_js_1.PublicKey('HysWcbHiYY9888pHbaqhwLYZQeZrcQMXKQWRqS7zcPK5'),
        decimals: 8
    },
    weSHIB: {
        symbol: 'weSHIB',
        name: 'shiba-inu',
        mintAddress: new web3_js_1.PublicKey('CiKu4eHsVrc1eueVQeHn7qhXTcVu95gSQmBpX4utjL9z'),
        decimals: 8
    },
    SLND: {
        symbol: 'SLND',
        name: 'solend',
        mintAddress: new web3_js_1.PublicKey('SLNDpmoWTVADgEdndyvWzroNL7zSi1dF9PC3xHGtPwp'),
        decimals: 6
    },
    weMANA: {
        symbol: 'weMANA',
        name: 'decentraland',
        mintAddress: new web3_js_1.PublicKey('7dgHoN8wBZCc5wbnQ2C47TDnBMAxG4Q5L3KjP67z8kNi'),
        decimals: 8
    },
    weSAND: {
        symbol: 'weSAND',
        name: 'the-sandbox',
        mintAddress: new web3_js_1.PublicKey('49c7WuCZkQgc3M4qH8WuEUNXfgwupZf1xqWkDQ7gjRGt'),
        decimals: 8
    },
    CAVE: {
        symbol: 'CAVE',
        name: 'cave',
        mintAddress: new web3_js_1.PublicKey('4SZjjNABoqhbd4hnapbvoEPEqT8mnNkfbEoAwALf1V8t'),
        decimals: 6
    },
    GENE: {
        symbol: 'GENE',
        name: 'genopets',
        mintAddress: new web3_js_1.PublicKey('GENEtH5amGSi8kHAtQoezp1XEXwZJ8vcuePYnXdKrMYz'),
        decimals: 9
    },
    SONAR: {
        symbol: 'SONAR',
        name: 'sonarwatch',
        mintAddress: new web3_js_1.PublicKey('sonarX4VtVkQemriJeLm6CKeW3GDMyiBnnAEMw1MRAE'),
        decimals: 9
    },
    DFL: {
        symbol: 'DFL',
        name: 'defi-land',
        mintAddress: new web3_js_1.PublicKey('DFL1zNkaGPWm1BqAVqRjCZvHmwTFrEaJtbzJWgseoNJh'),
        decimals: 9
    },
    wbWBNB: {
        symbol: 'wbWBNB',
        name: 'binancecoin',
        mintAddress: new web3_js_1.PublicKey('9gP2kCy3wA1ctvYWQk75guqXuHfrEomqydHLtcTCqiLa'),
        decimals: 8
    },
    REAL: {
        symbol: 'REAL',
        name: 'realy-metaverse',
        mintAddress: new web3_js_1.PublicKey('AD27ov5fVU2XzwsbvnFvb1JpCBaCB5dRXrczV9CqSVGb'),
        decimals: 9
    },
    MBS: {
        symbol: 'MBS',
        name: 'monkeyball',
        mintAddress: new web3_js_1.PublicKey('Fm9rHUTF5v3hwMLbStjZXqNBBoZyGriQaFM6sTFz3K8A'),
        decimals: 6
    },
    stSOL: {
        symbol: 'stSOL',
        name: 'lido-staked-sol',
        mintAddress: new web3_js_1.PublicKey('7dHbWXmci3dT8UFYWYZweBLXgycu7Y3iL6trKn1Y7ARj'),
        decimals: 9
    },
    wLDO: {
        symbol: 'wLDO',
        name: 'lido-dao',
        mintAddress: new web3_js_1.PublicKey('HZRCwxP2Vq9PCpPXooayhJ2bxTpo5xfpQrwB1svh332p'),
        decimals: 8
    },
    PRISM: {
        symbol: 'PRISM',
        name: 'prism',
        mintAddress: new web3_js_1.PublicKey('PRSMNsEPqhGVCH1TtWiJqPjJyh2cKrLostPZTNy1o5x'),
        decimals: 6
    },
    SHDW: {
        symbol: 'SHDW',
        name: 'genesysgo-shadow',
        mintAddress: new web3_js_1.PublicKey('SHDWyBxihqiCj6YekG2GUr7wqKLeLAMK1gHZck9pL6y'),
        decimals: 9
    },
    BASIS: {
        symbol: 'BASIS',
        name: 'basis-markets',
        mintAddress: new web3_js_1.PublicKey('Basis9oJw9j8cw53oMV7iqsgo6ihi9ALw4QR31rcjUJa'),
        decimals: 6
    },
    GRAPE: {
        symbol: 'GRAPE',
        name: 'grape-2',
        mintAddress: new web3_js_1.PublicKey('8upjSpvjcdpuzhfR1zriwg5NXkwDruejqNE9WNbPRtyA'),
        decimals: 6
    },
    wUST: {
        symbol: 'wUST',
        name: 'wUST',
        mintAddress: new web3_js_1.PublicKey('9vMJfxuKxXBoEa7rM12mYLMwTacLMLDJqHozw96WQL8i'),
        decimals: 6
    },
    wLUNA: {
        symbol: 'wLUNA',
        name: 'terra-luna',
        mintAddress: new web3_js_1.PublicKey('F6v4wfAdJB8D8p77bMXZgYt8TDKsYxLYxH5AFhUkYx9W'),
        decimals: 6
    },
    SLC: {
        symbol: 'SLC',
        name: 'solice',
        mintAddress: new web3_js_1.PublicKey('METAmTMXwdb8gYzyCPfXXFmZZw4rUsXX58PNsDg7zjL'),
        decimals: 6
    },
    MEAN: {
        symbol: 'MEAN',
        name: 'meanfi',
        mintAddress: new web3_js_1.PublicKey('MEANeD3XDdUmNMsRGjASkSWdC8prLYsoRJ61pPeHctD'),
        decimals: 6
    },
    ZBC: {
        symbol: 'ZBC',
        name: 'zebec-protocol',
        mintAddress: new web3_js_1.PublicKey('zebeczgi5fSEtbpfQKVZKCJ3WgYXxjkMUkNNx7fLKAF'),
        decimals: 9
    },
    ATS: {
        symbol: 'ATS',
        name: 'atlas-dex',
        mintAddress: new web3_js_1.PublicKey('HJbNXx2YMRxgfUJ6K4qeWtjatMK5KYQT1QnsCdDWywNv'),
        decimals: 9
    },
    GST: {
        symbol: 'GST',
        name: 'green-satoshi-token',
        mintAddress: new web3_js_1.PublicKey("AFbX8oGjGpmVFywbVouvhQSRmiW2aR1mohfahi4Y2AdB"),
        decimals: 9
    },
    GMT: {
        symbol: 'GMT',
        name: 'stepn',
        mintAddress: new web3_js_1.PublicKey("7i5KKsX2weiTkry7jA4ZwSuXGhs5eJBEjY8vVxR4pfRx"),
        decimals: 9
    },
    prANA: {
        symbol: 'prANA',
        name: 'nirvana-ana',
        mintAddress: new web3_js_1.PublicKey("PRAxfbouRoJ9yZqhyejEAH6RvjJ86Y82vfiZTBSM3xG"),
        decimals: 6
    },
    RATIO: {
        symbol: 'RATIO',
        name: 'ratio-finance',
        mintAddress: new web3_js_1.PublicKey("ratioMVg27rSZbSvBopUvsdrGUzeALUfFma61mpxc8J"),
        decimals: 6
    },
    TTT: {
        symbol: 'TTT',
        name: 'tabtrader',
        mintAddress: new web3_js_1.PublicKey("FNFKRV3V8DtA3gVJN6UshMiLGYA8izxFwkNWmJbFjmRj"),
        decimals: 6
    },
    RUN: {
        symbol: 'RUN',
        name: 'run',
        mintAddress: new web3_js_1.PublicKey("6F9XriABHfWhit6zmMUYAQBSy6XK5VF1cHXuW5LDpRtC"),
        decimals: 9
    },
    SLCL: {
        symbol: 'SLCL',
        name: 'solcial',
        mintAddress: new web3_js_1.PublicKey("SLCLww7nc1PD2gQPQdGayHviVVcpMthnqUz2iWKhNQV"),
        decimals: 9
    }
};
