// @mercurial-finance/vault-sdk/dist/cjs/src/vault/strategy/francium.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const web3_js_1 = require("@solana/web3.js");
const spl_token_1 = require("@solana/spl-token");
const francium_sdk_1 = __importStar(require("francium-sdk")), francium = francium_sdk_1;
const anchor = __importStar(require("@project-serum/anchor"));
const constants_1 = require("../constants");
class FranciumHandler {
    constructor(connection) {
        this.connection = connection;
        this.franciumSDK = new francium_sdk_1.default({
            connection,
        });
    }
    withdraw(walletPubKey, program, strategy, vault, tokenVault, feeVault, lpMint, userToken, userLp, amount, preInstructions, postInstructions, opt) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!walletPubKey)
                throw new Error('No user wallet public key');
            const vaultState = yield program.account.vault.fetch(vault);
            // https://github.com/Francium-DeFi/francium-sdk/blob/master/src/constants/lend/pools.ts#L59
            const lendingPools = francium.LENDING_CONFIG;
            const lendingPool = Object.values(lendingPools).find((lendingPool) => lendingPool.lendingPoolInfoAccount.equals(new web3_js_1.PublicKey(strategy.state.reserve)));
            if (!lendingPool)
                throw new Error('Cannot find francium lending pool');
            const collateralMint = lendingPool.lendingPoolShareMint;
            const strategyBuffer = new web3_js_1.PublicKey(strategy.pubkey).toBuffer();
            const [collateralVault] = yield web3_js_1.PublicKey.findProgramAddress([Buffer.from(constants_1.SEEDS.COLLATERAL_VAULT_PREFIX), strategyBuffer], program.programId);
            const accounts = [
                { pubkey: lendingPool.lendingPoolTknAccount, isWritable: true },
                { pubkey: lendingPool.marketInfoAccount, isWritable: true },
                { pubkey: lendingPool.lendingMarketAuthority },
                { pubkey: collateralMint, isWritable: true },
                { pubkey: web3_js_1.SYSVAR_CLOCK_PUBKEY },
            ];
            const remainingAccounts = [];
            for (const account of accounts) {
                remainingAccounts.push({
                    pubkey: account.pubkey,
                    isWritable: !!account.isWritable,
                    isSigner: false,
                });
            }
            const txAccounts = {
                vault,
                strategy: new web3_js_1.PublicKey(strategy.pubkey),
                reserve: new web3_js_1.PublicKey(strategy.state.reserve),
                strategyProgram: lendingPool.programId,
                collateralVault,
                feeVault,
                tokenVault,
                userToken,
                userLp,
                tokenProgram: spl_token_1.TOKEN_PROGRAM_ID,
            };
            if (opt === null || opt === void 0 ? void 0 : opt.affiliate) {
                const tx = yield opt.affiliate.affiliateProgram.methods
                    .withdrawDirectlyFromStrategy(new anchor.BN(amount), new anchor.BN(0))
                    .accounts(Object.assign(Object.assign({}, txAccounts), { partner: opt.affiliate.partner, user: opt.affiliate.user, vaultProgram: program.programId, vaultLpMint: lpMint, owner: walletPubKey }))
                    .remainingAccounts(remainingAccounts)
                    .preInstructions(preInstructions)
                    .postInstructions(postInstructions)
                    .transaction();
                return tx;
            }
            const tx = yield program.methods
                .withdrawDirectlyFromStrategy(new anchor.BN(amount), new anchor.BN(0))
                .accounts(Object.assign(Object.assign({}, txAccounts), { lpMint, user: walletPubKey }))
                .remainingAccounts(remainingAccounts)
                .preInstructions(preInstructions)
                .postInstructions(postInstructions)
                .transaction();
            return tx;
        });
    }
}
exports.default = FranciumHandler;
