// @port.finance/port-sdk/dist/cjs/models/AssetDepositConfig.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AssetDepositConfig = void 0;
const big_js_1 = __importDefault(require("big.js"));
class AssetDepositConfig {
    constructor(reserveId, args) {
        this.reserveId = reserveId;
        this.min = (args === null || args === void 0 ? void 0 : args.min) ? new big_js_1.default(args.min) : undefined;
        this.max = (args === null || args === void 0 ? void 0 : args.max) ? new big_js_1.default(args.max) : undefined;
        this.remain = (args === null || args === void 0 ? void 0 : args.remain) ? new big_js_1.default(args.remain) : undefined;
    }
    getReserveId() {
        return this.reserveId;
    }
    getMin() {
        return this.min;
    }
    getMax() {
        return this.max;
    }
    getRemain() {
        return this.remain;
    }
}
exports.AssetDepositConfig = AssetDepositConfig;
