// @port.finance/port-sdk/dist/cjs/models/AssetPrice.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AssetPrice = void 0;
const big_js_1 = __importDefault(require("big.js"));
const basic_1 = require("./basic");
class AssetPrice extends basic_1.Decimal {
    constructor(mintId, value) {
        super(value);
        this.mintId = mintId;
    }
    static of(mintId, value) {
        return new AssetPrice(mintId, value);
    }
    getMintId() {
        return this.mintId;
    }
    addFee(pct) {
        return this.multiply(new big_js_1.default(1).add(pct.getRaw()));
    }
    print(config) {
        const decimals = config.getPriceDecimals();
        return "$" + this.raw.round(decimals, 1).toFixed(decimals); // RoundHalfUp
    }
    replaceWithValue(value) {
        return new AssetPrice(this.getMintId(), value);
    }
    isCompatibleWith(that) {
        return this.getMintId().equals(that.getMintId());
    }
}
exports.AssetPrice = AssetPrice;
