// @port.finance/port-sdk/dist/cjs/models/Asset.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Asset = void 0;
const big_js_1 = __importDefault(require("big.js"));
const basic_1 = require("./basic");
const QuoteValue_1 = require("./QuoteValue");
const Share_1 = require("./Share");
const MintId_1 = require("./MintId");
class Asset extends basic_1.Token {
    constructor(mintId, lamport) {
        super(mintId, lamport);
    }
    static fromString(str, mintId, context) {
        const increment = context.multiplier;
        const lamport = basic_1.Lamport.of(new big_js_1.default(str).mul(increment).round(0, 0));
        return new Asset(mintId, lamport);
    }
    static zero(mintId) {
        return Asset.of(mintId);
    }
    static max(mintId) {
        return Asset.of(mintId, basic_1.Lamport.max());
    }
    static native(lamport) {
        return Asset.of(MintId_1.MintId.native(), lamport);
    }
    static fromTokenAccount(account) {
        return Asset.of(account.getMintId(), account.getAmount());
    }
    static of(mintId, lamport) {
        return new Asset(mintId, lamport || basic_1.Lamport.zero());
    }
    isNative() {
        return this.getMintId().isNative();
    }
    toValue(price, quantityContext) {
        console.assert(this.getMintId().equals(price.getMintId()), `asset id: ${this.getMintId()} price id: ${price.getMintId()}`);
        if (!price) {
            return QuoteValue_1.QuoteValue.zero();
        }
        const increment = quantityContext.multiplier;
        const value = this.getRaw().div(increment).mul(price.getRaw());
        return QuoteValue_1.QuoteValue.of(value);
    }
    toInterest(supplyApy) {
        if (!supplyApy.isPresent()) {
            return Asset.zero(this.getMintId());
        }
        const lamport = basic_1.Lamport.of(this.getRaw().mul(supplyApy.getUnchecked()));
        return Asset.of(this.getMintId(), lamport);
    }
    toShare(exchangeRatio) {
        console.assert(this.getMintId().equals(exchangeRatio.getAssetMintId()));
        if (!exchangeRatio.isPresent()) {
            return Share_1.Share.zero(exchangeRatio.getShareMintId());
        }
        const lamport = basic_1.Lamport.of(this.getRaw().mul(exchangeRatio.getUnchecked()).round(0));
        return Share_1.Share.of(exchangeRatio.getShareMintId(), lamport);
    }
    addFee(pct) {
        return this.multiply(new big_js_1.default(1).add(pct.getRaw()));
    }
    toNumber(context) {
        const multiplier = context.multiplier;
        const decimals = context.decimals;
        return this.getRaw().div(multiplier).round(decimals, 0).toNumber();
    }
    plain(context) {
        return this.toLimitRoundNumber(context).toString();
    }
    toLimitRoundNumber(context) {
        const multiplier = context.multiplier;
        const decimals = context.decimals;
        return this.getRaw()
            .div(multiplier)
            .round(Math.min(decimals, 6), 0)
            .toNumber();
    }
    print(context, symbol) {
        if (!context) {
            return "--";
        }
        const num = this.toLimitRoundNumber(context);
        const formatted = num > Asset.LARGE_THRESHOLD
            ? Asset.FORMATTER_LARGE.format(num)
            : Asset.FORMATTER_NORMAL.format(num);
        if (!symbol) {
            return formatted;
        }
        return formatted + " " + symbol;
    }
    wrap(value) {
        return Asset.of(this.getMintId(), value);
    }
}
exports.Asset = Asset;
Asset.MIN_NATIVE_LAMPORT = Asset.native(basic_1.Lamport.of(5000000));
Asset.SIGNIFICANT_DIGITS = 6;
Asset.LARGE_THRESHOLD = new big_js_1.default(10).pow(6).toNumber();
Asset.FORMATTER_NORMAL = new Intl.NumberFormat("en-US", {
    style: "decimal",
    maximumSignificantDigits: Asset.SIGNIFICANT_DIGITS,
});
Asset.FORMATTER_LARGE = new Intl.NumberFormat("en-US", {
    style: "decimal",
    maximumFractionDigits: 0,
});
