// @port.finance/port-sdk/dist/cjs/models/Loan.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Loan = void 0;
const basic_1 = require("./basic");
const ProfileEntry_1 = require("./ProfileEntry");
class Loan extends ProfileEntry_1.ProfileEntry {
    constructor(reserveId, amount, cumulativeBorrowRate) {
        super(reserveId, amount);
        this.cumulativeBorrowRate = cumulativeBorrowRate;
    }
    static zero(reserve) {
        return new Loan(reserve.getReserveId(), basic_1.Lamport.zero(), reserve.asset.getCumulativeBorrowRate());
    }
    accrueInterest(newCumulativeBorrowRate) {
        const compoundedInterestRate = newCumulativeBorrowRate.divide(this.cumulativeBorrowRate.getRaw());
        const newAmount = this.getAmount().multiply(compoundedInterestRate.getRaw());
        return new Loan(this.getReserveId(), newAmount, newCumulativeBorrowRate);
    }
    getCumulativeBorrowRate() {
        return this.cumulativeBorrowRate;
    }
    wrap(value) {
        return new Loan(this.getReserveId(), value, this.cumulativeBorrowRate);
    }
}
exports.Loan = Loan;
