// @port.finance/port-sdk/dist/cjs/models/Share.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Share = void 0;
const Asset_1 = require("./Asset");
const basic_1 = require("./basic");
class Share extends basic_1.Token {
    constructor(mintId, lamport) {
        super(mintId, lamport);
    }
    static zero(mintId) {
        return Share.of(mintId, basic_1.Lamport.zero());
    }
    static max(mintId) {
        return Share.of(mintId, basic_1.Lamport.max());
    }
    static fromTokenAccount(account) {
        return Share.of(account.getMintId(), account.getAmount());
    }
    static of(mintId, lamport) {
        return new Share(mintId, lamport);
    }
    toAsset(exchangeRatio) {
        console.assert(this.getMintId().equals(exchangeRatio.getShareMintId()));
        if (!exchangeRatio.isPresent()) {
            return Asset_1.Asset.zero(exchangeRatio.getAssetMintId());
        }
        const pct = exchangeRatio.getUnchecked();
        const lamport = basic_1.Lamport.of(this.getRaw().div(pct).round(0));
        return Asset_1.Asset.of(exchangeRatio.getAssetMintId(), lamport);
    }
    wrap(value) {
        return Share.of(this.getMintId(), value);
    }
}
exports.Share = Share;
