// francium-sdk/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.utils = exports.FARM_POOLS_CONFIG = exports.TOKENS_LIST = exports.PRICE_CONFIG = exports.ORCA_FARM_IDL = exports.RAYDIUM_FARM_IDL = exports.RAYDIUM_FARM_CONFIG = exports.ORCA_FARM_CONFIG = exports.LENDING_REWARD_CONFIG = exports.LENDING_CONFIG = void 0;
const main_1 = require("./main");
const info_1 = require("./constants/farm/orca/info");
Object.defineProperty(exports, "ORCA_FARM_CONFIG", { enumerable: true, get: function () { return info_1.ORCA_FARM_CONFIG; } });
const info_2 = require("./constants/farm/raydium/info");
Object.defineProperty(exports, "RAYDIUM_FARM_CONFIG", { enumerable: true, get: function () { return info_2.RAYDIUM_FARM_CONFIG; } });
const pools_1 = require("./constants/lend/pools");
const rewards_1 = require("./constants/lend/rewards");
const tools_1 = require("./utils/tools");
const math_1 = require("./utils/math");
const rebalanceKit = require("./utils/rebalance");
const price_1 = require("./constants/price");
const tokens_1 = require("./constants/tokens");
const farm_1 = require("./constants/farm");
const config_1 = require("./constants/farm/raydium/config");
exports.RAYDIUM_FARM_IDL = config_1.default;
const config_2 = require("./constants/farm/orca/config");
exports.ORCA_FARM_IDL = config_2.default;
exports.default = main_1.FranciumSDK;
exports.LENDING_CONFIG = pools_1.lendingPools;
exports.LENDING_REWARD_CONFIG = rewards_1.lendRewardInfo;
exports.PRICE_CONFIG = price_1.default;
exports.TOKENS_LIST = tokens_1.TOKENS;
exports.FARM_POOLS_CONFIG = farm_1.farmPools;
exports.utils = {
    getTokenDecimals: tools_1.getTokenDecimals,
    getAmountByDecimals: math_1.getAmountByDecimals,
    splitMultipleAccountsInfo: tools_1.splitMultipleAccountsInfo,
    getAprInfo: math_1.getAprInfo,
    rebalanceKit,
};
