// @apricot-lend/solana-swaps-js/dist/raydium/raydiumConstants.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RAYDIUM_whETH_USDC_MARKET = exports.RAYDIUM_stSOL_USDC_MARKET = exports.RAYDIUM_SRM_USDT_MARKET = exports.RAYDIUM_mSOL_USDT_MARKET = exports.RAYDIUM_RAY_USDT_MARKET = exports.RAYDIUM_ETH_USDT_MARKET = exports.RAYDIUM_BTC_USDT_MARKET = exports.RAYDIUM_MNDE_mSOL_MARKET = exports.RAYDIUM_ORCA_USDC_MARKET = exports.RAYDIUM_FTT_USDC_MARKET = exports.RAYDIUM_SBR_USDC_MARKET = exports.RAYDIUM_SRM_USDC_MARKET = exports.RAYDIUM_APT_USDC_MARKET = exports.RAYDIUM_mSOL_USDC_MARKET = exports.RAYDIUM_RAY_USDC_MARKET = exports.RAYDIUM_USDT_USDC_MARKET = exports.RAYDIUM_SOL_USDC_MARKET = exports.RAYDIUM_ETH_USDC_MARKET = exports.RAYDIUM_BTC_USDC_MARKET = exports.RaydiumMarket = exports.RAYDIUM_AMM_PROGRAM = void 0;
var web3_js_1 = require("@solana/web3.js");
var spl_token_1 = require("@solana/spl-token");
var types_1 = require("../types");
var mints_1 = require("../mints");
var serumConstants_1 = require("../serum/serumConstants");
var Parser_1 = require("../utils/Parser");
exports.RAYDIUM_AMM_PROGRAM = new web3_js_1.PublicKey("675kPX9MHTjS2zt1qfr1NYHuzeLXfQM9H24wFSUt1Mp8");
var RaydiumMarket = /** @class */ (function (_super) {
    __extends(RaydiumMarket, _super);
    function RaydiumMarket(name, tokenIdA, tokenIdB, amm, ammAuthority, openOrders, targetOrders, raydiumVaultA, raydiumVaultB, serumMarket, serumBids, serumAsks, serumEvents, serumVaultA, serumVaultB, serumVaultSigner) {
        var _this = _super.call(this, name, [tokenIdA, tokenIdB]) || this;
        _this.tokenIdA = tokenIdA;
        _this.tokenIdB = tokenIdB;
        _this.amm = amm;
        _this.ammAuthority = ammAuthority;
        _this.openOrders = openOrders;
        _this.targetOrders = targetOrders;
        _this.raydiumVaultA = raydiumVaultA;
        _this.raydiumVaultB = raydiumVaultB;
        _this.serumMarket = serumMarket;
        _this.serumBids = serumBids;
        _this.serumAsks = serumAsks;
        _this.serumEvents = serumEvents;
        _this.serumVaultA = serumVaultA;
        _this.serumVaultB = serumVaultB;
        _this.serumVaultSigner = serumVaultSigner;
        _this.INST_LAYOUT = new Parser_1.Parser()
            .u8("cmd")
            .u64("in_amount")
            .u64("min_out_amount");
        if (name !== tokenIdA + "/" + tokenIdB) {
            throw new Error("Incorrect name!");
        }
        _this.mintA = mints_1.MINTS[tokenIdA];
        _this.mintB = mints_1.MINTS[tokenIdB];
        return _this;
    }
    RaydiumMarket.prototype.getSwapper = function () {
        return this;
    };
    RaydiumMarket.prototype.createSwapInstructions = function (_fromToken, fromAmount, fromTokenAccount, _toToken, minToAmount, toTokenAccount, tradeOwner) {
        return __awaiter(this, void 0, void 0, function () {
            var buffer, ix;
            return __generator(this, function (_a) {
                buffer = this.INST_LAYOUT.encode({
                    cmd: 9,
                    in_amount: fromAmount,
                    min_out_amount: minToAmount
                });
                ix = new web3_js_1.TransactionInstruction({
                    programId: exports.RAYDIUM_AMM_PROGRAM,
                    keys: [
                        { pubkey: spl_token_1.TOKEN_PROGRAM_ID, isSigner: false, isWritable: false },
                        { pubkey: this.amm, isSigner: false, isWritable: true },
                        { pubkey: this.ammAuthority, isSigner: false, isWritable: false },
                        { pubkey: this.openOrders, isSigner: false, isWritable: true },
                        { pubkey: this.targetOrders, isSigner: false, isWritable: true },
                        { pubkey: this.raydiumVaultA, isSigner: false, isWritable: true },
                        { pubkey: this.raydiumVaultB, isSigner: false, isWritable: true },
                        { pubkey: serumConstants_1.SERUM_PROGRAM, isSigner: false, isWritable: false },
                        { pubkey: this.serumMarket, isSigner: false, isWritable: true },
                        { pubkey: this.serumBids, isSigner: false, isWritable: true },
                        { pubkey: this.serumAsks, isSigner: false, isWritable: true },
                        { pubkey: this.serumEvents, isSigner: false, isWritable: true },
                        { pubkey: this.serumVaultA, isSigner: false, isWritable: true },
                        { pubkey: this.serumVaultB, isSigner: false, isWritable: true },
                        { pubkey: this.serumVaultSigner, isSigner: false, isWritable: false },
                        { pubkey: fromTokenAccount, isSigner: false, isWritable: true },
                        { pubkey: toTokenAccount, isSigner: false, isWritable: true },
                        { pubkey: tradeOwner, isSigner: true, isWritable: false },
                    ],
                    data: buffer,
                });
                return [2 /*return*/, [ix]];
            });
        });
    };
    return RaydiumMarket;
}(types_1.Market));
exports.RaydiumMarket = RaydiumMarket;
exports.RAYDIUM_BTC_USDC_MARKET = new RaydiumMarket("BTC/USDC", types_1.TokenID.BTC, types_1.TokenID.USDC, new web3_js_1.PublicKey("6kbC5epG18DF2DwPEW34tBy5pGFS7pEGALR3v5MGxgc5"), // amm
new web3_js_1.PublicKey("5Q544fKrFoe6tsEbD7S8EmxGTJYAKtTVhAW5Q5pge4j1"), // ammAuthority
new web3_js_1.PublicKey("L6A7qW935i2HgaiaRx6xNGCGQfFr4myFU51dUSnCshd"), // openOrders
new web3_js_1.PublicKey("6DGjaczWfFthTYW7oBk3MXP2mMwrYq86PA3ki5YF6hLg"), // targetOrders
new web3_js_1.PublicKey("HWTaEDR6BpWjmyeUyfGZjeppLnH7s8o225Saar7FYDt5"), // vaultA
new web3_js_1.PublicKey("7iGcnvoLAxthsXY3AFSgkTDoqnLiuti5fyPNm2VwZ3Wz"), // vaultB
new web3_js_1.PublicKey("A8YFbxQYFVqKZaoYJLLUVcQiWP7G2MeEgW5wsAQgMvFw"), // market
new web3_js_1.PublicKey("6wLt7CX1zZdFpa6uGJJpZfzWvG6W9rxXjquJDYiFwf9K"), // bids
new web3_js_1.PublicKey("6EyVXMMA58Nf6MScqeLpw1jS12RCpry23u9VMfy8b65Y"), // asks
new web3_js_1.PublicKey("6NQqaa48SnBBJZt9HyVPngcZFW81JfDv9EjRX2M4WkbP"), // events
new web3_js_1.PublicKey("GZ1YSupuUq9kB28kX9t1j9qCpN67AMMwn4Q72BzeSpfR"), // vaultA
new web3_js_1.PublicKey("7sP9fug8rqZFLbXoEj8DETF81KasaRA1fr6jQb6ScKc5"), // vaultB
new web3_js_1.PublicKey("GBWgHXLf1fX4J1p5fAkQoEbnjpgjxUtr4mrVgtj9wW8a"));
exports.RAYDIUM_ETH_USDC_MARKET = new RaydiumMarket("ETH/USDC", types_1.TokenID.ETH, types_1.TokenID.USDC, new web3_js_1.PublicKey("AoPebtuJC4f2RweZSxcVCcdeTgaEXY64Uho8b5HdPxAR"), // amm
new web3_js_1.PublicKey("5Q544fKrFoe6tsEbD7S8EmxGTJYAKtTVhAW5Q5pge4j1"), // ammAuthority
new web3_js_1.PublicKey("7PwhFjfFaYp7w9N8k2do5Yz7c1G5ebp3YyJRhV4pkUJW"), // openOrders
new web3_js_1.PublicKey("BV2ucC7miDqsmABSkXGzsibCVWBp7gGPcvkhevDSTyZ1"), // targetOrders
new web3_js_1.PublicKey("EHT99uYfAnVxWHPLUMJRTyhD4AyQZDDknKMEssHDtor5"), // vaultA
new web3_js_1.PublicKey("58tgdkogRoMsrXZJubnFPsFmNp5mpByEmE1fF6FTNvDL"), // vaultB
new web3_js_1.PublicKey("4tSvZvnbyzHXLMTiFonMyxZoHmFqau1XArcRCVHLZ5gX"), // market
new web3_js_1.PublicKey("8tFaNpFPWJ8i7inhKSfAcSestudiFqJ2wHyvtTfsBZZU"), // bids
new web3_js_1.PublicKey("2po4TC8qiTgPsqcnbf6uMZRMVnPBzVwqqYfHP15QqREU"), // asks
new web3_js_1.PublicKey("Eac7hqpaZxiBtG4MdyKpsgzcoVN6eMe9tAbsdZRYH4us"), // events
new web3_js_1.PublicKey("7Nw66LmJB6YzHsgEGQ8oDSSsJ4YzUkEVAvysQuQw7tC4"), // vaultA
new web3_js_1.PublicKey("EsDTx47jjFACkBhy48Go2W7AQPk4UxtT4765f3tpK21a"), // vaultB
new web3_js_1.PublicKey("C5v68qSzDdGeRcs556YoEMJNsp8JiYEiEhw2hVUR8Z8y"));
exports.RAYDIUM_SOL_USDC_MARKET = new RaydiumMarket("SOL/USDC", types_1.TokenID.SOL, types_1.TokenID.USDC, new web3_js_1.PublicKey("58oQChx4yWmvKdwLLZzBi4ChoCc2fqCUWBkwMihLYQo2"), // amm
new web3_js_1.PublicKey("5Q544fKrFoe6tsEbD7S8EmxGTJYAKtTVhAW5Q5pge4j1"), // ammAuthority
new web3_js_1.PublicKey("HRk9CMrpq7Jn9sh7mzxE8CChHG8dneX9p475QKz4Fsfc"), // openOrders
new web3_js_1.PublicKey("CZza3Ej4Mc58MnxWA385itCC9jCo3L1D7zc3LKy1bZMR"), // targetOrders
new web3_js_1.PublicKey("DQyrAcCrDXQ7NeoqGgDCZwBvWDcYmFCjSb9JtteuvPpz"), // vaultA
new web3_js_1.PublicKey("HLmqeL62xR1QoZ1HKKbXRrdN1p3phKpxRMb2VVopvBBz"), // vaultB
new web3_js_1.PublicKey("9wFFyRfZBsuAha4YcuxcXLKwMxJR43S7fPfQLusDBzvT"), // market
new web3_js_1.PublicKey("14ivtgssEBoBjuZJtSAPKYgpUK7DmnSwuPMqJoVTSgKJ"), // bids
new web3_js_1.PublicKey("CEQdAFKdycHugujQg9k2wbmxjcpdYZyVLfV9WerTnafJ"), // asks
new web3_js_1.PublicKey("5KKsLVU6TcbVDK4BS6K1DGDxnh4Q9xjYJ8XaDCG5t8ht"), // events
new web3_js_1.PublicKey("36c6YqAwyGKQG66XEp2dJc5JqjaBNv7sVghEtJv4c7u6"), // vaultA
new web3_js_1.PublicKey("8CFo8bL8mZQK8abbFyypFMwEDd8tVJjHTTojMLgQTUSZ"), // vaultB
new web3_js_1.PublicKey("F8Vyqk3unwxkXukZFQeYyGmFfTG3CAX4v24iyrjEYBJV"));
exports.RAYDIUM_USDT_USDC_MARKET = new RaydiumMarket("USDT/USDC", types_1.TokenID.USDT, types_1.TokenID.USDC, new web3_js_1.PublicKey("7TbGqz32RsuwXbXY7EyBCiAnMbJq1gm1wKmfjQjuwoyF"), // amm
new web3_js_1.PublicKey("5Q544fKrFoe6tsEbD7S8EmxGTJYAKtTVhAW5Q5pge4j1"), // ammAuthority
new web3_js_1.PublicKey("6XXvXS3meWqnftEMUgdY8hDWGJfrb8t22x2k1WyVYwhF"), // openOrders
new web3_js_1.PublicKey("AXY75qWM1t5X16FaeUovd9ZjL1W698cV843sDHV5EMqb"), // targetOrders
new web3_js_1.PublicKey("Enb9jGaKzgDBfEbbUN3Ytx2ZLoZuBhBpjVX6DULiRmvu"), // vaultA
new web3_js_1.PublicKey("HyyZpz1JUZjsfyiVSt3qz6E9PkwnBcyhUg4zKGthMNeH"), // vaultB
new web3_js_1.PublicKey("77quYg4MGneUdjgXCunt9GgM1usmrxKY31twEy3WHwcS"), // market
new web3_js_1.PublicKey("37m9QdvxmKRdjm3KKV2AjTiGcXMfWHQpVFnmhtb289yo"), // bids
new web3_js_1.PublicKey("AQKXXC29ybqL8DLeAVNt3ebpwMv8Sb4csberrP6Hz6o5"), // asks
new web3_js_1.PublicKey("9MgPMkdEHFX7DZaitSh6Crya3kCCr1As6JC75bm3mjuC"), // events
new web3_js_1.PublicKey("H61Y7xVnbWVXrQQx3EojTEqf3ogKVY5GfGjEn5ewyX7B"), // vaultA
new web3_js_1.PublicKey("9FLih4qwFMjdqRAGmHeCxa64CgjP1GtcgKJgHHgz44ar"), // vaultB
new web3_js_1.PublicKey("FGBvMAu88q9d1Csz7ZECB5a2gbWwp6qicNxN2Mo7QhWG"));
exports.RAYDIUM_RAY_USDC_MARKET = new RaydiumMarket("RAY/USDC", types_1.TokenID.RAY, types_1.TokenID.USDC, new web3_js_1.PublicKey("6UmmUiYoBjSrhakAobJw8BvkmJtDVxaeBtbt7rxWo1mg"), // amm
new web3_js_1.PublicKey("5Q544fKrFoe6tsEbD7S8EmxGTJYAKtTVhAW5Q5pge4j1"), // ammAuthority
new web3_js_1.PublicKey("J8u8nTHYtvudyqwLrXZboziN95LpaHFHpd97Jm5vtbkW"), // openOrders
new web3_js_1.PublicKey("3cji8XW5uhtsA757vELVFAeJpskyHwbnTSceMFY5GjVT"), // targetOrders
new web3_js_1.PublicKey("FdmKUE4UMiJYFK5ogCngHzShuVKrFXBamPWcewDr31th"), // vaultA
new web3_js_1.PublicKey("Eqrhxd7bDUCH3MepKmdVkgwazXRzY6iHhEoBpY7yAohk"), // vaultB
new web3_js_1.PublicKey("2xiv8A5xrJ7RnGdxXB42uFEkYHJjszEhaJyKKt4WaLep"), // market
new web3_js_1.PublicKey("Hf84mYadE1VqSvVWAvCWc9wqLXak4RwXiPb4A91EAUn5"), // bids
new web3_js_1.PublicKey("DC1HsWWRCXVg3wk2NndS5LTbce3axwUwUZH1RgnV4oDN"), // asks
new web3_js_1.PublicKey("H9dZt8kvz1Fe5FyRisb77KcYTaN8LEbuVAfJSnAaEABz"), // events
new web3_js_1.PublicKey("GGcdamvNDYFhAXr93DWyJ8QmwawUHLCyRqWL3KngtLRa"), // vaultA
new web3_js_1.PublicKey("22jHt5WmosAykp3LPGSAKgY45p7VGh4DFWSwp21SWBVe"), // vaultB
new web3_js_1.PublicKey("FmhXe9uG6zun49p222xt3nG1rBAkWvzVz7dxERQ6ouGw"));
exports.RAYDIUM_mSOL_USDC_MARKET = new RaydiumMarket("mSOL/USDC", types_1.TokenID.mSOL, types_1.TokenID.USDC, new web3_js_1.PublicKey("ZfvDXXUhZDzDVsapffUyXHj9ByCoPjP4thL6YXcZ9ix"), // amm
new web3_js_1.PublicKey("5Q544fKrFoe6tsEbD7S8EmxGTJYAKtTVhAW5Q5pge4j1"), // ammAuthority
new web3_js_1.PublicKey("4zoatXFjMSirW2niUNhekxqeEZujjC1oioKCEJQMLeWF"), // openOrders
new web3_js_1.PublicKey("Kq9Vgb8ntBzZy5doEER2p4Zpt8SqW2GqJgY5BgWRjDn"), // targetOrders
new web3_js_1.PublicKey("8JUjWjAyXTMB4ZXcV7nk3p6Gg1fWAAoSck7xekuyADKL"), // vaultA
new web3_js_1.PublicKey("DaXyxj42ZDrp3mjrL9pYjPNyBp5P8A2f37am4Kd4EyrK"), // vaultB
new web3_js_1.PublicKey("6oGsL2puUgySccKzn9XA9afqF217LfxP5ocq4B3LWsjy"), // market
new web3_js_1.PublicKey("8qyWhEcpuvEsdCmY1kvEnkTfgGeWHmi73Mta5jgWDTuT"), // bids
new web3_js_1.PublicKey("PPnJy6No31U45SVSjWTr45R8Q73X6bNHfxdFqr2vMq3"), // asks
new web3_js_1.PublicKey("BC8Tdzz7rwvuYkJWKnPnyguva27PQP5DTxosHVQrEzg9"), // events
new web3_js_1.PublicKey("2y3BtF5oRBpLwdoaGjLkfmT3FY3YbZCKPbA9zvvx8Pz7"), // vaultA
new web3_js_1.PublicKey("6w5hF2hceQRZbaxjPJutiWSPAFWDkp3YbY2Aq3RpCSKe"), // vaultB
new web3_js_1.PublicKey("9dEVMESKXcMQNndoPc5ji9iTeDJ9GfToboy8prkZeT96"));
exports.RAYDIUM_APT_USDC_MARKET = new RaydiumMarket("APT/USDC", types_1.TokenID.APT, types_1.TokenID.USDC, new web3_js_1.PublicKey("4crhN3D8R5rnZd66q9b32P7K649e5XdzCfPMPiTzBceH"), // amm
new web3_js_1.PublicKey("5Q544fKrFoe6tsEbD7S8EmxGTJYAKtTVhAW5Q5pge4j1"), // ammAuthority
new web3_js_1.PublicKey("9ZkyYVUKZ3iWZnx6uJNUNKdv3NW3WcKNWZMg2YDYTxSx"), // openOrders
new web3_js_1.PublicKey("FWKNVdavvUKdcpCCU3XT1dsCEbHF1ak21q2EzoyMy1av"), // targetOrders
new web3_js_1.PublicKey("6egmkyieHa2R2TiVoLkwmy3fXG1F8EG8KmEMBN2Lahh7"), // vaultA
new web3_js_1.PublicKey("4dcKsdDe39Yp4NDzko1Jv6ViSDo2AUMh2KGxT6giidpA"), // vaultB
new web3_js_1.PublicKey("ATjWoJDChATL7E5WVeSk9EsoJAhZrHjzCZABNx3Miu8B"), // market
new web3_js_1.PublicKey("5M3bbs43jpQWkXccVbny317rKFFq9bZT3ccv3YoLSwRd"), // bids
new web3_js_1.PublicKey("EZYkKSRfdqbQbwBrVmkkWXmosYFB4cVhcT4jLT3Qjfxt"), // asks
new web3_js_1.PublicKey("7tnT8FCXaN5zryRpjJieFHLLVBUtZYR3LhYDh3da9HJh"), // events
new web3_js_1.PublicKey("GesJe56oHgbA9gTxNz5BFGXxhGdScteKNdmYeLj6PBmq"), // vaultA
new web3_js_1.PublicKey("GvjFcsncRnqfmRig7kkgoeur7QzkZaPurpHHyWyeriNu"), // vaultB
new web3_js_1.PublicKey("Hfn1km6sEcBnQ6S1SLYsJZkwQzx7kJJ9o8UqwWhPNiW3"));
exports.RAYDIUM_SRM_USDC_MARKET = new RaydiumMarket("SRM/USDC", types_1.TokenID.SRM, types_1.TokenID.USDC, new web3_js_1.PublicKey("8tzS7SkUZyHPQY7gLqsMCXZ5EDCgjESUHcB17tiR1h3Z"), // amm
new web3_js_1.PublicKey("5Q544fKrFoe6tsEbD7S8EmxGTJYAKtTVhAW5Q5pge4j1"), // ammAuthority
new web3_js_1.PublicKey("GJwrRrNeeQKY2eGzuXGc3KBrBftYbidCYhmA6AZj2Zur"), // openOrders
new web3_js_1.PublicKey("26LLpo8rscCpMxyAnJsqhqESPnzjMGiFdmXA4eF2Jrk5"), // targetOrders
new web3_js_1.PublicKey("zuLDJ5SEe76L3bpFp2Sm9qTTe5vpJL3gdQFT5At5xXG"), // vaultA
new web3_js_1.PublicKey("4usvfgPDwXBX2ySX11ubTvJ3pvJHbGEW2ytpDGCSv5cw"), // vaultB
new web3_js_1.PublicKey("ByRys5tuUWDgL73G8JBAEfkdFf8JWBzPBDHsBVQ5vbQA"), // market
new web3_js_1.PublicKey("AuL9JzRJ55MdqzubK4EutJgAumtkuFcRVuPUvTX39pN8"), // bids
new web3_js_1.PublicKey("8Lx9U9wdE3afdqih1mCAXy3unJDfzSaXFqAvoLMjhwoD"), // asks
new web3_js_1.PublicKey("6o44a9xdzKKDNY7Ff2Qb129mktWbsCT4vKJcg2uk41uy"), // events
new web3_js_1.PublicKey("Ecfy8et9Mft9Dkavnuh4mzHMa2KWYUbBTA5oDZNoWu84"), // vaultA
new web3_js_1.PublicKey("hUgoKy5wjeFbZrXDW4ecr42T4F5Z1Tos31g68s5EHbP"), // vaultB
new web3_js_1.PublicKey("GVV4ZT9pccwy9d17STafFDuiSqFbXuRTdvKQ1zJX6ttX"));
exports.RAYDIUM_SBR_USDC_MARKET = new RaydiumMarket("SBR/USDC", types_1.TokenID.SBR, types_1.TokenID.USDC, new web3_js_1.PublicKey("5cmAS6Mj4pG2Vp9hhyu3kpK9yvC7P6ejh9HiobpTE6Jc"), // amm
new web3_js_1.PublicKey("5Q544fKrFoe6tsEbD7S8EmxGTJYAKtTVhAW5Q5pge4j1"), // ammAuthority
new web3_js_1.PublicKey("8bEDWrUBqMV7ei55PgySABm8swC9WFW24NB6U5f5sPJT"), // openOrders
new web3_js_1.PublicKey("G2nswHPqZLXtMimXZtsiLHVZ5gJ9GTiKRdLxahDDdYag"), // targetOrders
new web3_js_1.PublicKey("8vwzjpW7KPGFLQdRuyoBBoiBCsNG6SLRGssKMNsofch2"), // vaultA
new web3_js_1.PublicKey("AcK6bv25Q7xofBUiXKwUgueSi3ELS6anMbmNn2NPV8FZ"), // vaultB
new web3_js_1.PublicKey("HXBi8YBwbh4TXF6PjVw81m8Z3Cc4WBofvauj5SBFdgUs"), // market
new web3_js_1.PublicKey("FdGKYpHxpQEkRitZw6KZ8b21Q2mYiATHXZgJjFDhnRWM"), // bids
new web3_js_1.PublicKey("cxqTRyeoGeh6TBEgo3NAieHaMkdmfZiCjSEfkNAe1Y3"), // asks
new web3_js_1.PublicKey("EUre4VPaLh7B95qG3JPS3atquJ5hjbwtX7XFcTtVNkc7"), // events
new web3_js_1.PublicKey("38r5pRYVzdScrJNZowNyrpjVbtRKQ5JMcQxn7PgKE45L"), // vaultA
new web3_js_1.PublicKey("4YqAGXQEQTQbn4uKX981yCiSjUuYPV8aCajc9qQh3QPy"), // vaultB
new web3_js_1.PublicKey("84aqZGKMzbr8ddA267ML7JUTAjieVJe8oR1yGUaKwP53"));
exports.RAYDIUM_FTT_USDC_MARKET = new RaydiumMarket("FTT/USDC", types_1.TokenID.FTT, types_1.TokenID.USDC, new web3_js_1.PublicKey("4C2Mz1bVqe42QDDTyJ4HFCFFGsH5YDzo91Cen5w5NGun"), // amm
new web3_js_1.PublicKey("5Q544fKrFoe6tsEbD7S8EmxGTJYAKtTVhAW5Q5pge4j1"), // ammAuthority
new web3_js_1.PublicKey("23WS5XY3srvBtnP6hXK64HAsXTuj1kT7dd7srjrJUNTR"), // openOrders
new web3_js_1.PublicKey("CYbPm6BCkMyX8NnnS7AoCUkpxHVwYyxvjQWwZLsrFcLR"), // targetOrders
new web3_js_1.PublicKey("4TaBaR1ZgHNuQM3QNHnjJdAT4Sws9cz46MtVWVebg7Ax"), // vaultA
new web3_js_1.PublicKey("7eDiHvsfcZf1VFC2sUDJwr5EMMr66TpQ2nmAreUjoASV"), // vaultB
new web3_js_1.PublicKey("2Pbh1CvRVku1TgewMfycemghf6sU9EyuFDcNXqvRmSxc"), // market
new web3_js_1.PublicKey("9HTDV2r7cQBUKL3fgcJZCUfmJsKA9qCP7nZAXyoyaQou"), // bids
new web3_js_1.PublicKey("EpnUJCMCQNZi45nCBoNs6Bugy67Kj3bCSTLYPfz6jkYH"), // asks
new web3_js_1.PublicKey("2XHxua6ZaPKpCGUNvSvTwc9teJBmexp8iMWCLu4mtzGb"), // events
new web3_js_1.PublicKey("4LXjM6rptNvhBZTcWk4AL49oF4oA8AH7D4CV6z7tmpX3"), // vaultA
new web3_js_1.PublicKey("2ycZAqQ3YNPfBZnKTbz2FqPiV7fmTQpzF95vjMUekP5z"), // vaultB
new web3_js_1.PublicKey("B5b9ddFHrjndUieLAKkyzB1xmq8sNqGGZPmbyYWPzCyu"));
exports.RAYDIUM_ORCA_USDC_MARKET = new RaydiumMarket("ORCA/USDC", types_1.TokenID.ORCA, types_1.TokenID.USDC, new web3_js_1.PublicKey("C5yXRTp39qv5WZrfiqoqeyK6wvbqS97oBqbsDUqfZyu"), // amm
new web3_js_1.PublicKey("5Q544fKrFoe6tsEbD7S8EmxGTJYAKtTVhAW5Q5pge4j1"), // ammAuthority
new web3_js_1.PublicKey("BUwThGpiXwei6xeAZyeSofZYAsQRwnqhyyZ3Xe3J1YAB"), // openOrders
new web3_js_1.PublicKey("3g7Ef2aZzvWo57Cggv8o8dnMLGz2NSB1BRNyvVnb8AYm"), // targetOrders
new web3_js_1.PublicKey("48uXZgcnxxDSipQoXMdFmvDsu3xwDsEjHnhKXVYpeHvF"), // vaultA
new web3_js_1.PublicKey("8eLo3ppAUnjwa4HekixbZ6wTkKGgcMXF3NzxYpduV3if"), // vaultB
new web3_js_1.PublicKey("8N1KkhaCYDpj3awD58d85n973EwkpeYnRp84y1kdZpMX"), // market
new web3_js_1.PublicKey("HaAjqsdR6CzDJAioL6s9RGYL7tNC84Hv65S1Gm6MeS9s"), // bids
new web3_js_1.PublicKey("BQUychhbQfWHsAdTtrcy3DxPRm3dbqZTfYy1W7PQS9e"), // asks
new web3_js_1.PublicKey("3ajZQLGpAiTnX9quZyoRw1T4E5emWbTAjFtdVyfevXds"), // events
new web3_js_1.PublicKey("4noUQEJF15yMVWHc7JkWid5EKoE6XLjQEHfdN3pT43NZ"), // vaultA
new web3_js_1.PublicKey("38DxyYjp4ZqAqjrvAPvDhdALYd4y91jxcpnj28hbvyky"), // vaultB
new web3_js_1.PublicKey("Dtz4cysczNNTUbHMqnZW2UfUm87bGecR98snGZePt2ot"));
exports.RAYDIUM_MNDE_mSOL_MARKET = new RaydiumMarket("MNDE/mSOL", types_1.TokenID.MNDE, types_1.TokenID.mSOL, new web3_js_1.PublicKey("2kPA9XUuHUifcCYTnjSuN7ZrC3ma8EKPrtzUhC86zj3m"), // amm
new web3_js_1.PublicKey("5Q544fKrFoe6tsEbD7S8EmxGTJYAKtTVhAW5Q5pge4j1"), // ammAuthority
new web3_js_1.PublicKey("G3qeShDT2w3Y9XnJbk5TZsx1qbxkBLFmRsnNVLMnkNZb"), // openOrders
new web3_js_1.PublicKey("DfMpzNeT4XHs2xtN74j5q94QfqPSJbng5BgGyyyChsVm"), // targetOrders
new web3_js_1.PublicKey("F1zwWuPfYLZfLykDYUqu43A74TUsv8mHuWL6BUrwVhL7"), // vaultA
new web3_js_1.PublicKey("TuT7ftAgCQGsETei4Q4nMBwp2QLcDwKnixAEgFSBuao"), // vaultB
new web3_js_1.PublicKey("AVxdeGgihchiKrhWne5xyUJj7bV2ohACkQFXMAtpMetx"), // market
new web3_js_1.PublicKey("9YBjtad6ZxR7hxNXyTjRRPnPgS7geiBMHbBp4BqHsgV2"), // bids
new web3_js_1.PublicKey("8UZpvreCr8bprUwstHMPb1pe5jQY82N9fJ1XLa3oKMXg"), // asks
new web3_js_1.PublicKey("3eeXmsg8byQEC6Q18NE7MSgSbnAJkxz8KNPbW2zfKyfY"), // events
new web3_js_1.PublicKey("aj1igzDQNRg18h9yFGvNqMPBfCGGWLDvKDp2NdYh92C"), // vaultA
new web3_js_1.PublicKey("3QjiyDAny7ZrwPohN8TecXL4jBwGWoSUe7hzTiX35Pza"), // vaultB
new web3_js_1.PublicKey("6Ysd8CE6KwC7KQYpPD9Ax8B77z3bWRnHt1SVrBM8AYC9"));
exports.RAYDIUM_BTC_USDT_MARKET = new RaydiumMarket("BTC/USDT", types_1.TokenID.BTC, types_1.TokenID.USDT, new web3_js_1.PublicKey("AMMwkf57c7ZsbbDCXvBit9zFehMr1xRn8ZzaT1iDF18o"), // amm
new web3_js_1.PublicKey("5Q544fKrFoe6tsEbD7S8EmxGTJYAKtTVhAW5Q5pge4j1"), // ammAuthority
new web3_js_1.PublicKey("G5rZ4Qfv5SxpJegVng5FuZftDrJkzLkxQUNjEXuoczX5"), // openOrders
new web3_js_1.PublicKey("DMEasFJLDw27MLkTBFqSX2duvV5GV6LzwtoVqVfBqeGR"), // targetOrders
new web3_js_1.PublicKey("7KwCHoQ9nqTnGea4XrcfLUr1pwEWp2maGBHWFqBTeoKW"), // vaultA
new web3_js_1.PublicKey("HwbXe9YJVez3BKK22jBH1i64YeX2fSKaYny5jrcPDxAk"), // vaultB
new web3_js_1.PublicKey("C1EuT9VokAKLiW7i2ASnZUvxDoKuKkCpDDeNxAptuNe4"), // market
new web3_js_1.PublicKey("2e2bd5NtEGs6pb758QHUArNxt6X9TTC5abuE1Tao6fhS"), // bids
new web3_js_1.PublicKey("F1tDtTDNzusig3kJwhKwGWspSu8z2nRwNXFWc6wJowjM"), // asks
new web3_js_1.PublicKey("FERWWtsZoSLcHVpfDnEBnUqHv4757kTUUZhLKBCbNfpS"), // events
new web3_js_1.PublicKey("DSf7hGudcxhhegMpZA1UtSiW4RqKgyEex9mqQECWwRgZ"), // vaultA
new web3_js_1.PublicKey("BD8QnhY2T96h6KwyJoCT9abMcPBkiaFuBNK9h6FUNX2M"), // vaultB
new web3_js_1.PublicKey("EPzuCsSzHwhYWn2j69HQPKWuWz6wuv4ANZiVigLGMBoD"));
exports.RAYDIUM_ETH_USDT_MARKET = new RaydiumMarket("ETH/USDT", types_1.TokenID.ETH, types_1.TokenID.USDT, new web3_js_1.PublicKey("He3iAEV5rYjv6Xf7PxKro19eVrC3QAcdic5CF2D2obPt"), // amm
new web3_js_1.PublicKey("5Q544fKrFoe6tsEbD7S8EmxGTJYAKtTVhAW5Q5pge4j1"), // ammAuthority
new web3_js_1.PublicKey("8x4uasC632WSrk3wgwoCWHy7MK7Xo2WKAe9vV93tj5se"), // openOrders
new web3_js_1.PublicKey("G1eji3rrfRFfvHUbPEEbvnjmJ4eEyXeiJBVbMTUPfKL1"), // targetOrders
new web3_js_1.PublicKey("DZZwxvJakqbraXTbjRW3QoGbW5GK4R5nmyrrGrFMKWgh"), // vaultA
new web3_js_1.PublicKey("HoGPb5Rp44TyR1EpM5pjQQyFUdgteeuzuMHtimGkAVHo"), // vaultB
new web3_js_1.PublicKey("7dLVkUfBVfCGkFhSXDCq1ukM9usathSgS716t643iFGF"), // market
new web3_js_1.PublicKey("J8a3dcUkMwrE5kxN86gsL1Mwrg63RnGdvWsPbgdFqC6X"), // bids
new web3_js_1.PublicKey("F6oqP13HNZho3bhwuxTmic4w5iNgTdn89HdihMUNR24i"), // asks
new web3_js_1.PublicKey("CRjXyfAxboMfCAmsvBw7pdvkfBY7XyGxB7CBTuDkm67v"), // events
new web3_js_1.PublicKey("2CZ9JbDYPux5obFXb9sefwKyG6cyteNBSzbstYQ3iZxE"), // vaultA
new web3_js_1.PublicKey("D2f4NG1NC1yeBM2SgRe5YUF91w3M4naumGQMWjGtxiiE"), // vaultB
new web3_js_1.PublicKey("CVVGPFejAj3A75qPy2116iJFma7zGEuL8DgnxhwUaFBF"));
exports.RAYDIUM_RAY_USDT_MARKET = new RaydiumMarket("RAY/USDT", types_1.TokenID.RAY, types_1.TokenID.USDT, new web3_js_1.PublicKey("DVa7Qmb5ct9RCpaU7UTpSaf3GVMYz17vNVU67XpdCRut"), // amm
new web3_js_1.PublicKey("5Q544fKrFoe6tsEbD7S8EmxGTJYAKtTVhAW5Q5pge4j1"), // ammAuthority
new web3_js_1.PublicKey("7UF3m8hDGZ6bNnHzaT2YHrhp7A7n9qFfBj6QEpHPv5S8"), // openOrders
new web3_js_1.PublicKey("3K2uLkKwVVPvZuMhcQAPLF8hw95somMeNwJS7vgWYrsJ"), // targetOrders
new web3_js_1.PublicKey("3wqhzSB9avepM9xMteiZnbJw75zmTBDVmPFLTQAGcSMN"), // vaultA
new web3_js_1.PublicKey("5GtSbKJEPaoumrDzNj4kGkgZtfDyUceKaHrPziazALC1"), // vaultB
new web3_js_1.PublicKey("teE55QrL4a4QSfydR9dnHF97jgCfptpuigbb53Lo95g"), // market
new web3_js_1.PublicKey("AvKStCiY8LTp3oDFrMkiHHxxhxk4sQUWnGVcetm4kRpy"), // bids
new web3_js_1.PublicKey("Hj9kckvMX96mQokfMBzNCYEYMLEBYKQ9WwSc1GxasW11"), // asks
new web3_js_1.PublicKey("58KcficuUqPDcMittSddhT8LzsPJoH46YP4uURoMo5EB"), // events
new web3_js_1.PublicKey("2kVNVEgHicvfwiyhT2T51YiQGMPFWLMSp8qXc1hHzkpU"), // vaultA
new web3_js_1.PublicKey("5AXZV7XfR7Ctr6yjQ9m9dbgycKeUXWnWqHwBTZT6mqC7"), // vaultB
new web3_js_1.PublicKey("HzWpBN6ucpsA9wcfmhLAFYqEUmHjE9n2cGHwunG5avpL"));
exports.RAYDIUM_mSOL_USDT_MARKET = new RaydiumMarket("mSOL/USDT", types_1.TokenID.mSOL, types_1.TokenID.USDT, new web3_js_1.PublicKey("BhuMVCzwFVZMSuc1kBbdcAnXwFg9p4HJp7A9ddwYjsaF"), // amm
new web3_js_1.PublicKey("5Q544fKrFoe6tsEbD7S8EmxGTJYAKtTVhAW5Q5pge4j1"), // ammAuthority
new web3_js_1.PublicKey("67xxC7oyzGFMVX8AaAHqcT3UWpPt4fMsHuoHrHvauhog"), // openOrders
new web3_js_1.PublicKey("HrNUwbZF4NPRSdZ9hwD7EWV1cwQoJ9Yhu9Jf7ybXALpe"), // targetOrders
new web3_js_1.PublicKey("FaoMKkKzMDQaURce1VLewT6K38F6FQS5UQXD1mTXJ2Cb"), // vaultA
new web3_js_1.PublicKey("GE8m3rHHejrNf4jE96n5gzMmLbxTfPPcmv9Ppaw24FZa"), // vaultB
new web3_js_1.PublicKey("HxkQdUnrPdHwXP5T9kewEXs3ApgvbufuTfdw9v1nApFd"), // market
new web3_js_1.PublicKey("wNv6YZ31PX5hS42XCijwgd7SuMAu63aPvDWjMNTM2UP"), // bids
new web3_js_1.PublicKey("7g28QYJPPNypyPvoAdir8WzPT2Me78u78jufiG7M3wym"), // asks
new web3_js_1.PublicKey("Ee9UPY9CH2jHx2LLW2daLyc9VS5Bnp4yTykw4aveeXLX"), // events
new web3_js_1.PublicKey("FgVVda2Wnp2PuDpuh23B341qZx2cnArqVNSgxsU877Y"), // vaultA
new web3_js_1.PublicKey("2PtdrUGJd7aYoMKXpQ5d19r5Aa1z8dkRj6NNRCNGTE3D"), // vaultB
new web3_js_1.PublicKey("QMhH9Mnv1jg8tLNanAvKf3ymbuzh7sDENyjCgiyn3Kk"));
exports.RAYDIUM_SRM_USDT_MARKET = new RaydiumMarket("SRM/USDT", types_1.TokenID.SRM, types_1.TokenID.USDT, new web3_js_1.PublicKey("af8HJg2ffWoKJ6vKvkWJUJ9iWbRR83WgXs8HPs26WGr"), // amm
new web3_js_1.PublicKey("5Q544fKrFoe6tsEbD7S8EmxGTJYAKtTVhAW5Q5pge4j1"), // ammAuthority
new web3_js_1.PublicKey("8E2GLzSgLmzWdpdXjjEaHbPXRXsA5CFehg6FP6N39q2e"), // openOrders
new web3_js_1.PublicKey("8R5TVxXvRfCaYvT493FWAJyLt8rVssUHYVGbGupAbYaQ"), // targetOrders
new web3_js_1.PublicKey("D6b4Loa4LoidUor2ffouE5BTMt6tLP6MtkNrsfBWG2C3"), // vaultA
new web3_js_1.PublicKey("4gNeJniq6yqEygFmbAJa82TQjH1j3Fczm4bdeBHhwGJ1"), // vaultB
new web3_js_1.PublicKey("AtNnsY1AyRERWJ8xCskfz38YdvruWVJQUVXgScC1iPb"), // market
new web3_js_1.PublicKey("EE2CYFBSoMvcUR9mkEF6tt8kBFhW9zcuFmYqRM9GmqYb"), // bids
new web3_js_1.PublicKey("nkNzrV3ZtkWCft6ykeNGXXCbNSemqcauYKiZdf5JcKQ"), // asks
new web3_js_1.PublicKey("2i34Kriz23ZaQaJK6FVhzkfLhQj8DSqdQTmMwz4FF9Cf"), // events
new web3_js_1.PublicKey("GxPFMyeb7BUnu2mtGV2Zvorjwt8gxHqwL3r2kVDe6rZ8"), // vaultA
new web3_js_1.PublicKey("149gvUQZeip4u8bGra5yyN11btUDahDVHrixzknfKFrL"), // vaultB
new web3_js_1.PublicKey("4yWr7H2p8rt11QnXb2yxQF3zxSdcToReu5qSndWFEJw"));
exports.RAYDIUM_stSOL_USDC_MARKET = new RaydiumMarket("stSOL/USDC", types_1.TokenID.stSOL, types_1.TokenID.USDC, new web3_js_1.PublicKey("6a1CsrpeZubDjEJE9s1CMVheB6HWM5d7m1cj2jkhyXhj"), // amm
new web3_js_1.PublicKey("5Q544fKrFoe6tsEbD7S8EmxGTJYAKtTVhAW5Q5pge4j1"), // ammAuthority
new web3_js_1.PublicKey("28NQqHxrqMYMQ67aWyn9AzZ1F16PYd4zvLpiiKnEZpsD"), // openOrders
new web3_js_1.PublicKey("B8nmqinHQjyqAnMWNiqSzs1Jb8VbMpX5k9VUMnDp1gUA"), // targetOrders
new web3_js_1.PublicKey("DD6oh3HRCvMzqHkGeUW3za4pLgWNPJdV6aNYW3gVjXXi"), // vaultA
new web3_js_1.PublicKey("6KR4qkJN91LGko2gdizheri8LMtCwsJrhtsQt6QPwCi5"), // vaultB
new web3_js_1.PublicKey("5F7LGsP1LPtaRV7vVKgxwNYX4Vf22xvuzyXjyar7jJqp"), // market
new web3_js_1.PublicKey("HjJSzUbis6VhBZLCbSFN1YtvWLLdxutb7WEvymCLrBJt"), // bids
new web3_js_1.PublicKey("9e37wf6QUqe2s4J6UUNsuv6REQkwTxd47hXhDanm1adp"), // asks
new web3_js_1.PublicKey("CQY7LwdZJrfLRZcmEzUYp34XJbxhnxgF4UXmLKqJPLCk"), // events
new web3_js_1.PublicKey("4gqecEySZu6SEgCNhBJm7cEn2TFqCMsMNoiyski5vMTD"), // vaultA
new web3_js_1.PublicKey("6FketuhRzyTpevhgjz4fFgd5GL9fHeBeRsq9uJvu8h9m"), // vaultB
new web3_js_1.PublicKey("x1vRSsrhXkSn7xzJfu9mYP2i19SPqG1gjyj3vUWhim1"));
exports.RAYDIUM_whETH_USDC_MARKET = new RaydiumMarket("whETH/USDC", types_1.TokenID.whETH, types_1.TokenID.USDC, new web3_js_1.PublicKey("EoNrn8iUhwgJySD1pHu8Qxm5gSQqLK3za4m8xzD2RuEb"), // amm
new web3_js_1.PublicKey("5Q544fKrFoe6tsEbD7S8EmxGTJYAKtTVhAW5Q5pge4j1"), // ammAuthority
new web3_js_1.PublicKey("6iwDsRGaQucEcfXX8TgDW1eyTfxLAGrypxdMJ5uqoYcp"), // openOrders
new web3_js_1.PublicKey("EGZL5PtEnSHrNmeoQF64wXG6b5oqiTArDvAQuSRyomX5"), // targetOrders
new web3_js_1.PublicKey("DVWRhoXKCoRbvC5QUeTECRNyUSU1gwUM48dBMDSZ88U"), // vaultA
new web3_js_1.PublicKey("HftKFJJcUTu6xYcS75cDkm3y8HEkGgutcbGsdREDWdMr"), // vaultB
new web3_js_1.PublicKey("8Gmi2HhZmwQPVdCwzS7CM66MGstMXPcTVHA7jF19cLZz"), // market
new web3_js_1.PublicKey("3nXzH1gYKM1FKdSLHM7GCRG76mhKwyDjwinJxAg8jjx6"), // bids
new web3_js_1.PublicKey("b3L5dvehk48X4mDoKzZUZKA4nXGpPAMFkYxHZmsZ98n"), // asks
new web3_js_1.PublicKey("3z4QQPFdgNSxazqEAzmZD5C5tJWepczimVqWak2ZPY8v"), // events
new web3_js_1.PublicKey("8cCoWNtgCL7pMapGZ6XQ6NSyD1KC9cosUEs4QgeVq49d"), // vaultA
new web3_js_1.PublicKey("C7KrymKrLWhCsSjFaUquXU3SYRmgYLRmMjQ4dyQeFiGE"), // vaultB
new web3_js_1.PublicKey("FG3z1H2BBsf5ekEAxSc1K6DERuAuiXpSdUGkYecQrP5v"));
