// @apricot-lend/sdk-ts/dist/utils/Logger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LogEmergency = exports.LogAlert = exports.LogError = exports.LogWarning = exports.LogInfo = exports.LogDebug = exports.TAG_EMERGENCY = exports.TAG_ALERT = exports.TAG_ERROR = exports.TAG_WARN = exports.TAG_INFO = exports.TAG_DEBUG = exports.INFO_LOG_ENV = exports.DEBUG_LOG_ENV = void 0;
exports.DEBUG_LOG_ENV = 'APRICOT_LOG_DEBUG';
exports.INFO_LOG_ENV = 'APRICOT_LOG_INFO';
exports.TAG_DEBUG = 'DEBUG:';
exports.TAG_INFO = 'INFO:';
exports.TAG_WARN = 'WARN:';
exports.TAG_ERROR = 'ERROR:';
exports.TAG_ALERT = 'ALERT:';
exports.TAG_EMERGENCY = 'EMERGENCY:';
function LogDebug(content) {
    // info logging is silenced by default. Requires the APRICOT_LOG_INFO env to be set to actually log
    if (process.env[exports.DEBUG_LOG_ENV] === 'true') {
        console.log(exports.TAG_DEBUG + content);
    }
}
exports.LogDebug = LogDebug;
function LogInfo(content) {
    // info logging is silenced by default. Requires the APRICOT_LOG_INFO env to be set to actually log
    if (process.env[exports.INFO_LOG_ENV] !== 'false') {
        console.log(exports.TAG_INFO + content);
    }
}
exports.LogInfo = LogInfo;
function LogWarning(content) {
    console.log(exports.TAG_WARN + content);
}
exports.LogWarning = LogWarning;
function LogError(content) {
    console.log(exports.TAG_ERROR + content);
}
exports.LogError = LogError;
function LogAlert(content) {
    console.log(exports.TAG_ALERT + content);
}
exports.LogAlert = LogAlert;
function LogEmergency(content) {
    console.log(exports.TAG_EMERGENCY + content);
}
exports.LogEmergency = LogEmergency;
