// @port.finance/port-sdk/dist/cjs/instructions/lending/createReserve.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.initReserveInstruction = void 0;
const spl_token_1 = require("@solana/spl-token");
const web3_js_1 = require("@solana/web3.js");
const constants_1 = require("../../constants");
const ReserveData_1 = require("../../structs/ReserveData");
const instruction_1 = require("./instruction");
const BufferLayout = __importStar(require("@solana/buffer-layout"));
const Layout = __importStar(require("../../serialization/layout"));
const bn_js_1 = __importDefault(require("bn.js"));
const Instructions_1 = require("../../utils/Instructions");
// interface Data {
//   instruction: number;
//   liquidityAmount: bigint;
//   option: number;
//   marketPrice: BN;
//   config: ReserveConfigProto;
// }
// Initializes a new lending market reserve.
//
// Accounts expected by this instruction:
//
//   0. `[writable]` Source liquidity token account.
//                     $authority can transfer $liquidity_amount.
//   1. `[writable]` Destination collateral token account - uninitialized.
//   2. `[writable]` Reserve account - uninitialized.
//   3. `[]` Reserve liquidity SPL Token mint.
//   4. `[writable]` Reserve liquidity supply SPL Token account - uninitialized.
//   5. `[writable]` Reserve liquidity fee receiver - uninitialized.
//   6. `[writable]` Reserve collateral SPL Token mint - uninitialized.
//   7. `[writable]` Reserve collateral token supply - uninitialized.
//   8 `[]` Lending market account.
//   9 `[]` Derived lending market authority.
//   10 `[signer]` Lending market owner.
//   11 `[signer]` User transfer authority ($authority).
//   12 `[]` Clock sysvar.
//   13 `[]` Rent sysvar.
//   14 `[]` Token program id.
//   15 `[optional]` Oracle price account, pyth or switchboard.
//           This will be used as the reserve liquidity oracle account.
const initReserveInstruction = (liquidityAmount, option, price, config, sourceLiquidity, destinationCollateral, reserve, liquidityMint, liquiditySupply, liquidityFeeReceiver, pythPrice, collateralMint, collateralSupply, lendingMarket, lendingMarketAuthority, lendingMarketOwner, transferAuthority, lendingProgramId = constants_1.PORT_LENDING) => {
    const dataLayout = BufferLayout.struct([
        BufferLayout.u8("instruction"),
        Layout.uint64("liquidityAmount"),
        BufferLayout.u32("option"),
        Layout.uint128("marketPrice"),
        // eslint-disable-next-line new-cap
        (0, ReserveData_1.ReserveConfigProtoLayout)("config"),
    ]);
    const data = Buffer.alloc(dataLayout.span);
    dataLayout.encode({
        instruction: instruction_1.LendingInstruction.InitReserve,
        option,
        marketPrice: new bn_js_1.default(price),
        liquidityAmount: new bn_js_1.default(liquidityAmount),
        config: Object.assign({}, config),
    }, data);
    const keys = [
        (0, Instructions_1.getAccess)(sourceLiquidity, Instructions_1.AccessType.WRITE),
        (0, Instructions_1.getAccess)(destinationCollateral, Instructions_1.AccessType.WRITE),
        (0, Instructions_1.getAccess)(reserve, Instructions_1.AccessType.WRITE),
        (0, Instructions_1.getAccess)(liquidityMint, Instructions_1.AccessType.READ),
        (0, Instructions_1.getAccess)(liquiditySupply, Instructions_1.AccessType.WRITE),
        (0, Instructions_1.getAccess)(liquidityFeeReceiver, Instructions_1.AccessType.WRITE),
        (0, Instructions_1.getAccess)(collateralMint, Instructions_1.AccessType.WRITE),
        (0, Instructions_1.getAccess)(collateralSupply, Instructions_1.AccessType.WRITE),
        (0, Instructions_1.getAccess)(lendingMarket, Instructions_1.AccessType.READ),
        (0, Instructions_1.getAccess)(lendingMarketAuthority, Instructions_1.AccessType.READ),
        (0, Instructions_1.getAccess)(lendingMarketOwner, Instructions_1.AccessType.SIGNER),
        (0, Instructions_1.getAccess)(transferAuthority, Instructions_1.AccessType.SIGNER),
        (0, Instructions_1.getAccess)(web3_js_1.SYSVAR_CLOCK_PUBKEY, Instructions_1.AccessType.READ),
        (0, Instructions_1.getAccess)(web3_js_1.SYSVAR_RENT_PUBKEY, Instructions_1.AccessType.READ),
        (0, Instructions_1.getAccess)(spl_token_1.TOKEN_PROGRAM_ID, Instructions_1.AccessType.READ),
    ];
    if (pythPrice && option === 0) {
        keys.push((0, Instructions_1.getAccess)(pythPrice, Instructions_1.AccessType.READ));
    }
    return new web3_js_1.TransactionInstruction({
        keys,
        programId: lendingProgramId,
        data,
    });
};
exports.initReserveInstruction = initReserveInstruction;
