// @port.finance/port-sdk/dist/cjs/instructions/lending/liquidateObligation.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.liquidateObligationInstruction = void 0;
const web3_js_1 = require("@solana/web3.js");
const BufferLayout = __importStar(require("@solana/buffer-layout"));
const Layout = __importStar(require("../../serialization/layout"));
const bn_js_1 = __importDefault(require("bn.js"));
const spl_token_1 = require("@solana/spl-token");
const instruction_1 = require("./instruction");
const constants_1 = require("../../constants");
// Repay borrowed liquidity to a reserve to receive collateral at a discount from an unhealthy
// obligation. Requires a refreshed obligation and reserves.
//
// Accounts expected by this instruction:
//
//   0. `[writable]` Source liquidity token account.
//                     Minted by repay reserve liquidity mint.
//                     $authority can transfer $liquidity_amount.
//   1. `[writable]` Destination collateral token account.
//                     Minted by withdraw reserve collateral mint.
//   2. `[writable]` Repay reserve account - refreshed.
//   3. `[writable]` Repay reserve liquidity supply SPL Token account.
//   4. `[]` Withdraw reserve account - refreshed.
//   5. `[writable]` Withdraw reserve collateral supply SPL Token account.
//   6. `[writable]` Obligation account - refreshed.
//   7. `[]` Lending market account.
//   8. `[]` Derived lending market authority.
//   9. `[signer]` User transfer authority ($authority).
//   10 `[]` Clock sysvar.
//   11 `[]` Token program id.
const liquidateObligationInstruction = (liquidityAmount, sourceLiquidity, destinationCollateral, repayReserve, repayReserveLiquiditySupply, withdrawReserve, withdrawReserveCollateralSupply, obligation, lendingMarket, lendingMarketAuthority, transferAuthority, lendingProgramId = constants_1.PORT_LENDING, stakingPool, stakeAccount, stakingProgramId = constants_1.PORT_STAKING) => {
    const dataLayout = BufferLayout.struct([
        BufferLayout.u8("instruction"),
        Layout.uint64("liquidityAmount"),
    ]);
    const data = Buffer.alloc(dataLayout.span);
    dataLayout.encode({
        instruction: instruction_1.LendingInstruction.LiquidateObligation,
        liquidityAmount: new bn_js_1.default(liquidityAmount),
    }, data);
    let keys = [
        { pubkey: sourceLiquidity, isSigner: false, isWritable: true },
        { pubkey: destinationCollateral, isSigner: false, isWritable: true },
        { pubkey: repayReserve, isSigner: false, isWritable: true },
        { pubkey: repayReserveLiquiditySupply, isSigner: false, isWritable: true },
        { pubkey: withdrawReserve, isSigner: false, isWritable: false },
        {
            pubkey: withdrawReserveCollateralSupply,
            isSigner: false,
            isWritable: true,
        },
        { pubkey: obligation, isSigner: false, isWritable: true },
        { pubkey: lendingMarket, isSigner: false, isWritable: false },
        { pubkey: lendingMarketAuthority, isSigner: false, isWritable: false },
        { pubkey: transferAuthority, isSigner: true, isWritable: false },
        { pubkey: web3_js_1.SYSVAR_CLOCK_PUBKEY, isSigner: false, isWritable: false },
        { pubkey: spl_token_1.TOKEN_PROGRAM_ID, isSigner: false, isWritable: false },
    ];
    if (stakingPool !== undefined && stakeAccount !== undefined) {
        keys = keys.concat([
            { pubkey: stakeAccount, isSigner: false, isWritable: true },
            { pubkey: stakingPool, isSigner: false, isWritable: true },
            { pubkey: stakingProgramId, isSigner: false, isWritable: false },
        ]);
    }
    return new web3_js_1.TransactionInstruction({
        keys,
        programId: lendingProgramId,
        data,
    });
};
exports.liquidateObligationInstruction = liquidateObligationInstruction;
