// francium-sdk/dist/constants/auto/raydium.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AUTO_RAYDIUM_EXTRA_CONFIG = exports.AUTO_RAYDIUM_FARM_CONFIG = exports.LendingInfo = exports.MarketInfo = exports.FRANCIUM_AUTO_VAULTS_PROGRAM_ID = void 0;
const web3_js_1 = require("@solana/web3.js");
exports.FRANCIUM_AUTO_VAULTS_PROGRAM_ID = new web3_js_1.PublicKey("FVAUXEVi4qwpGCDSvb1wxMSJ6hXWzSPdfZqYXa19hVDz");
exports.MarketInfo = {
    MarketInfoAccount: "Ch15dYReFdK7Dewoz6eYAwXZeh9i4kqrKy3VBjsS5JZE",
    owner: "7MBLg6oV5phip11YBbJPuq7u38kdzSi9PM3BifKSpLaR",
    feeReceiver: "7MBLg6oV5phip11YBbJPuq7u38kdzSi9PM3BifKSpLaR",
};
exports.LendingInfo = {
    "programId": "FC81tbGt6JWRXidaWYFXxGnTk4VgobhJHATvTRVMqgWj",
    "marketInfoAccount": "4XNif294wbrxj6tJ8K5Rg7SuaEACnu9s2L27i28MQB6E",
    "marketAuthority": "sCDiYj7X7JmXg5fVq2nqED2q1Wqjo7PnqMgH3casMem",
    "marketOwner": "7MBLg6oV5phip11YBbJPuq7u38kdzSi9PM3BifKSpLaR",
};
exports.AUTO_RAYDIUM_FARM_CONFIG = {
    "L4x-SOLUSDC-RAY1": {
        pair: 'SOL-USDC',
        "tknMint0": "EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v",
        "tknMint1": "So11111111111111111111111111111111111111112",
        "lpMint": "8HoQnePLqPj4M7PUDzfw8e3Ymdwgc7NLGnaTUapubyvu",
        "rewardsMint": "4k3Dyjzvzp8eMZWUXbBCjEvwSkkk59S5iCNLY3QrkX6R",
        "rewardsMintB": "SRMuApVNdxXokk5GT7XD5cUUgXMBCoAz2LHeuAoKWRt",
        "poolInfoAccount": "6mEFMM9EzAmCoeG75syF3y9mYc8rCbycU5BS2hppPrDb",
        "poolAuthority": "A7yL7NrMWtnvmD5Zaup53pBBPVADDpX3t13PLZiSvcAL",
        "poolTknAccount0": "HqRmPF1bbtp7EzypYB9k81UsNoCV63izWrHTMP8wvDG6",
        "poolTknAccount1": "G3z5ouUivqwyqLhCoKzhNtjcwEBjnUwkfSKNnADbAnKU",
        "poolLpAccount": "9wTPKas9Ke8pvoP57r3zgv3dEcZ8D8d5Pi3iEenXCzYq",
        "poolBorrowingCreditTokenAccount0": "J8vwYWBU7ixoZWCqMHrkj1jiSKUVjd9D5eVKUYLbwSHx",
        "poolBorrowingCreditTokenAccount1": "2zqzatKmt8Td2sdvqjWVMRK2kJ9kLoGCMbFGfibDp8Gx",
        "poolRewardsTokenAccount": "3qSi6kagCi6kA1ppmzVU1EUFVMhsantgsNunV7iDvRno",
        "poolRewardsTokenAccountB": "DUPZ8DkGLHQwW8nmvh2uQcK9CEvuZEyjkVcG8bMSWwFG",
        "poolFarmInfo": "6J5JjqMeYjxU2yqCas1BVp7ZXUYpvdCzXs11sPhJzDE1",
        "lendingPool0": {
            "lendingPoolInfoAccount": "Hx6LbkMHe69DYawhPyVNs8Apa6tyfogfzQV6a7XkwBUU",
            "lendingPoolTknAccount": "CFp9kt8z3Epb1QSiEp3xA44KbSwuJxhFR3wQoerFqYS9",
            "lendingPoolFeeAccount": "6h6vVc42moDmRs7KahQpFs8YWL5GfWBrxT9UrsUvpApn",
            "lendingPoolShareMint": "62fDf5daUJ9jBz8Xtj6Bmw1bh1DvHn8AG4L9hMmxCzpu",
            "lendingPoolShareAccount": "Gyc1V1xbA9NjzuURE662ATw6W4AdhwvsL26yUnSGhbkz",
            "lendingPoolCreditMint": "CZ2s85dnuAVyGbRWBNauHZwj9oTV2i9xTvFwc1Cedr7E",
            "lendingPoolCreditAccount": "Bd71jEvypKFs8N5gByyhD2149tSqLy4PVAzFFShCTJTB",
        },
        "lendingPool1": {
            "lendingPoolInfoAccount": "EnZC8MB6QLTxwN1LNqCXYC7XMpvXqitnYQPf5y5AcQRn",
            "lendingPoolTknAccount": "GiJwrxQW5NQWWAva4rnDMXwHxfrhyUzwMX5to1pmCCgF",
            "lendingPoolFeeAccount": "wLq5xWzkgfH1T2bbByBoUFxKk5g4K2urQv8q71gCsbM",
            "lendingPoolShareMint": "92Zst3rEoJsyZUW2yNc2811GtkBwFrm3tqgxMTtFWuSB",
            "lendingPoolShareAccount": "Hyi1jMgc9MMEZyaG9ziL2PvT7zbeRNKhb5K1kj9bxz4x",
            "lendingPoolCreditMint": "7x61NrCzeekHwgW5Nk4afMg2sxjMK6RyMkx8tCzzX4qG",
            "lendingPoolCreditAccount": "WuPCgCuyLSXatdjyWqkGjCPSQFgKLwcfsNR9fDq3hVL",
        },
        "raydiumProgramId": "675kPX9MHTjS2zt1qfr1NYHuzeLXfQM9H24wFSUt1Mp8",
        "ammId": "58oQChx4yWmvKdwLLZzBi4ChoCc2fqCUWBkwMihLYQo2",
        "ammAuthority": "5Q544fKrFoe6tsEbD7S8EmxGTJYAKtTVhAW5Q5pge4j1",
        "ammOpenOrders": "HRk9CMrpq7Jn9sh7mzxE8CChHG8dneX9p475QKz4Fsfc",
        "ammTargetOrders": "CZza3Ej4Mc58MnxWA385itCC9jCo3L1D7zc3LKy1bZMR",
        "ammCoinAccount": "DQyrAcCrDXQ7NeoqGgDCZwBvWDcYmFCjSb9JtteuvPpz",
        "ammPcAccount": "HLmqeL62xR1QoZ1HKKbXRrdN1p3phKpxRMb2VVopvBBz",
        "poolWithdrawQueue": "G7xeGGLevkRwB5f44QNgQtrPKBdMfkT6ZZwpS9xcC97n",
        "poolTempLpTokenAccount": "Awpt6N7ZYPBa4vG4BQNFhFxDj4sxExAA9rpBAoBw2uok",
        "serumProgramId": "9xQeWvG816bUx9EPjHmaT23yvVM2ZWbrrpZb9PusVFin",
        "serumMarketId": "9wFFyRfZBsuAha4YcuxcXLKwMxJR43S7fPfQLusDBzvT",
        "serumBids": "14ivtgssEBoBjuZJtSAPKYgpUK7DmnSwuPMqJoVTSgKJ",
        "serumAsks": "CEQdAFKdycHugujQg9k2wbmxjcpdYZyVLfV9WerTnafJ",
        "serumEventQueue": "5KKsLVU6TcbVDK4BS6K1DGDxnh4Q9xjYJ8XaDCG5t8ht",
        "serumCoinVault": "36c6YqAwyGKQG66XEp2dJc5JqjaBNv7sVghEtJv4c7u6",
        "serumPCVault": "8CFo8bL8mZQK8abbFyypFMwEDd8tVJjHTTojMLgQTUSZ",
        "serumVaultSigner": "F8Vyqk3unwxkXukZFQeYyGmFfTG3CAX4v24iyrjEYBJV",
        "stakePoolProgramId": "9KEPoZmtHUrBbhWN1v1KWLMkkvwY6WLtAVUCPRtRjP4z",
        "stakePoolId": "GUzaohfNuFbBqQTnPgPSNciv3aUvriXYjQduRE3ZkqFw",
        "stakePoolAuthority": "DgbCWnbXg43nmeiAveMCkUUPEpAr3rZo3iop3TyP6S63",
        "stakePoolLpAccount": "J6ECnRDZEXcxuruvErXDWsPZn9czowKynUr9eDSQ4QeN",
        "stakePoolRewardAccount": "38YS2N7VUb856QDsXHS1h8zv5556YgEy9zKbbL2mefjf",
        "stakePoolRewardAccountB": "ANDJUfDryy3jY6DngwGRXVyxCJBT5JfojLDXwZYSpnEL",
        "ammIdForRewards": "6UmmUiYoBjSrhakAobJw8BvkmJtDVxaeBtbt7rxWo1mg",
        "ammIdForRewardsB": "8tzS7SkUZyHPQY7gLqsMCXZ5EDCgjESUHcB17tiR1h3Z",
    }
};
exports.AUTO_RAYDIUM_EXTRA_CONFIG = {
    "L4x-SOLUSDC-RAY1": {
        priceKey: 'SOL-USDC[Raydium]',
        lpDecimals: 9,
        leverage: 4,
        rebalanceType: 'delta',
        threshold: 4,
        cap: 10000000
    }
};
